---
title: "Compare pairwise dN/dS for clades affected by GC jumps"
output: html_notebook
---

### Background
Previously, for a subset of datasets, I have tried estimating dN, dS using a non-stationary codon model. However, we now uspect that this model does not fit the data well leading to biased estimates of dN,dS on long branches. Therefore, I will now attempt an alternate approach. I will try to test for changes in other genomic proxies which can give indication of evolutionary forces acting on the genomes. These include- 
1. dN/dS (inceased dN/dS indicates increased drift)
2. CUB (decreased CUb indicates decreased Ne.s)
3. rate of 16s rRNA evolution (increased rate indicates increased drift/mutation)
4. genome size (reductions can be due to increased drift or increased streamlining selection)
5. coding density (decrease indicates drift, increase indicates streamlinign selection)
6. length of intergenic region (decrease indicates increased streamlining selection)

estimate pairwise dN/dS within taxa of the focal clades (affected by jumps) and compare it with dN/dS of pairs in the corresponding sister or related clades.

#### Common data and input processing
```{r common_code }
library(parallel)

library(phytools)
library(rncl)
library(seqinr)

library(tidyverse)
library(reshape2)

library(cowplot)
library(viridis)
library(ggpubr)
library(patchwork)
library(dendextend)

n_cores_av <- 3

main_dir <- "../"
input_dir <- "../Input_data/"
levolution_data_dir <- "../Input_data/levolution_files_inference_actual_data/"
levolution_out_dir <- "/Results/macrodynamics/GTDB_levolution/"
jump_common_dir <- "/Results/jump_analysis/"

GTDB_14k_tree <- read_newick_phylo(file = paste0(main_dir, "Input_data/phylogenies/GTDB_bac120_r80_pruned0.01.newick"))
GTDB_14k_metadata <- read.delim(file = paste0(main_dir, "Input_data/phylogenies/GTDB_bac_metadata_r80_pruned0.01.tsv"), header = T, row.names = 1, stringsAsFactors = F)

levolution_best_fit_parameters <- read.table(file = paste0(levolution_data_dir, "levolution_best_fit_parameters_summary.txt"), row.names = 1, header = T, stringsAsFactors = F)

levolution_all_focal_clade_details <- read.csv(file = paste0(input_dir, "AS1.csv"), header = T, stringsAsFactors = F)

gd_base <- paste0(main_dir, "Input_data/GTDB_data/GTDB_")
fn_types <- c("ass_fn", "ft_fn", "cds_fn", "rna_fn", "ftcnt_fn", "fna_fn", "gb_fn", "prot_fn")
gdirs_oi <- c("assembly_reports/", "feature_tables/", "cds/", "rna/", "feature_counts/", "genomes/", "gb/", "proteins/")
names(gdirs_oi) <- fn_types

download_script <- paste0(main_dir, "Codes/download_all_genome_files_from_GCFonly.py")
source(file = paste0(main_dir, "Codes/plotTree.wBars.R"))

OLclades_oi <- c("Sphingomonadale", "Rhizobiale", "Rhodobacterale", "Alpha1", "Bacteroidale", "Cytophagale", "Flavobacteriale","Pseudomonadale", "Enterobacterales", "Betaproteo")
p_folders <- c(rep("Alphaproteo",4), rep("Bacteroidetes",3), rep("Gammaproteo", 3))
pp_cutoffs <- c(0.75,0.75,0.75,0.95,0.95,0.95,0.75, 0.9,0.9,0.9)
alphas_oi <- c(1,1,1,0.5,0.25,0.25,1,0.25,0.5,0.5)
config_oi <- 1

get_genera_sp_names <- function(full_GTDB_taxonomy){
  GTDB_fields <- strsplit(x = full_GTDB_taxonomy, split = ";g__")[[1]]
  return(paste0("g__", GTDB_fields[2]))
}

clade_size_cutoff <- 20

cbPalette <- c("#999999", "#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7")
##colorBlind_friendly_palette= black, orange, sky blue, bluish green, yellow, blue, vermillion, reddish purple
```


### Filtering datasets according to available clades and number of taxa
I will first find which focal jumps have appropriate sister groups and outgroup clades for further analysis. Clades should have at least a sister group for dN/dS analysis, whereas at least one outgroup for analysis based on branch lengths.

I will also ignore focal jumps where sister or outgroup clades have more than 20 taxa, in order to make the downstream analysis more reliable and simpler. A large number of taxa in sister or outgroup clade increases the chances of heterogeneity in these groups used for comparison.
```{r levolution_jumpClades_filtering }
for(OLclade_n in 1:length(OLclades_oi)){
  OLclade_oi <- OLclades_oi[OLclade_n]
  print(paste0("Analysing jumps within ", OLclade_oi))
  pp_cutoff <- pp_cutoffs[OLclade_n]
  alpha_oi <- alphas_oi[OLclade_n]
  p_folder <- p_folders[OLclade_n]
  res_dir <- out_dir <-  paste0(levolution_out_dir ,p_folder,"_GTDB_14k_levolution/")

  tree_pp <- read.newick(file = paste0(res_dir,OLclade_oi,"_GTDB_14k_GC_config", config_oi, "_alpha", alpha_oi,".post"))
  tree_in <- read.newick(paste0(res_dir,OLclade_oi,"_GTDB_14k_tree.newick"))
  OLclade_nTips <- length(tree_in$tip.label) ##OL=OrderLevel
  
  OLclade_jumpBranches <- which(tree_pp$edge.length>pp_cutoff)
  
  # jumpClade_details_colNames <- c("jumpBranch", "depth", "nTaxa_focal", "nTaxa_sis", "nTaxa_out1", "nTaxa_out2")
  # jumpClade_details <- data.frame(matrix(NA, nrow = length(OLclade_jumpBranches), ncol = length(jumpClade_details_colNames), dimnames = list(paste0("jump", 1:length(OLclade_jumpBranches)), jumpClade_details_colNames)))
  jumpClade_details <- read.table(file = paste0(jump_common_dir, OLclade_oi, "_jumpSummaries4.tab"), stringsAsFactors = F)
  
  for(jumpNum_oi in 1:length(OLclade_jumpBranches)){
    print(jumpNum_oi)
    jumpDir <- paste0(jump_common_dir, OLclade_oi, "_jump", jumpNum_oi, "/")
    system2(command = "mkdir", args = c(jumpDir), stdout = F, stderr = F)

    jumpBranch_oi <- OLclade_jumpBranches[jumpNum_oi]

    anc_node <- tree_in$edge[jumpBranch_oi,1] #parent node of jump branch
    focal_root <- tree_in$edge[jumpBranch_oi,2] #descendant node of jump branch
    focal_descendants <- getDescendants(tree=tree_in, node=focal_root)
    focal_tips <- intersect(1:OLclade_nTips, focal_descendants)
    focal_tip_labels <- tree_in$tip.label[focal_tips]
    write.table(focal_tip_labels, file = paste0(jumpDir, OLclade_oi, "_jump", jumpNum_oi, "_focalGTDBLabels.tab"), row.names = F, col.names = F)
    nTaxa_focal <- length(focal_tips)
    
    ##check if there are any other, nested jumps in this focal clade
    all_nestedFocalTips <- c()
    nested_jumpBranches <- setdiff(intersect(OLclade_jumpBranches, which(tree_in$edge[,2] %in% focal_descendants)), jumpBranch_oi)
    nested_focalNodes <- sapply(nested_jumpBranches, function(x) getDescendants(tree_in, tree_in$edge[x,2]))
    nested_focalTips <- intersect(unlist(nested_focalNodes), 1:length(tree_in$tip.label))
    all_nestedFocalTips <- nested_focalTips

    sis_root <- setdiff(tree_in$edge[which(tree_in$edge[,1]==anc_node), 2], focal_root) #the other descendant of ancestral node
    sis_branch <- intersect(which(tree_in$edge[,1]==anc_node),which(tree_in$edge[,2]==sis_root))
    sis_descendants <- getDescendants(tree = tree_in, node = sis_root)
    sis_tips <- intersect(1:length(tree_in$tip.label), sis_descendants)
    sis_tip_labels <- tree_in$tip.label[sis_tips]
    write.table(sis_tip_labels, file = paste0(jumpDir, OLclade_oi, "_jump", jumpNum_oi, "_sisGTDBLabels.tab"), row.names = F, col.names = F)
    nTaxa_sis <- length(sis_tips)
    
    ##check if there are any additional, nested jumps in the sister clade
    nested_jumpBranches <- intersect(OLclade_jumpBranches, which(tree_in$edge[,2] %in% c(sis_root, sis_descendants)))
    nested_focalNodes <- sapply(nested_jumpBranches, function(x) getDescendants(tree_in, tree_in$edge[x,2]))
    nested_focalTips <- intersect(unlist(nested_focalNodes), 1:length(tree_in$tip.label))
    all_nestedFocalTips <- c(all_nestedFocalTips, nested_focalTips)

    anc_out1 <- tree_in$edge[which(tree_in$edge[,2]==anc_node),1] #parent of ancestral node
    
    if(identical(anc_out1, numeric(0))){ #if ancestor of outgroup cant be found, most likely because focal clade diverges from the root, so there is no outgroup
      print("cant find outgroup! Is focal clade at the root?")
      if(all(c(nTaxa_focal, nTaxa_sis) <= clade_size_cutoff)){
        jumpClade_details[paste0("jump", jumpNum_oi), ] <- c(jumpBranch_oi, "sister", nTaxa_focal, nTaxa_sis, NA, NA)
      }else{
        jumpClade_details[paste0("jump", jumpNum_oi), ] <- c(jumpBranch_oi, "focal", nTaxa_focal, nTaxa_sis, NA, NA)
      }
      write.table(tree_in$tip.label[all_nestedFocalTips], file = paste0(jumpDir, "nestedFocalTips.tab"), quote = F, row.names = F, col.names = F)
      
      next #skip this jump for remaining analysis
    }
    
    out1_root <- setdiff(tree_in$edge[which(tree_in$edge[,1]==anc_out1),2],anc_node)
    out1_branch <- intersect(which(tree_in$edge[,1]==anc_out1),which(tree_in$edge[,2]==out1_root))
    out1_descendants <- getDescendants(tree=tree_in, node=out1_root)
    out1_tips <- intersect(1:length(tree_in$tip.label), out1_descendants)
    out1_tip_labels <- tree_in$tip.label[out1_tips]
    write.table(out1_tip_labels, file = paste0(jumpDir, OLclade_oi, "_jump", jumpNum_oi, "_out1GTDBLabels.tab"), row.names = F, col.names = F)
    nTaxa_out1 <- length(out1_tips)
    
    ##check if there are any additional, nested jumps in the outgroup clades
    nested_jumpBranches <- intersect(OLclade_jumpBranches, which(tree_in$edge[,2] %in% c(out1_root, out1_descendants)))
    nested_focalNodes <- sapply(nested_jumpBranches, function(x) getDescendants(tree_in, tree_in$edge[x,2]))
    nested_focalTips <- intersect(unlist(nested_focalNodes), 1:length(tree_in$tip.label))
    all_nestedFocalTips <- c(all_nestedFocalTips, nested_focalTips)
    
    anc_out2 <- tree_in$edge[which(tree_in$edge[,2]==anc_out1),1] #parent of ancestor of first outgroup
    if(identical(anc_out2, numeric(0))){ #if ancestor of 2nd outgroup cant be found, most likely because 1st outgroup diverges from root, so there is no other outgroup
      print("cant find 2nd outgroup! Is 1st outgroup at the root?")
      
      ##if focal, sister and outgroup1 have limited taxa
      if(all(c(nTaxa_focal, nTaxa_sis, nTaxa_out1)  <=  clade_size_cutoff)){
        #extract focal_clade
        focal_clade_tree <- extract.clade(tree_in, anc_out1)
        jumpClade_details[paste0("jump", jumpNum_oi), ] <- c(jumpBranch_oi, "outgroup1", nTaxa_focal, nTaxa_sis, nTaxa_out1, NA)
        write.table(tree_in$tip.label[all_nestedFocalTips], file = paste0(jumpDir, "nestedFocalTips.tab"), quote = F, row.names = F, col.names = F)
      }else if(all(c(nTaxa_focal, nTaxa_sis) <= clade_size_cutoff)){
        # focal_clade_tree <- extract.clade(tree_in, anc_node)
        # jumpClade_details[paste0("jump", jumpNum_oi), ] <- c(jumpBranch_oi, "sister", nTaxa_focal, nTaxa_sis, nTaxa_out1, NA)
        # write.table(tree_in$tip.label[all_nestedFocalTips], file = paste0(jumpDir, "nestedFocalTips.tab"), quote = F, row.names = F, col.names = F)
        print("Either focal, sister, or outgroup clade is too large. Will not analyze this jump directly.")
        jumpClade_details[paste0("jump", jumpNum_oi), ] <- c(jumpBranch_oi, "outgroup1", nTaxa_focal, nTaxa_sis, nTaxa_out1, NA)
        write.table(tree_in$tip.label[all_nestedFocalTips], file = paste0(jumpDir, "nestedFocalTips.tab"), quote = F, row.names = F, col.names = F)
        too_large <- T
      }else{
        print("Either focal, sister, or outgroup clade is too large. Will not analyze this jump directly.")
        jumpClade_details[paste0("jump", jumpNum_oi), ] <- c(jumpBranch_oi, "outgroup1", nTaxa_focal, nTaxa_sis, nTaxa_out1, NA)
        write.table(tree_in$tip.label[all_nestedFocalTips], file = paste0(jumpDir, "nestedFocalTips.tab"), quote = F, row.names = F, col.names = F)
        too_large <- T
      } ##end of if, focal, sister, and outgroup1 have limited taxa
      
      if(exists("too_large")) next ##skip this jump if focal or sister clades are too large #these are dealt with separately
      
      #if focal, sister, and first outgroup are small, identify edges for coloring
      if(nTaxa_focal == 1){
        focal_edges <- which(focal_clade_tree$edge[,2]==which(focal_clade_tree$tip.label==focal_tip_labels))
      }else if(nTaxa_focal > 1){
        clade_tree_focal_root <- getMRCA(focal_clade_tree, focal_tip_labels)
        focal_edges <- which(focal_clade_tree$edge[,2] %in% c(clade_tree_focal_root, getDescendants(focal_clade_tree, clade_tree_focal_root)))
      }
  
      if(nTaxa_sis == 1){
        sis_edges <- which(focal_clade_tree$edge[,2]==which(focal_clade_tree$tip.label==sis_tip_labels))
      }else if(nTaxa_sis > 1){
        clade_tree_sis_root <- getMRCA(focal_clade_tree, sis_tip_labels)
        sis_edges <- which(focal_clade_tree$edge[,2] %in% c(clade_tree_sis_root, getDescendants(focal_clade_tree, clade_tree_sis_root)))
      }
      
      edge_colors <- rep("black", length(focal_clade_tree$edge.length))
      edge_colors[focal_edges] <- "orange"
      edge_colors[sis_edges] <- "purple"
  
      #change labels to species names for plotting
      focal_clade_tree_labeled <- focal_clade_tree
      focal_clade_tree_labeled$tip.label <- sapply(focal_clade_tree$tip.label, function(x) get_genera_sp_names(GTDB_14k_metadata[x,"gtdb_taxonomy"]), simplify = T, USE.NAMES = F)
      
      #plot tree for reference
      svg(filename = paste0(jumpDir, OLclade_oi, "_jump", jumpNum_oi, "_cladeTree.svg"), width = max(max(nodeHeights(focal_clade_tree))*50,5), height = max(length(focal_clade_tree$tip.label)*0.5, 3))
      plot(focal_clade_tree, edge.color = edge_colors, edge.width = 2)
      add.scale.bar()
      dev.off()
      
      svg(filename = paste0(jumpDir, OLclade_oi, "_jump", jumpNum_oi, "_cladeTree_labeled.svg"), width = max(max(nodeHeights(focal_clade_tree_labeled))*50,5), height = max(length(focal_clade_tree_labeled$tip.label)*0.5, 5))
      plot(focal_clade_tree_labeled, edge.color = edge_colors, edge.width = 2)
      add.scale.bar()
      dev.off()
      
      write.tree(focal_clade_tree, file = paste0(jumpDir, OLclade_oi, "_jump", jumpNum_oi, ".newick"))
      write.tree(focal_clade_tree_labeled, file = paste0(jumpDir, OLclade_oi, "_jump", jumpNum_oi, "_labeled.newick"))
      
      next #skip this jump now
    }##end of if, outgroup 2 not found
    
    ##if outgroup2 is found, we continue
    out2_root <- setdiff(tree_in$edge[which(tree_in$edge[,1]==anc_out2),2],anc_out1)
    out2_branch <- intersect(which(tree_in$edge[,1]==anc_out2),which(tree_in$edge[,2]==out2_root))
    out2_descendants <- getDescendants(tree=tree_in, node=out2_root)
    out2_tips <- intersect(1:length(tree_in$tip.label), out2_descendants)
    out2_tip_labels <- tree_in$tip.label[out2_tips]
    write.table(out2_tip_labels, file = paste0(jumpDir, OLclade_oi, "_jump", jumpNum_oi, "_out2GTDBLabels.tab"), row.names = F, col.names = F)
    nTaxa_out2 <- length(out2_tips)
    
    ##check if there are any additional, nested jumps in the outgroup2 clades
    nested_jumpBranches <- intersect(OLclade_jumpBranches, which(tree_in$edge[,2] %in% c(out2_root, out2_descendants)))
    nested_focalNodes <- sapply(nested_jumpBranches, function(x) getDescendants(tree_in, tree_in$edge[x,2]))
    nested_focalTips <- intersect(unlist(nested_focalNodes), 1:length(tree_in$tip.label))
    all_nestedFocalTips <- c(all_nestedFocalTips, nested_focalTips)
    write.table(tree_in$tip.label[all_nestedFocalTips], file = paste0(jumpDir, "nestedFocalTips.tab"), quote = F, row.names = F, col.names = F)
    
    if(all(c(nTaxa_focal, nTaxa_sis, nTaxa_out1, nTaxa_out2) <=  clade_size_cutoff)){
      focal_clade_tree <- extract.clade(tree_in, anc_out2)
      jumpClade_details[paste0("jump", jumpNum_oi), ] <- c(jumpBranch_oi, "outgroup2", nTaxa_focal, nTaxa_sis, nTaxa_out1, nTaxa_out2)
    }else if(all(c(nTaxa_focal, nTaxa_sis, nTaxa_out1) <=  clade_size_cutoff)){
      focal_clade_tree <- extract.clade(tree_in, anc_out1)
      jumpClade_details[paste0("jump", jumpNum_oi), ] <- c(jumpBranch_oi, "outgroup1", nTaxa_focal, nTaxa_sis, nTaxa_out1, nTaxa_out2)
    }else if(all(c(nTaxa_focal, nTaxa_sis)  <=  clade_size_cutoff)){
      print("Either focal, sister, or outgroup clade is too large. Will not analyze this jump directly.")
      jumpClade_details[paste0("jump", jumpNum_oi), ] <- c(jumpBranch_oi, "outgroup1", nTaxa_focal, nTaxa_sis, nTaxa_out1, NA)
      write.table(tree_in$tip.label[all_nestedFocalTips], file = paste0(jumpDir, "nestedFocalTips.tab"), quote = F, row.names = F, col.names = F)
      too_large <- T
    }else{
      print("Either focal, sister, or outgroup clade is too large. Will not analyze this jump directly.")
      jumpClade_details[paste0("jump", jumpNum_oi), ] <- c(jumpBranch_oi, "outgroup1", nTaxa_focal, nTaxa_sis, nTaxa_out1, NA)
      write.table(tree_in$tip.label[all_nestedFocalTips], file = paste0(jumpDir, "nestedFocalTips.tab"), quote = F, row.names = F, col.names = F)
      too_large <- T
    }
    
    if(exists("too_large")) next ##skip this jump if focal or sister clades are too large #these are dealt with separately
    
    ##idenityf edges for coloring
    if(nTaxa_focal == 1){
      focal_edges <- which(focal_clade_tree$edge[,2]==which(focal_clade_tree$tip.label==focal_tip_labels))
    }else if(nTaxa_focal > 1){
      clade_tree_focal_root <- getMRCA(focal_clade_tree, focal_tip_labels)
      focal_edges <- which(focal_clade_tree$edge[,2] %in% c(clade_tree_focal_root, getDescendants(focal_clade_tree, clade_tree_focal_root)))
    }
    
    if(nTaxa_sis == 1){
      sis_edges <- which(focal_clade_tree$edge[,2]==which(focal_clade_tree$tip.label==sis_tip_labels))
    }else if(nTaxa_sis > 1){
      clade_tree_sis_root <- getMRCA(focal_clade_tree, sis_tip_labels)
      sis_edges <- which(focal_clade_tree$edge[,2] %in% c(clade_tree_sis_root, getDescendants(focal_clade_tree, clade_tree_sis_root)))
    }
    
    edge_colors <- rep("black", length(focal_clade_tree$edge.length))
    edge_colors[focal_edges] <- "orange"
    edge_colors[sis_edges] <- "purple"

    #change labels to species names for plotting
    focal_clade_tree_labeled <- focal_clade_tree
    focal_clade_tree_labeled$tip.label <- sapply(focal_clade_tree$tip.label, function(x) get_genera_sp_names(GTDB_14k_metadata[x,"gtdb_taxonomy"]), simplify = T, USE.NAMES = F)
    
    #plot tree for reference
    svg(filename = paste0(jumpDir, OLclade_oi, "_jump", jumpNum_oi, "_cladeTree.svg"), width = max(max(nodeHeights(focal_clade_tree))*50,5), height = max(length(focal_clade_tree$tip.label)*0.5, 3))
    plot(focal_clade_tree, edge.color = edge_colors, edge.width = 2)
    add.scale.bar()
    dev.off()
    
    svg(filename = paste0(jumpDir, OLclade_oi, "_jump", jumpNum_oi, "_cladeTree_labeled.svg"), width = max(max(nodeHeights(focal_clade_tree_labeled))*50,5), height = max(length(focal_clade_tree_labeled$tip.label)*0.5, 5))
    plot(focal_clade_tree_labeled, edge.color = edge_colors, edge.width = 2)
    add.scale.bar()
    dev.off()
    
    write.tree(focal_clade_tree, file = paste0(jumpDir, OLclade_oi, "_jump", jumpNum_oi, ".newick"))
    write.tree(focal_clade_tree_labeled, file = paste0(jumpDir, OLclade_oi, "_jump", jumpNum_oi, "_labeled.newick"))
  }#finish looping over jumps within a order-level clade
  write.table(jumpClade_details, file = paste0(jump_common_dir, OLclade_oi, "_jumpSummaries4.tab"), quote = F)
}#finish looping over order-level clades
```

### Downloading required genomes from filtered clades
```{r download_genomes, eval=F }
for(OLclade_n in 1:length(OLclades_oi)){
  OLclade_oi <- OLclades_oi[OLclade_n]
  print(paste0("Downloading genomes within ", OLclade_oi))
  pp_cutoff <- pp_cutoffs[OLclade_n]
  alpha_oi <- alphas_oi[OLclade_n]
  p_folder <- p_folders[OLclade_n]
  res_dir <- out_dir <-  paste0(levolution_out_dir,p_folder,"_GTDB_14k_levolution/")

  tree_pp <- read.newick(file = paste0(res_dir,OLclade_oi,"_GTDB_14k_GC_config", config_oi, "_alpha", alpha_oi,".post"))
  tree_in <- read.newick(paste0(res_dir,OLclade_oi,"_GTDB_14k_tree.newick"))
  OLclade_nTips <- length(tree_in$tip.label) ##OL=OrderLevel
  
  OLclade_jumpBranches <- which(tree_pp$edge.length>pp_cutoff)
  jumpClade_details <- read.table(file = paste0(jump_common_dir, OLclade_oi, "_jumpSummaries4.tab"), stringsAsFactors = F)
  jumpClade_details$genomes_downloaded <- NA
  
  for(jumpNum_oi in which(jumpClade_details$depth %in% c("sister", "outgroup1", "outgroup2"))){
    print(jumpNum_oi)
    jumpBranch_oi <- OLclade_jumpBranches[jumpNum_oi]
    
    if((jumpClade_details[paste0("jump", jumpNum_oi),"depth"]=="outgroup2" & all(jumpClade_details[paste0("jump", jumpNum_oi),c("nTaxa_focal", "nTaxa_sis", "nTaxa_out1", "nTaxa_out2")]<=(clade_size_cutoff), na.rm = T))  || (jumpClade_details[paste0("jump", jumpNum_oi),"depth"]=="outgroup1" & all(jumpClade_details[paste0("jump", jumpNum_oi),c("nTaxa_focal", "nTaxa_sis", "nTaxa_out1")]<=(clade_size_cutoff), na.rm = T)) || (jumpClade_details[paste0("jump", jumpNum_oi),"depth"]=="sister" & all(jumpClade_details[paste0("jump", jumpNum_oi),c("nTaxa_focal", "nTaxa_sis")]<=(clade_size_cutoff), na.rm = T))){
      jumpClade_details[paste0("jump", jumpNum_oi),"genomes_downloaded"] <- T
    }else{
      print("Either clade has too many taxa, cant analyze results")
      jumpClade_details[paste0("jump", jumpNum_oi),"genomes_downloaded"] <- F
      next #skip this jump
    }
    
    jumpDir <- paste0(jump_common_dir, OLclade_oi, "_jump", jumpNum_oi, "/")
    focal_clade_tree <- read.tree(paste0(jumpDir, OLclade_oi, "_jump", jumpNum_oi, ".newick"))
    
    if(file.exists(paste0(jumpDir,"ass_fn_GCFids_found.txt"))){
      print("The genomes are already downloaded!")
      next ##skip this jump
    }else{
      write.table(x = focal_clade_tree$tip.label, file = paste0(jumpDir, "required_GCFonly_ids.txt"), append = F, quote = F, row.names = F, col.names = F)
    
      for(fn_type in fn_types){
        system2(command = "python", args = c(download_script, "-wd", jumpDir, "-list_fn", "required_GCFonly_ids.txt", "-fn_type", fn_type, "-od", paste0(gd_base, gdirs_oi[fn_type]), "-sd", jumpDir), stdout = paste0(jumpDir,"download.log"), stderr = paste0(jumpDir,"download.error"))
      }#finish downloading files
    }
  }#finish looping over jumps within a order-level clade
  write.table(jumpClade_details, file = paste0(jump_common_dir, OLclade_oi, "_jumpSummaries4.tab"), quote = F)
}#finish looping over order-level clades
```

### dN/dS analysis
Here I will analyze if there are any changes in the dN/dS ratio between the focal and related clades. I will use the YN00 method to estimated dN/dS.
#### A function required to extract orthologs by reciprocal BLAST best hits method
```{r rbbs_function }
get_rbbs <- function(ref_genome, GCF_oi){
  shared_ortholog_list <- rep(NA, length(ref_HEG_prots))
  shared_ortholog_list <- setNames(shared_ortholog_list,names(ref_HEG_prots))
  
  blast_df <- read.csv(file = paste0(gd_base,"blast_op/",ref_genome,"_HEG_",GCF_oi,"_blast_op.dat"), header = F, col.names = col_names, stringsAsFactors = F)
  
  valid_hits_df <- blast_df[which(blast_df$length/blast_df$qlen > 0.75),]
  qacc_unique <- unique(valid_hits_df$qacc)
  rblast_df <- read.csv(file = paste0(gd_base,"blast_op/",GCF_oi,"_",ref_genome,"_HEG_blast_op.dat"), header = F, col.names = col_names, stringsAsFactors = F)
  rvalid_hits_df <- rblast_df[which(rblast_df$length/rblast_df$qlen > 0.75),]
  
  #identify hit with best e-value, this is usually the first hit in the list,
  #so I will just identify first occurence of query in the list of query-hit pairs
  f_best_hits <- sapply(qacc_unique, function(x) valid_hits_df[which(valid_hits_df$qacc==x)[1],"sacc"])
  for(q_oi in names(f_best_hits)){
    hit_oi <- f_best_hits[q_oi]
    if(any(grepl(pattern = hit_oi, x = rvalid_hits_df$qacc))){
      if(grepl(pattern = rvalid_hits_df$sacc[grep(pattern = hit_oi, x = rvalid_hits_df$qacc)[1]],q_oi)){
        shared_ortholog_list[[q_oi]] <- rvalid_hits_df$qacc[grep(pattern = hit_oi, x = rvalid_hits_df$qacc)[1]]
      }
    }
  }
  return(shared_ortholog_list)
}
```

#### Delete ortholog sequence files
A snippet to delete existing sequence files corresponding to orthologs. This is required during repeat analysis in order to avoid appending same sequences multiple times.
```{r delete_ortholog_files, eval=F, echo=F}
for(OLclade_n in 1:length(OLclades_oi)){
  OLclade_oi <- OLclades_oi[OLclade_n]
  print(paste0("Finding orthologs within ", OLclade_oi))
  pp_cutoff <- pp_cutoffs[OLclade_n]
  alpha_oi <- alphas_oi[OLclade_n]
  p_folder <- p_folders[OLclade_n]
  res_dir <- out_dir <-  paste0(levolution_out_dir,p_folder,"_GTDB_14k_levolution/")

  tree_pp <- read.newick(file = paste0(res_dir,OLclade_oi,"_GTDB_14k_GC_config", config_oi, "_alpha", alpha_oi,".post"))
  tree_in <- read.newick(paste0(res_dir,OLclade_oi,"_GTDB_14k_tree.newick"))
  OLclade_nTips <- length(tree_in$tip.label) ##OL=OrderLevel
  
  OLclade_jumpBranches <- which(tree_pp$edge.length>pp_cutoff)
  jumpClade_details <- read.table(file = paste0(jump_common_dir, OLclade_oi, "_jumpSummaries4.tab"), stringsAsFactors = F)
  
  for(jumpNum_oi in which(jumpClade_details$yn00_analysis==T)){
    print(jumpNum_oi)
    jumpBranch_oi <- OLclade_jumpBranches[jumpNum_oi]
    
    jumpDir <- paste0(jump_common_dir, OLclade_oi, "_jump", jumpNum_oi, "/")
    cds_aln_dir <- paste0(jumpDir,"cds_alignments/")
    
    # if(!file.exists(paste0(cds_aln_dir,"ref_orthologs_list.txt"))){
      # next
    # }
    shared_orthologs <- read.table(file = paste0(cds_aln_dir,"ref_orthologs_list.txt"), stringsAsFactors = F, header = F)
    shared_orthologs <- shared_orthologs$V1
    for(ortho_oi in shared_orthologs){
      system2(command = "rm", args = c(paste0(cds_aln_dir,ortho_oi,"_prot.fas")))
      system2(command = "rm", args = c(paste0(cds_aln_dir,ortho_oi,"_cds.fas")))
      
    }
  }
}
```

#### Extract single copy orthologs
Here the strategy is to identify orthologs of highly expressed genes (e.g. ribosomal protein, elongation factor Tu and RNA polymerase genes) based on reciprocal BLAST best hits. I start with HEGs from a well-annotated reference genome and look for reciprocal best hits in other genomes.

```{r extract_HEG_sequences }
for(OLclade_n in 1:length(OLclades_oi)){
  OLclade_oi <- OLclades_oi[OLclade_n]
  print(paste0("Finding orthologs within ", OLclade_oi))
  pp_cutoff <- pp_cutoffs[OLclade_n]
  alpha_oi <- alphas_oi[OLclade_n]
  p_folder <- p_folders[OLclade_n]
  res_dir <- out_dir <-  paste0(levolution_out_dir,p_folder,"_GTDB_14k_levolution/")

  tree_pp <- read.newick(file = paste0(res_dir,OLclade_oi,"_GTDB_14k_GC_config", config_oi, "_alpha", alpha_oi,".post"))
  tree_in <- read.newick(paste0(res_dir,OLclade_oi,"_GTDB_14k_tree.newick"))
  OLclade_nTips <- length(tree_in$tip.label) ##OL=OrderLevel
  
  OLclade_jumpBranches <- which(tree_pp$edge.length>pp_cutoff)
  jumpClade_details <- read.table(file = paste0(jump_common_dir, OLclade_oi, "_jumpSummaries4.tab"), stringsAsFactors = F)
  jumpClade_details$yn00_analysis <- NA
  
  for(jumpNum_oi in which(jumpClade_details$genomes_downloaded==T)){
    print(jumpNum_oi)
    
    ##This was implemented to reanalyze only those clades that were left out earlier
    # if(jumpClade_details[paste0("jump",jumpNum_oi), "yn00_analysis"]==T){
      # print("This clade was already analyzed!")
      # next
    # }

    jumpBranch_oi <- OLclade_jumpBranches[jumpNum_oi]
    
    jumpDir <- paste0(jump_common_dir, OLclade_oi, "_jump", jumpNum_oi, "/")
    
    all_accNums <- read.table(paste0(jumpDir, "ass_fn_GCFids_found.txt"), stringsAsFactors = F)
    GTDB_accNums_map <- all_accNums$V1
    GTDB_labels <- read.table(paste0(jumpDir, "required_GCFonly_ids.txt"), stringsAsFactors = F)
    names(GTDB_accNums_map) <- GTDB_labels$V1
    
    all_cdsAccNums <- read.table(file = paste0(jumpDir,"cds_fn_GCFids_found.txt"),stringsAsFactors = F, header = F)
    all_cdsAccNums <- all_cdsAccNums$V1
    
    all_protAccNums <- read.table(file = paste0(jumpDir,"prot_fn_GCFids_found.txt"),stringsAsFactors = F, header = F)
    all_protAccNums <- all_protAccNums$V1
    
    focal_GTDBLabels <- read.table(paste0(jumpDir, OLclade_oi,"_jump",jumpNum_oi,"_focalGTDBLabels.tab"), stringsAsFactors = F, header = F)
    focal_accNums <- sapply(focal_GTDBLabels$V1, function(x) GTDB_accNums_map[x], simplify = T, USE.NAMES = F)
    sis_GTDBLabels <- read.table(paste0(jumpDir,OLclade_oi,"_jump",jumpNum_oi,"_sisGTDBLabels.tab"), stringsAsFactors = F, header = F)
    sis_accNums <- sapply(sis_GTDBLabels$V1, function(x) GTDB_accNums_map[x], simplify = T, USE.NAMES = F)
    
    out1_accNums <- NA
    if(jumpClade_details[paste0("jump", jumpNum_oi), "depth"] %in% c("outgroup1", "outgroup2")){
      out1_GTDBLabels <- read.table(paste0(jumpDir,OLclade_oi,"_jump",jumpNum_oi,"_out1GTDBLabels.tab"), stringsAsFactors = F, header = F)
      out1_accNums <- sapply(out1_GTDBLabels$V1, function(x) GTDB_accNums_map[x], simplify = T, USE.NAMES = F)  
    }
    
    if(length(setdiff(intersect(focal_accNums, all_cdsAccNums), NA)) < 2 || length(setdiff(intersect(c(sis_accNums, out1_accNums), all_cdsAccNums), NA)) < 2 ){
      print("Cannot estimate dN/dS. No pair available in focal clade or related clade.")
      jumpClade_details[paste0("jump",jumpNum_oi), "yn00_analysis"] <- F
      next
    }
    jumpClade_details[paste0("jump",jumpNum_oi), "yn00_analysis"] <- T
    
    valid_cdsAccNums <- setdiff(intersect(c(focal_accNums, sis_accNums,out1_accNums), all_cdsAccNums), NA)
    valid_protAccNums <- setdiff(intersect(c(focal_accNums, sis_accNums,out1_accNums), all_protAccNums), NA)

    get_HEGnums <- function(prot_accNum_oi){
      prot_list <- read.fasta(file = paste0(gd_base,"proteins/",prot_accNum_oi,"_protein.faa"), seqtype = "AA", set.attributes = T, strip.desc = T)
      HEG_annotations <- grep(pattern = "50S ribosomal protein|LSU ribosomal protein|30S ribosomal protein|SSU ribosomal protein|translation elongation factor|RNA polymerase", x = getAnnot(prot_list), fixed = F)
      return(length(HEG_annotations))
    }

    HEG_nums <- sapply(valid_cdsAccNums, function(x) get_HEGnums(x))
    ref_genome <- valid_cdsAccNums[which(HEG_nums==max(HEG_nums))][1]

    ref_prot_list <- read.fasta(file = paste0(gd_base,"proteins/",ref_genome,"_protein.faa"), seqtype = "AA", set.attributes = T, strip.desc = T)
    HEG_prots <- grep(pattern = "50S ribosomal protein|LSU ribosomal protein|30S ribosomal protein|SSU ribosomal protein|translation elongation factor|RNA polymerase", x = getAnnot(ref_prot_list), fixed = F)
    write.fasta(sequences = ref_prot_list[HEG_prots], names = getAnnot(ref_prot_list)[HEG_prots], file.out = paste0(gd_base,"proteins/",ref_genome,"_HEG_protein.faa"))

    for(prot_accNum in valid_cdsAccNums){
      prot_fn <- paste0(gd_base,"proteins/",prot_accNum,"_protein.faa")
      prot_db_fn <- paste0(gd_base,"prot_db/",prot_accNum,"_prot.db")
      system2(command = "makeblastdb", args = c("-in", prot_fn, "-out", prot_db_fn, "-parse_seqids", "-dbtype", "prot"), stdout = F, stderr=F)
    }

    no_cores <- n_cores_av ##choose the appropriate number of cores
    cl_oi <- makeCluster(no_cores, type = "FORK")
    parLapply(cl = cl_oi, X = valid_cdsAccNums, function(x) system2(command = "blastp", args = c("-query", paste0(gd_base,"proteins/",ref_genome,"_HEG_protein.faa"), "-db", paste0(gd_base,"prot_db/",x,"_prot.db"), "-out", paste0(gd_base,"blast_op/",ref_genome,"_HEG_",x,"_blast_op.dat"), "-evalue", "1e-5", "-outfmt", shQuote("10 qaccver saccver qstart qend qlen slen evalue length")), stdout = paste0(jumpDir,"blastp.stdout"), stderr = paste0(jumpDir,"blastp.stderr")))
    stopCluster(cl_oi)

    prot_fn <- paste0(gd_base,"proteins/",ref_genome,"_HEG_protein.faa")
    prot_db_fn <- paste0(gd_base,"prot_db/",ref_genome,"_HEG_prot.db")
    system2(command = "makeblastdb", args = c("-in", prot_fn, "-out", prot_db_fn, "-parse_seqids", "-dbtype", "prot"), stdout = F, stderr = F)

    cl_oi <- makeCluster(no_cores, type = "FORK")
    parLapply(cl = cl_oi, X = valid_cdsAccNums, function(x) system2(command = "blastp", args = c("-query", paste0(gd_base,"proteins/",x,"_protein.faa"), "-db", paste0(gd_base,"prot_db/",ref_genome,"_HEG_prot.db"), "-out", paste0(gd_base,"blast_op/",x, "_",ref_genome,"_HEG_blast_op.dat"), "-evalue", "1e-5", "-outfmt", shQuote("10 qaccver saccver qstart qend qlen slen evalue length")), stdout = paste0(jumpDir,"blastp.stdout"), stderr = paste0(jumpDir,"blastp.stderr")))
    stopCluster(cl_oi)

    ref_HEG_prots <- ref_prot_list[HEG_prots]
    col_names <- c("qacc", "sacc", "qstart", "qend", "qlen", "slen", "evalue", "length")
                                                                              cl_oi <- makeCluster(n_cores_av, type = "FORK")
    clusterExport(cl=cl_oi, varlist=c("ref_genome"))
    shared_ortholog_list_of_list <- parLapply(cl = cl_oi, X=valid_cdsAccNums, function(x) get_rbbs(ref_genome, x))
    stopCluster(cl_oi)

    shared_ortholog_df <- data.frame(shared_ortholog_list_of_list, row.names = names(shared_ortholog_list_of_list[[1]]), stringsAsFactors = F)
    colnames(shared_ortholog_df) <- valid_cdsAccNums

    shared_ortholog_bool <- sapply(rownames(shared_ortholog_df), function(x) !anyNA(shared_ortholog_df[x,]))
    shared_orthologs <- rownames(shared_ortholog_df)[shared_ortholog_bool]
    shared_ortholog_scp_df <- shared_ortholog_df[shared_ortholog_bool,]

    cds_aln_dir <- paste0(jumpDir,"cds_alignments/")
    system2(command = "mkdir", args = c(cds_aln_dir))
    write.table(x = shared_orthologs,file = paste0(cds_aln_dir,"ref_orthologs_list.txt"), quote = F, row.names = F, col.names = F)
    write.table(x = shared_ortholog_scp_df,file = paste0(cds_aln_dir,"ref_orthologs_sets_list.txt"), quote = F)

    for(genome_n in 1:length(valid_cdsAccNums)){
      genome_oi <- valid_cdsAccNums[genome_n]
      prot_seqs_oi <- read.fasta(file = paste0(gd_base,"proteins/",genome_oi,"_protein.faa"), seqtype = "AA", set.attributes = T, strip.desc = F)
      cds_seqs_oi <- read.fasta(file = paste0(gd_base,"cds/",genome_oi,"_cds_from_genomic.fna"), seqtype = "DNA", set.attributes = T, strip.desc = F)
      for(ortho_n in 1:nrow(shared_ortholog_scp_df)){
        ref_ortho <- rownames(shared_ortholog_scp_df)[ortho_n]
        ortho_oi <- shared_ortholog_scp_df[ref_ortho,genome_oi]
        write.fasta(sequences = prot_seqs_oi[ortho_oi], names = genome_oi, file.out = paste0(cds_aln_dir,ref_ortho,"_prot.fas"), open = "a")
        write.fasta(sequences = cds_seqs_oi[which(grepl(pattern = ortho_oi,names(cds_seqs_oi)))], names = genome_oi, file.out = paste0(cds_aln_dir,ref_ortho,"_cds.fas"), open = "a")
      }##finsih looping over orthologs
    }##finish looping over genomes

  }##loop over jumps
  write.table(jumpClade_details, file = paste0(jump_common_dir, OLclade_oi, "_jumpSummaries4.tab"), quote = F)
}##loop over order level clades
```


#### Align the orthologs
I use PRANK to align and then Gblocks to clean up and concatenate the alignments.
```{r align_orthologs }
for(OLclade_n in 1:length(OLclades_oi)){
  OLclade_oi <- OLclades_oi[OLclade_n]
  print(paste0("Finding orthologs within ", OLclade_oi))
  pp_cutoff <- pp_cutoffs[OLclade_n]
  alpha_oi <- alphas_oi[OLclade_n]
  p_folder <- p_folders[OLclade_n]
  res_dir <- out_dir <-  paste0(levolution_out_dir,p_folder,"_GTDB_14k_levolution/")

  tree_pp <- read.newick(file = paste0(res_dir,OLclade_oi,"_GTDB_14k_GC_config", config_oi, "_alpha", alpha_oi,".post"))
  tree_in <- read.newick(paste0(res_dir,OLclade_oi,"_GTDB_14k_tree.newick"))
  OLclade_nTips <- length(tree_in$tip.label) ##OL=OrderLevel
  
  OLclade_jumpBranches <- which(tree_pp$edge.length>pp_cutoff)
  jumpClade_details <- read.table(file = paste0(jump_common_dir, OLclade_oi, "_jumpSummaries4.tab"), stringsAsFactors = F)
  
  for(jumpNum_oi in which(jumpClade_details$yn00_analysis)){
    print(jumpNum_oi)
    jumpBranch_oi <- OLclade_jumpBranches[jumpNum_oi]
    
    jumpDir <- paste0(jump_common_dir, OLclade_oi, "_jump", jumpNum_oi, "/")
    # if(dir.exists(paste0(jumpDir, "YN00/"))){
    #   print("YN00 estimates already available!")
    #   next
    # }
    cds_aln_dir <- paste0(jumpDir,"cds_alignments/")
    
    all_accNums <- read.table(paste0(jumpDir, "ass_fn_GCFids_found.txt"), stringsAsFactors = F)
    GTDB_accNums_map <- all_accNums$V1
    GTDB_labels <- read.table(paste0(jumpDir, "required_GCFonly_ids.txt"), stringsAsFactors = F)
    names(GTDB_accNums_map) <- GTDB_labels$V1
    
    all_cdsAccNums <- read.table(file = paste0(jumpDir,"cds_fn_GCFids_found.txt"),stringsAsFactors = F, header = F)
    all_cdsAccNums <- all_cdsAccNums$V1
    
    focal_GTDBLabels <- read.table(paste0(jumpDir, OLclade_oi,"_jump",jumpNum_oi,"_focalGTDBLabels.tab"), stringsAsFactors = F, header = F)
    focal_accNums <- sapply(focal_GTDBLabels$V1, function(x) GTDB_accNums_map[x], simplify = T, USE.NAMES = F)
    sis_GTDBLabels <- read.table(paste0(jumpDir,OLclade_oi,"_jump",jumpNum_oi,"_sisGTDBLabels.tab"), stringsAsFactors = F, header = F)
    sis_accNums <- sapply(sis_GTDBLabels$V1, function(x) GTDB_accNums_map[x], simplify = T, USE.NAMES = F)
    out1_accNums <- NA
    if(jumpClade_details[paste0("jump", jumpNum_oi), "depth"] %in% c("outgroup1", "outgroup2")){
      out1_GTDBLabels <- read.table(paste0(jumpDir,OLclade_oi,"_jump",jumpNum_oi,"_out1GTDBLabels.tab"), stringsAsFactors = F, header = F)
      out1_accNums <- sapply(out1_GTDBLabels$V1, function(x) GTDB_accNums_map[x], simplify = T, USE.NAMES = F)  
    }
    
    valid_cdsAccNums <- setdiff(intersect(c(focal_accNums, sis_accNums,out1_accNums), all_cdsAccNums), NA)
    
    shared_orthologs <- read.table(file = paste0(cds_aln_dir,"ref_orthologs_list.txt"), stringsAsFactors = F, header = F)
    shared_orthologs <- shared_orthologs$V1
    
    prank_cds_alignment_code <- paste0(main_dir, "Codes/prank_alignment_cds_v3.sh")
    system2(command = "parallel", args = c(paste0("-j",n_cores_av), "-a", paste0(cds_aln_dir,"ref_orthologs_list.txt"), prank_cds_alignment_code, ":::", cds_aln_dir))
    
    orthoSeqs_pathlist_fn <- paste0(cds_aln_dir,"jump", jumpNum_oi,"_codon_alns.txt")
    system2(command = "ls", args = c("-d", paste0(cds_aln_dir,"*.nuc.fas")), stdout = orthoSeqs_pathlist_fn)
    
    ##rewrite the sequences in a standard order
    ##this is required by Gblocks to correctly concatenate the alignments
    for(ortho_cds_aln_fn in readLines(orthoSeqs_pathlist_fn)){
      ortho_cds_aln <- read.alignment(ortho_cds_aln_fn, format = "fasta")
      write.fasta(sequences = ortho_cds_aln$seq[match(valid_cdsAccNums, ortho_cds_aln$nam)], names = valid_cdsAccNums, file.out = ortho_cds_aln_fn, open = "w")
    }
    
    system2(command = "Gblocks", args = c(orthoSeqs_pathlist_fn, "-t=c", "-b3=2", "-b4=10", "-b5=n", "-w=y", "-e=.aln", ">>", paste0(cds_aln_dir, "gblocks_alignment.log")))
  }
}
```


#### A function to remove stop codons
```{r remove_stop_codons }
removeStopCodons <- function(aln_oi){
  aln_oi_matrix <- t(sapply(aln_oi, function(x) getSequence(x), USE.NAMES = T))
  aln_oi_matrix <- aln_oi_matrix[,-(which(aln_oi_matrix[1,]==" "))] ##remove empty columns
  stopCodon_cols <- unique(unlist(sapply(1:nrow(aln_oi_matrix), function(y) which(sapply(0:((ncol(aln_oi_matrix)/3)-1), function(x) paste0(aln_oi_matrix[y,(1+3*x):(3+3*x)], collapse = "") %in% c("tga","tag","taa"))))))
  stopCodon_cols <- stopCodon_cols - 1  
  aln_oi_matrix <- aln_oi_matrix[,-c(sapply(stopCodon_cols, function(x) (1+x*3):(3+x*3)))]
  return(aln_oi_matrix)
}
```


####Actual dN/dS estimation using YN00
```{r estimate_yn00_dNdS }
yn00_ds_n <- 0
for(OLclade_n in 1:length(OLclades_oi)){
  OLclade_oi <- OLclades_oi[OLclade_n]
  jumpClade_details <- read.table(file = paste0(jump_common_dir, OLclade_oi, "_jumpSummaries4.tab"), stringsAsFactors = F)
  yn00_ds_n <- yn00_ds_n + length(which(jumpClade_details$yn00_analysis==T))
}

colNames_dNdS_yn00 <- c("focal_jump", "gc_focal", "gc_other", "median_dNdS_focal", "max_dNdS_focal", "min_dNdS_focal", "median_dNdS_others", "max_dNdS_others", "min_dNdS_others", "median_dN_focal", "max_dN_focal", "min_dN_focal", "median_dN_others", "max_dN_others", "min_dN_others", "median_dS_focal", "max_dS_focal", "min_dS_focal", "median_dS_others", "max_dS_others", "min_dS_others")
HEG_mean_dNdS <- data.frame(matrix(NA, nrow = yn00_ds_n, ncol = length(colNames_dNdS_yn00), dimnames = list(1:yn00_ds_n,colNames_dNdS_yn00)))

ds_n <- 1
for(OLclade_n in 1:length(OLclades_oi)){
  OLclade_oi <- OLclades_oi[OLclade_n]
  print(paste0("Calculating dN/dS in ", OLclade_oi))
  pp_cutoff <- pp_cutoffs[OLclade_n]
  alpha_oi <- alphas_oi[OLclade_n]
  p_folder <- p_folders[OLclade_n]
  res_dir <- out_dir <-  paste0(levolution_out_dir,p_folder,"_GTDB_14k_levolution/")

  tree_pp <- read.newick(file = paste0(res_dir,OLclade_oi,"_GTDB_14k_GC_config", config_oi, "_alpha", alpha_oi,".post"))
  tree_in <- read.newick(paste0(res_dir,OLclade_oi,"_GTDB_14k_tree.newick"))
  OLclade_nTips <- length(tree_in$tip.label) ##OL=OrderLevel
  OLclade_jumpBranches <- which(tree_pp$edge.length>pp_cutoff)
  
  jumpClade_details <- read.table(file = paste0(jump_common_dir, OLclade_oi, "_jumpSummaries4.tab"), stringsAsFactors = F)
  
  for(jumpNum_oi in which(jumpClade_details$yn00_analysis==T)){
    print(jumpNum_oi)
    jumpBranch_oi <- OLclade_jumpBranches[jumpNum_oi]
    
    jumpDir <- paste0(jump_common_dir, OLclade_oi, "_jump", jumpNum_oi, "/")
    cds_aln_dir <- paste0(jumpDir,"cds_alignments/")
    YN00_dir <- paste0(jumpDir,"YN00/")
    system2(command = "mkdir", args = c(paste0(YN00_dir)))
    
    all_accNums <- read.table(paste0(jumpDir, "ass_fn_GCFids_found.txt"), stringsAsFactors = F)
    GTDB_accNums_map <- all_accNums$V1
    GTDB_labels <- read.table(paste0(jumpDir, "required_GCFonly_ids.txt"), stringsAsFactors = F)
    names(GTDB_accNums_map) <- GTDB_labels$V1
    
    focal_GTDBLabels <- read.table(paste0(jumpDir, OLclade_oi,"_jump",jumpNum_oi,"_focalGTDBLabels.tab"), stringsAsFactors = F, header = F)
    focal_accNums <- sapply(focal_GTDBLabels$V1, function(x) GTDB_accNums_map[x], simplify = T, USE.NAMES = F)
    sis_GTDBLabels <- read.table(paste0(jumpDir,OLclade_oi,"_jump",jumpNum_oi,"_sisGTDBLabels.tab"), stringsAsFactors = F, header = F)
    sis_accNums <- sapply(sis_GTDBLabels$V1, function(x) GTDB_accNums_map[x], simplify = T, USE.NAMES = F)
    out1_accNums <- NA
    if(jumpClade_details[paste0("jump", jumpNum_oi), "depth"] %in% c("outgroup1", "outgroup2")){
      out1_GTDBLabels <- read.table(paste0(jumpDir,OLclade_oi,"_jump",jumpNum_oi,"_out1GTDBLabels.tab"), stringsAsFactors = F, header = F)
      out1_accNums <- sapply(out1_GTDBLabels$V1, function(x) GTDB_accNums_map[x], simplify = T, USE.NAMES = F)  
    }
    
    nested_focalTip_labels <- tryCatch({
      read.table(file = paste0(jumpDir, "nestedFocalTips.tab"), stringsAsFactors = F)$V1
    }, error=function(e){
      print(e)
      return(NA)
    })
    
    alnFile_fn <- paste0(cds_aln_dir,"jump", jumpNum_oi,"_codon_alns.txt--.seq")
    aln_oi <- read.fasta(file = alnFile_fn, as.string = F)
    
    valid_accNums <- getName(aln_oi)
    valid_focal_accNums <- setdiff(intersect(focal_accNums,valid_accNums), GTDB_accNums_map[nested_focalTip_labels])
    valid_sis_accNums <- setdiff(intersect(sis_accNums,valid_accNums), GTDB_accNums_map[nested_focalTip_labels])
    valid_other_accNums <- setdiff(setdiff(valid_accNums, valid_focal_accNums), GTDB_accNums_map[nested_focalTip_labels])
    
    if(length(valid_focal_accNums)<2 || length(valid_other_accNums)<2){
      print("Looks like either focal or related clade have a nested jump and there is no unambiguous pair to analyse")
      jumpClade_details[paste0("jump", jumpNum_oi),"yn00_analysis"] <- F
      next
    }
    
    ###perform yn00 estimation separately for focal clade, sister claed, and related clade
    focal_aln_oi <- aln_oi[valid_focal_accNums]
    renamed_focal_aln_fn <- paste0(cds_aln_dir,"jump", jumpNum_oi,"_codon_alns_focal.fa")
    write.fasta(sequences = focal_aln_oi, names = paste0("seq", 1:length(focal_aln_oi)), file.out = renamed_focal_aln_fn)
    focal_aln_oi_mod <- removeStopCodons(read.fasta(renamed_focal_aln_fn,as.string = F))
    mod_focal_aln_fn <- paste0(cds_aln_dir,"jump", jumpNum_oi,"_codon_alns_mod_focal.fa")
    write.dna(x = focal_aln_oi_mod, file = mod_focal_aln_fn, format = "fasta", nbcol = 10, colsep = "")
    
    focal_aln_oi_mod <- read.fasta(mod_focal_aln_fn, as.string = F)
    gc3_seqs <- sapply(focal_aln_oi_mod, function(x) GC3(seq = x))
    
    ds_YN00_ctl_fn_focal <- paste0(YN00_dir, OLclade_oi, "_", "jump", jumpNum_oi, "_YN00_focal.ctl")
    system2(command = "cp", args = c(paste0(input_dir, "yn00_for_R.ctl"), ds_YN00_ctl_fn_focal))
    
    seqfile_line <- paste0("seqfile = ", mod_focal_aln_fn)
    write(x = seqfile_line, file = ds_YN00_ctl_fn_focal, append = T)
      
    yn00_out_fn_focal <- paste0(YN00_dir, OLclade_oi, "_", "jump", jumpNum_oi, "_HEG_yn00_focal")
    outfile_line <- paste0("outfile = ", yn00_out_fn_focal)
    write(x = outfile_line, file = ds_YN00_ctl_fn_focal, append = T)
    
    yn00_sh_fn <- paste0(jumpDir, "call_yn00_focal.sh")
    writeLines(text = paste0("cd ", jumpDir, "\n", "yn00 ", paste0("YN00/", OLclade_oi,"_jump",jumpNum_oi, "_YN00_focal.ctl")), con = yn00_sh_fn)
    
    system2(command = "bash", args = c(yn00_sh_fn),stdout = F, stderr = F)
    
    yn00_output_focal <- readLines(yn00_out_fn_focal)
    yn00_res_start <- grep(pattern = "seq. ", x = yn00_output_focal, fixed = T)
    yn00_tab_start <- yn00_res_start + 2
    yn00_tab_end <- yn00_tab_start - 1 + (length(focal_aln_oi_mod)*(length(focal_aln_oi_mod)-1))/2
    
    yn00_tab_focal <- do.call(rbind, strsplit(yn00_output_focal[yn00_tab_start:yn00_tab_end], "\\s+"))
    yn00_tab_focal<- data.frame(yn00_tab_focal, stringsAsFactors = F)
    yn00_tab_focal$X2 <- as.integer(yn00_tab_focal$X2)
    yn00_tab_focal$X3 <- as.integer(yn00_tab_focal$X3)
    
    colNames_ynTab <- c("dataset", "parent", "category", "gene_set", "acc1", "acc2", "name1", "name2", "dN", "dS", "dN/dS", "gc1", "gc2")
    
    n_focal_pairs <- length(valid_focal_accNums)*(length(valid_focal_accNums)-1)/2
    pairwise_yn00_est_focal <- as.data.frame(matrix(NA, nrow = n_focal_pairs, ncol = 13, dimnames = list(1:n_focal_pairs, colNames_ynTab)))
    focal_combinations <- combn(valid_focal_accNums, 2)
    for(focal_pair in 1:n_focal_pairs){
      acc1 <- focal_combinations[1,focal_pair]
      acc2 <- focal_combinations[2,focal_pair]
      n1_in_aln <- which(valid_focal_accNums==acc1)
      n2_in_aln <- which(valid_focal_accNums==acc2)
      
      name1 <- get_genera_sp_names(GTDB_14k_metadata[names(GTDB_accNums_map)[which(GTDB_accNums_map==acc1)],"gtdb_taxonomy"])
      name2 <- get_genera_sp_names(GTDB_14k_metadata[names(GTDB_accNums_map)[which(GTDB_accNums_map==acc2)],"gtdb_taxonomy"])
      
      row_oi <- which(yn00_tab_focal$X2==n2_in_aln & yn00_tab_focal$X3==n1_in_aln)
      if(identical(row_oi, integer(0))) row_oi <- which(yn00_tab_focal$X2==n1_in_aln & yn00_tab_focal$X3==n2_in_aln)
      pairwise_yn00_est_focal[focal_pair,] <- c(OLclade_oi, paste0("jump", jumpNum_oi), "focal", "HEG", acc2, acc1, name2, name1, yn00_tab_focal[row_oi, 9], yn00_tab_focal[row_oi, 12], yn00_tab_focal[row_oi, 8], gc3_seqs[n2_in_aln], gc3_seqs[n1_in_aln])
    }
    
    write.table(x = pairwise_yn00_est_focal, file = paste0(YN00_dir, OLclade_oi,"_", paste0("jump", jumpNum_oi), "_HEG_YN00_estimates_focal.tab"),col.names = T)
    pairwise_yn00_est_focal <- read.table(file = paste0(YN00_dir, OLclade_oi,"_", paste0("jump", jumpNum_oi), "_HEG_YN00_estimates_focal.tab"), stringsAsFactors = F)
    
    ###for related clade
    related_aln_oi <- aln_oi[valid_other_accNums]
    renamed_related_aln_fn <- paste0(cds_aln_dir,"jump", jumpNum_oi,"_codon_alns_related.fa")
    write.fasta(sequences = related_aln_oi, names = paste0("seq", 1:length(related_aln_oi)), file.out = renamed_related_aln_fn)
    related_aln_oi_mod <- removeStopCodons(read.fasta(renamed_related_aln_fn,as.string = F))
    mod_related_aln_fn <- paste0(cds_aln_dir,"jump", jumpNum_oi,"_codon_alns_mod_related.fa")
    write.dna(x = related_aln_oi_mod, file = mod_related_aln_fn, format = "fasta", nbcol = 10, colsep = "")
    
    related_aln_oi_mod <- read.fasta(mod_related_aln_fn, as.string = F)
    gc3_seqs <- sapply(related_aln_oi_mod, function(x) GC3(seq = x))
    
    ds_YN00_ctl_fn_related <- paste0(YN00_dir, OLclade_oi, "_", "jump", jumpNum_oi, "_YN00_related.ctl")
    system2(command = "cp", args = c(paste0(input_dir, "yn00_for_R.ctl"), ds_YN00_ctl_fn_related))
    
    seqfile_line <- paste0("seqfile = ", mod_related_aln_fn)
    write(x = seqfile_line, file = ds_YN00_ctl_fn_related, append = T)
      
    yn00_out_fn_related <- paste0(YN00_dir, OLclade_oi, "_", "jump", jumpNum_oi, "_HEG_yn00_related")
    outfile_line <- paste0("outfile = ", yn00_out_fn_related)
    write(x = outfile_line, file = ds_YN00_ctl_fn_related, append = T)
    
    yn00_sh_fn <- paste0(jumpDir, "call_yn00_related.sh")
    writeLines(text = paste0("cd ", jumpDir, "\n", "yn00 ", paste0("YN00/", OLclade_oi,"_jump",jumpNum_oi, "_YN00_related.ctl")), con = yn00_sh_fn)
    
    system2(command = "bash", args = c(yn00_sh_fn),stdout = F, stderr = F)
    
    yn00_output_related <- readLines(yn00_out_fn_related)
    yn00_res_start <- grep(pattern = "seq. ", x = yn00_output_related, fixed = T)
    yn00_tab_start <- yn00_res_start + 2
    yn00_tab_end <- yn00_tab_start - 1 + (length(related_aln_oi_mod)*(length(related_aln_oi_mod)-1))/2
    
    yn00_tab_related <- do.call(rbind, strsplit(yn00_output_related[yn00_tab_start:yn00_tab_end], "\\s+"))
    yn00_tab_related<- data.frame(yn00_tab_related, stringsAsFactors = F)
    yn00_tab_related$X2 <- as.integer(yn00_tab_related$X2)
    yn00_tab_related$X3 <- as.integer(yn00_tab_related$X3)
    
    
    n_sister_pairs <- length(valid_sis_accNums)*(length(valid_sis_accNums)-1)/2
    n_other_pairs <- length(valid_other_accNums)*(length(valid_other_accNums) - 1)/2
    pairwise_yn00_est_related <- as.data.frame(matrix(NA, nrow = n_other_pairs, ncol = 13, dimnames = list(1:n_other_pairs, colNames_ynTab)))
    other_combinations <- combn(valid_other_accNums, 2)
    for(other_pair in 1:n_other_pairs){
      acc1 <- other_combinations[1,other_pair]
      acc2 <- other_combinations[2,other_pair]
      n1_in_aln <- which(valid_other_accNums==acc1)
      n2_in_aln <- which(valid_other_accNums==acc2)
      
      name1 <- get_genera_sp_names(GTDB_14k_metadata[names(GTDB_accNums_map)[which(GTDB_accNums_map==acc1)],"gtdb_taxonomy"])
      name2 <- get_genera_sp_names(GTDB_14k_metadata[names(GTDB_accNums_map)[which(GTDB_accNums_map==acc2)],"gtdb_taxonomy"])
      
      row_oi <- which(yn00_tab_related$X2==n2_in_aln & yn00_tab_related$X3==n1_in_aln)
      if(identical(row_oi, integer(0))) row_oi <- which(yn00_tab_related$X2==n1_in_aln & yn00_tab_related$X3==n2_in_aln)
      if(acc1 %in% valid_sis_accNums & acc2 %in% valid_sis_accNums){
        pairwise_yn00_est_related[n_focal_pairs+other_pair,] <- c(OLclade_oi, paste0("jump", jumpNum_oi), "sister", "HEG", acc2, acc1, name2, name1, yn00_tab_related[row_oi, 9], yn00_tab_related[row_oi, 12], yn00_tab_related[row_oi,8], gc3_seqs[n2_in_aln], gc3_seqs[n1_in_aln])
      }else{
        pairwise_yn00_est_related[n_focal_pairs+other_pair,] <- c(OLclade_oi, paste0("jump", jumpNum_oi), "other", "HEG", acc2, acc1, name2, name1, yn00_tab_related[row_oi, 9], yn00_tab_related[row_oi, 12], yn00_tab_related[row_oi,8], gc3_seqs[n2_in_aln], gc3_seqs[n1_in_aln])
      }
    }#end of other genome combinations
      
    write.table(x = pairwise_yn00_est_related, file = paste0(YN00_dir, OLclade_oi,"_", paste0("jump", jumpNum_oi), "_HEG_YN00_estimates_related.tab"),col.names = T)
    pairwise_yn00_est_related <- read.table(file = paste0(YN00_dir, OLclade_oi,"_", paste0("jump", jumpNum_oi), "_HEG_YN00_estimates_related.tab"), stringsAsFactors = F)

    focal_rows <- which(pairwise_yn00_est_focal$category=="focal")
    focal_rows <- intersect(focal_rows, which(pairwise_yn00_est_focal$dS<2.5))
    other_rows <- which(pairwise_yn00_est_related$category %in% c("sister","other"))
    other_rows <- intersect(other_rows, which(pairwise_yn00_est_related$dS<2.5))
    
    focal_GC <- mean(c(pairwise_yn00_est_focal$gc1[focal_rows], pairwise_yn00_est_focal$gc2[focal_rows]))
    other_GC <- mean(c(pairwise_yn00_est_related$gc1[other_rows], pairwise_yn00_est_related$gc2[other_rows]))
    
    median_dNdS_focal <- median(pairwise_yn00_est_focal$`dN.dS`[focal_rows], na.rm = T)
    max_dNdS_focal <- max(pairwise_yn00_est_focal$`dN.dS`[focal_rows], na.rm = T)
    min_dNdS_focal <- min(pairwise_yn00_est_focal$`dN.dS`[focal_rows], na.rm = T)
    median_dNdS_other <- median(pairwise_yn00_est_related$`dN.dS`[other_rows], na.rm = T)
    max_dNdS_other <- max(pairwise_yn00_est_related$`dN.dS`[other_rows], na.rm = T)
    min_dNdS_other <- min(pairwise_yn00_est_related$`dN.dS`[other_rows], na.rm = T)
    median_dN_focal <- median(pairwise_yn00_est_focal$`dN`[focal_rows], na.rm = T)
    max_dN_focal <- max(pairwise_yn00_est_focal$`dN`[focal_rows], na.rm = T)
    min_dN_focal <- min(pairwise_yn00_est_focal$`dN`[focal_rows], na.rm = T)
    median_dN_other <- median(pairwise_yn00_est_related$`dN`[other_rows], na.rm = T)
    max_dN_other <- max(pairwise_yn00_est_related$`dN`[other_rows], na.rm = T)
    min_dN_other <- min(pairwise_yn00_est_related$`dN`[other_rows], na.rm = T)
    median_dS_focal <- median(pairwise_yn00_est_focal$`dS`[focal_rows], na.rm = T)
    max_dS_focal <- max(pairwise_yn00_est_focal$`dS`[focal_rows], na.rm = T)
    min_dS_focal <- min(pairwise_yn00_est_focal$`dS`[focal_rows], na.rm = T)
    median_dS_other <- median(pairwise_yn00_est_related$`dS`[other_rows], na.rm = T)
    max_dS_other <- max(pairwise_yn00_est_related$`dS`[other_rows], na.rm = T)
    min_dS_other <- min(pairwise_yn00_est_related$`dS`[other_rows], na.rm = T)
    
    
    HEG_mean_dNdS[ds_n, ] <- c(paste0(OLclade_oi, "_jump", jumpNum_oi), round(focal_GC,3), round(other_GC,3), round(median_dNdS_focal,3), round(max_dNdS_focal,3), round(min_dNdS_focal,3), round(median_dNdS_other,3), round(max_dNdS_other,3), round(min_dNdS_other,3), round(median_dN_focal,3), round(max_dN_focal,3), round(min_dN_focal,3), round(median_dN_other,3), round(max_dN_other,3), round(min_dN_other,3), round(median_dS_focal,3), round(max_dS_focal,3), round(min_dS_focal,3), round(median_dS_other,3), round(max_dS_other,3), round(min_dS_other,3))
    ds_n <- ds_n + 1
  }
  write.table(jumpClade_details, file = paste0(jump_common_dir, OLclade_oi, "_jumpSummaries4.tab"), quote = F)
}
HEG_mean_dNdS <- HEG_mean_dNdS[!is.na(HEG_mean_dNdS$focal_jump),]
rownames(HEG_mean_dNdS) <- HEG_mean_dNdS[,"focal_jump"]
write.table(HEG_mean_dNdS, file = paste0(jump_common_dir,"levolution_jumps_yn00_comparison2.tab"), row.names = T, col.names = T, quote = F)
```

#### Plotting dN/dS comparisons
```{r plot_dNdS_comparison }
HEG_mean_dNdS <- read.table(file = paste0(jump_common_dir,"levolution_jumps_yn00_comparison2.tab"), stringsAsFactors = F)
HEG_mean_dNdS$dNdSRatio <- HEG_mean_dNdS$median_dNdS_focal/HEG_mean_dNdS$median_dNdS_others

dNdS_focal_sis <- ggplot(HEG_mean_dNdS, aes(x=median_dNdS_others, y=median_dNdS_focal, col=gc_focal - gc_other )) + geom_point(shape=21, stroke=1.2)
dNdS_focal_sis <- dNdS_focal_sis + scale_x_continuous(name = "median dN/dS (related clade)") + scale_y_continuous(name = "median dN/dS (focal clade)") + scale_color_gradient2()
dNdS_focal_sis <- dNdS_focal_sis + labs(color=expression(Delta*"GC3")) + theme(legend.title.align = 0.5)
dNdS_focal_sis <- dNdS_focal_sis + geom_abline(slope = 1, intercept = 0, color="grey")

dNdSratio_plt <- ggplot(HEG_mean_dNdS, aes(x=gc_focal - gc_other, y=median_dNdS_focal/median_dNdS_others)) + geom_point(col="darkblue", shape=21, stroke=1.5)
dNdSratio_plt <- dNdSratio_plt + scale_x_continuous(name = expression(Delta*"GC3 (focal clade - sister clade)")) + scale_y_continuous(name = "focal clade dN/dS/\n related clade dN/dS", trans = "log2", breaks = c(0.5,1,2), limits = c(0.5,4)) + geom_hline(yintercept = 1, color="grey")

dNdS_plt <- plot_grid(plot_grid(dNdS_focal_sis + theme(legend.position = "none"), dNdSratio_plt, align = "hv", ncol = 1, labels = "AUTO", label_size = 20), plot_grid(get_legend(dNdS_focal_sis), ggplot(), ncol = 1), rel_widths = c(7,3))

save_plot(filename = paste0(jump_common_dir, "HEG_dNdS_yn00_comparisonns.svg"), plot = dNdS_plt, ncol = 1, base_width = 5, base_height = 6)
save_plot(filename = paste0(jump_common_dir, "HEG_dNdS_yn00_comparisonns.png"), plot = dNdS_plt, ncol = 1, base_width = 5, base_height = 6)

dNdS_plt
```

#### Plot dN/dS temporal
```{r plot_dNdS_temporal }
for(OLclade_n in 1:length(OLclades_oi)){
  OLclade_oi <- OLclades_oi[OLclade_n]
  print(paste0("Calculating dN/dS in ", OLclade_oi))
  pp_cutoff <- pp_cutoffs[OLclade_n]
  alpha_oi <- alphas_oi[OLclade_n]
  p_folder <- p_folders[OLclade_n]
  res_dir <- out_dir <-  paste0(levolution_out_dir,p_folder,"_GTDB_14k_levolution/")

  tree_pp <- read.newick(file = paste0(res_dir,OLclade_oi,"_GTDB_14k_GC_config", config_oi, "_alpha", alpha_oi,".post"))
  tree_in <- read.newick(paste0(res_dir,OLclade_oi,"_GTDB_14k_tree.newick"))
  OLclade_nTips <- length(tree_in$tip.label) ##OL=OrderLevel
  OLclade_jumpBranches <- which(tree_pp$edge.length>pp_cutoff)
  
  jumpClade_details2 <- read.table(file = paste0(jump_common_dir, OLclade_oi, "_jumpSummaries4.tab"), stringsAsFactors = F)
  
  for(jumpNum_oi in which(jumpClade_details2$yn00_analysis==T)){
    print(jumpNum_oi)
    jumpBranch_oi <- OLclade_jumpBranches[jumpNum_oi]
    
    jumpDir <- paste0(jump_common_dir, OLclade_oi, "_jump", jumpNum_oi, "/")
    cds_aln_dir <- paste0(jumpDir,"cds_alignments/")
    YN00_dir <- paste0(jumpDir,"YN00/")
    
    pairwise_yn00_est_focal <- read.table(file = paste0(YN00_dir, OLclade_oi,"_", paste0("jump", jumpNum_oi), "_HEG_YN00_estimates_focal.tab"), stringsAsFactors = F)
    
    pairwise_yn00_est_related <- read.table(file = paste0(YN00_dir, OLclade_oi,"_", paste0("jump", jumpNum_oi), "_HEG_YN00_estimates_related.tab"), stringsAsFactors = F)
    
    # png(filename = paste0(jumpDir,"YN00/", OLclade_oi, "_jump", jumpNum_oi, "dNdS_vs_dS.png"))
    plot(pairwise_yn00_est_focal$dS, pairwise_yn00_est_focal$dN.dS, col="orange", xlim=c(0,4), ylim=c(0,max(c(pairwise_yn00_est_focal$dN.dS, pairwise_yn00_est_related$dN.dS), na.rm = T)), xlab = "dS", ylab="dN/dS", main=paste0(OLclade_oi,", jump", jumpNum_oi), lwd=2)
    points(pairwise_yn00_est_related$dS, pairwise_yn00_est_related$dN.dS, col="purple", lwd=2)
    # dev.off()
  }
}
```

#### Analysis of dN/dS, accounting for dS
```{r dN_dS_analysis_dS }
valid_median_dNdS_ratios <- c()
names_ds <- c()
for(OLclade_n in 1:length(OLclades_oi)){
  OLclade_oi <- OLclades_oi[OLclade_n]
  jumpClade_details2 <- read.table(file = paste0(jump_common_dir, OLclade_oi, "_jumpSummaries4.tab"), stringsAsFactors = F)
  valid_median_dNdS_ratios <- c(valid_median_dNdS_ratios, rep(NA, length(which(jumpClade_details2$yn00_analysis==T))))
  if(!identical(which(jumpClade_details2$yn00_analysis==T),integer(0))){
    names_ds <- c(names_ds, paste0(OLclade_oi,"_jump",which(jumpClade_details2$yn00_analysis==T)))
  }
}

valid_median_gc3_focal <- valid_median_gc3_related <- valid_median_dNdS_focal <- valid_median_dNdS_related <- valid_median_dS_focal <- valid_median_dS_related <- valid_median_dNdS_ratios

names(valid_median_gc3_focal) <- names(valid_median_gc3_related) <- names(valid_median_dNdS_related) <- names(valid_median_dNdS_focal) <- names(valid_median_dS_related) <- names(valid_median_dS_focal) <- names(valid_median_dNdS_ratios) <- names_ds

for(OLclade_n in 1:length(OLclades_oi)){
  OLclade_oi <- OLclades_oi[OLclade_n]
  print(paste0("Calculating dN/dS in ", OLclade_oi))
  pp_cutoff <- pp_cutoffs[OLclade_n]
  alpha_oi <- alphas_oi[OLclade_n]
  p_folder <- p_folders[OLclade_n]
  res_dir <- out_dir <-  paste0(levolution_out_dir,p_folder,"_GTDB_14k_levolution/")

  tree_pp <- read.newick(file = paste0(res_dir,OLclade_oi,"_GTDB_14k_GC_config", config_oi, "_alpha", alpha_oi,".post"))
  tree_in <- read.newick(paste0(res_dir,OLclade_oi,"_GTDB_14k_tree.newick"))
  OLclade_nTips <- length(tree_in$tip.label) ##OL=OrderLevel
  OLclade_jumpBranches <- which(tree_pp$edge.length>pp_cutoff)
  
  jumpClade_details2 <- read.table(file = paste0(jump_common_dir, OLclade_oi, "_jumpSummaries4.tab"), stringsAsFactors = F)
  
  for(jumpNum_oi in which(jumpClade_details2$yn00_analysis==T)){
    print(jumpNum_oi)
    jumpBranch_oi <- OLclade_jumpBranches[jumpNum_oi]
    
    jumpDir <- paste0(jump_common_dir, OLclade_oi, "_jump", jumpNum_oi, "/")
    cds_aln_dir <- paste0(jumpDir,"cds_alignments/")
    YN00_dir <- paste0(jumpDir,"YN00/")
    
    alnFile_fn <- paste0(cds_aln_dir,"jump", jumpNum_oi,"_codon_alns.txt--.seq")
    aln_oi <- read.fasta(file = alnFile_fn, as.string = F)
    gc3_seqs <- sapply(aln_oi, function(x) GC3(seq = x))
    
    pairwise_yn00_est_focal <- read.table(file = paste0(YN00_dir, OLclade_oi,"_", paste0("jump", jumpNum_oi), "_HEG_YN00_estimates_focal.tab"), stringsAsFactors = F)
    
    pairwise_yn00_est_related <- read.table(file = paste0(YN00_dir, OLclade_oi,"_", paste0("jump", jumpNum_oi), "_HEG_YN00_estimates_related.tab"), stringsAsFactors = F)
    
    valid_dS_focal <- valid_dS_related <- valid_dNdS_focal <- valid_dNdS_related <- valid_dNdS_ratios <- c()
    valid_acc_focal <- valid_acc_related <- c()
    for(focal_pair in 1:nrow(pairwise_yn00_est_focal)){
      if(pairwise_yn00_est_focal[focal_pair,"dS"]>1){
          print("this pair may be saturated!")
        
          valid_dNdS_focal <- c(valid_dNdS_focal, NA)
          valid_dNdS_related <- c(valid_dNdS_related, NA)
          valid_dS_focal <- c(valid_dS_focal, NA)
          valid_dS_related <- c(valid_dS_related, NA)
          valid_dNdS_ratios <- c(valid_dNdS_ratios, NA)
          
          valid_acc_focal <- c(valid_acc_focal, NA)
          valid_acc_related <- c(valid_acc_related, NA)
          next
      }
      valid_related_pairs <- which(abs(pairwise_yn00_est_related$dS - pairwise_yn00_est_focal[focal_pair,"dS"]) <= 0.2)
      
      valid_dNdS_focal <- c(valid_dNdS_focal, pairwise_yn00_est_focal[focal_pair, "dN.dS"])
      valid_dNdS_related <- c(valid_dNdS_related, pairwise_yn00_est_related[valid_related_pairs, "dN.dS"])
      valid_dS_focal <- c(valid_dS_focal, pairwise_yn00_est_focal[focal_pair, "dS"])
      valid_dS_related <- c(valid_dS_related, pairwise_yn00_est_related[valid_related_pairs, "dS"])
      valid_dNdS_ratios <- c(valid_dNdS_ratios, pairwise_yn00_est_focal[focal_pair,"dN.dS"]/pairwise_yn00_est_related[valid_related_pairs,"dN.dS"])
      
      valid_acc_focal <- c(valid_acc_focal, pairwise_yn00_est_focal[focal_pair,"acc1"], pairwise_yn00_est_focal[focal_pair,"acc2"])
      valid_acc_related <- c(valid_acc_related, pairwise_yn00_est_related[valid_related_pairs,"acc1"], pairwise_yn00_est_related[valid_related_pairs,"acc2"])
    }
    
    if(all(is.na(valid_dNdS_focal))){ ##no valid focal pair
      print("No focal pair with less divergence found!")
      next ##skips this jump
    }
    
    valid_median_dNdS_focal[paste0(OLclade_oi,"_jump", jumpNum_oi)] <- median(valid_dNdS_focal, na.rm = T)
    valid_median_dNdS_related[paste0(OLclade_oi,"_jump", jumpNum_oi)] <- median(valid_dNdS_related, na.rm = T)
    valid_median_dS_focal[paste0(OLclade_oi,"_jump", jumpNum_oi)] <- median(valid_dS_focal, na.rm = T)
    valid_median_dS_related[paste0(OLclade_oi,"_jump", jumpNum_oi)] <- median(valid_dS_related, na.rm = T)
    valid_median_dNdS_ratios[paste0(OLclade_oi,"_jump", jumpNum_oi)] <- median(valid_dNdS_ratios, na.rm = T)
    
    valid_median_gc3_focal[paste0(OLclade_oi,"_jump", jumpNum_oi)]  <- median(gc3_seqs[valid_acc_focal], na.rm = T)
    valid_median_gc3_related[paste0(OLclade_oi,"_jump", jumpNum_oi)]  <- median(gc3_seqs[valid_acc_related], na.rm = T)
  }
}

YN00_dS_corrected_omega <- cbind(valid_median_dNdS_focal, valid_median_dNdS_related, valid_median_dS_focal, valid_median_dS_related, valid_median_dNdS_ratios, valid_median_gc3_focal, valid_median_gc3_related)
write.table(x = YN00_dS_corrected_omega, file = paste0(jump_common_dir,"levolution_jumps_yn00_dS_corrected_comparison2.tab"), row.names = T, col.names = T, quote = F)
```


### CUB analysis: comparison of CUB between focal and related clades
Here the idea is to test whether changes in codon usage bias (CUB) accompany GC jumps. Since CUB depends on selection, it is likely to be affected by effective population size. Thus, changes in CUB can indicate changes in effective population size.

#### Some common code required for CUB analysis
```{r common_codon_code }
codon_table <- read.csv(file = "codon_aa_map.txt", header = T, stringsAsFactors=F)
degen2_aas <- unique(codon_table[codon_table$degen==2, "aa"])
degen4_aas <- unique(codon_table[codon_table$degen==4, "aa"])
degen6_aas <- unique(codon_table[codon_table$degen==6, "aa"])
aa_unique <- setdiff(unique(codon_table$aa), c("Stop", "Met", "Trp"))
codons_aa <- sapply(aa_unique, function(x) codon_table[codon_table$aa==x, "codon"])

cub_dir <- "/home/saurabh/Documents/Deepas_lab_Backup/Deepas_lab/Work/GC_macro/Results/CUB"
CUB_code_dir <- "/home/saurabh/Documents/Deepas_lab_Backup/Deepas_lab/Work/tRNA_CUB_aa-master/"
```


#### Count codons using ENCprime and calculate CUB
Here, CUB is calculated according to Sharp or Higgs and Ran, i.e. average CUB is calculated as the average of CUB of 4 amino acids with known optimal codons.
```{r count_codons, eval=F }
for(OLclade_n in 1:length(OLclades_oi)){
  OLclade_oi <- OLclades_oi[OLclade_n]
  print(paste0("Analysing jumps within ", OLclade_oi))

  jumpClade_details <- read.table(file = paste0(jump_common_dir, OLclade_oi, "_jumpSummaries4.tab"), stringsAsFactors = F)
  jumpClade_details[,"codon_counting"] <- rep(NA, nrow(jumpClade_details))
  
  for(jumpNum_oi in which(jumpClade_details[,"genomes_downloaded"])){#only for those jumps which are analyzable
    print(paste0("Counting codons for jump number ", jumpNum_oi))
    
    jumpDir <- paste0(jump_common_dir, OLclade_oi, "_jump", jumpNum_oi, "/")
    all_accNums <- read.table(paste0(jumpDir, "ass_fn_GCFids_found.txt"), stringsAsFactors = F)
    GTDB_accNums_map <- all_accNums$V1
    GTDB_labels <- read.table(paste0(jumpDir, "required_GCFonly_ids.txt"), stringsAsFactors = F)
    names(GTDB_accNums_map) <- GTDB_labels$V1
    
    all_cdsAccNums <- read.table(file = paste0(jumpDir,"cds_fn_GCFids_found.txt"),stringsAsFactors = F, header = F)
    all_cdsAccNums <- all_cdsAccNums$V1
    
    focal_GTDBLabels <- read.table(paste0(jumpDir, OLclade_oi,"_jump",jumpNum_oi,"_focalGTDBLabels.tab"), stringsAsFactors = F, header = F)
    focal_accNums <- sapply(focal_GTDBLabels$V1, function(x) GTDB_accNums_map[x], simplify = T, USE.NAMES = F)
    sis_GTDBLabels <- read.table(paste0(jumpDir,OLclade_oi,"_jump",jumpNum_oi,"_sisGTDBLabels.tab"), stringsAsFactors = F, header = F)
    sis_accNums <- sapply(sis_GTDBLabels$V1, function(x) GTDB_accNums_map[x], simplify = T, USE.NAMES = F)
    out1_accNums <- NA
    if(jumpClade_details[paste0("jump", jumpNum_oi), "depth"] %in% c("outgroup1", "outgroup2")){
      out1_GTDBLabels <- read.table(paste0(jumpDir,OLclade_oi,"_jump",jumpNum_oi,"_out1GTDBLabels.tab"), stringsAsFactors = F, header = F)
      out1_accNums <- sapply(out1_GTDBLabels$V1, function(x) GTDB_accNums_map[x], simplify = T, USE.NAMES = F)  
    }
    
    if(!any(focal_accNums %in% setdiff(all_cdsAccNums, NA)) || !(any(c(sis_accNums, out1_accNums) %in% setdiff(all_cdsAccNums, NA)) )){
      print("Cannot count codons! cds files not available for focal clade OR sister/first outgroup clade")
      jumpClade_details[paste0("jump",jumpNum_oi), "codon_counting"] <- F
      next
    }
    jumpClade_details[paste0("jump",jumpNum_oi), "codon_counting"] <- T
    
    available_cdsAccNums <- c(focal_accNums[focal_accNums %in% all_cdsAccNums], sis_accNums[sis_accNums %in% all_cdsAccNums], out1_accNums[out1_accNums %in% all_cdsAccNums])
    available_cdsAccNums <- setdiff(available_cdsAccNums, NA)
    write.table(available_cdsAccNums, paste0(jumpDir,"available_cdsAccNums.txt"), quote = F, row.names = F, col.names = F)

    # system2(command = "parallel", args = c(paste0("-j", n_cores_av), "-a", paste0(jumpDir, "available_cdsAccNums.txt"),  paste0(CUB_code_dir, "codon_counting_script.sh"), ":::", "10"), stdout = paste0(jumpDir, "codon_counting_log.txt"), stderr = paste0(jumpDir, "codon_counting_error.txt"))

    ##calculate CUB
    aawise_S_all <- matrix(data = NA, nrow = length(available_cdsAccNums), ncol = length(degen2_aas)+2, dimnames = list(available_cdsAccNums, c(degen2_aas, "Ile", "S_avg")))
    for(genome_oi in available_cdsAccNums){ #for each genome
      tryCatch({
        codCnt_matrix <- read.table(file = paste0(cub_dir,"/", genome_oi,"/", genome_oi,"_cds_from_genomic.fna.codcnt_matrix"), header = T, sep = "\t", row.names = 1)

        ribosomal_rows <- rownames(codCnt_matrix)[grepl("ribosomal protein", rownames(codCnt_matrix))]
        EF_rows <- rownames(codCnt_matrix)[grepl("elongation factor Tu", rownames(codCnt_matrix))]
        RNApol_rows <- rownames(codCnt_matrix)[grepl("\\[protein=DNA-directed RNA polymerase\\]", rownames(codCnt_matrix)) | (grepl("DNA-directed RNA polymerase", rownames(codCnt_matrix)) & grepl("subunit A'", rownames(codCnt_matrix)))  | (grepl("DNA-directed RNA polymerase", rownames(codCnt_matrix)) & grepl("subunit B", rownames(codCnt_matrix)))]
        HEG_rows_def_ribo <- c(ribosomal_rows, EF_rows, RNApol_rows)
        HEG_rows <- HEG_rows_def_ribo
        remaining_rows <- setdiff(rownames(codCnt_matrix), HEG_rows_def_ribo)
        # print(paste0("There are ", length(HEG_rows), " HEGs and ", length(remaining_rows), " other genes."))
        codCnt_HEG <- colSums(x = codCnt_matrix[HEG_rows,], na.rm = T) #pool codon counts for all HEG
        codCnt_rest <- colSums(x = codCnt_matrix[remaining_rows,], na.rm = T) #pool codon counts for all remaining genes

        S_CUB  <- sapply(c(degen2_aas, "Ile", "avg"), function(x) 0)
        for(aa_oi in c(degen2_aas, "Ile")){
          codons_oi <- codon_table[codon_table$aa==aa_oi, "codon"]
          S_CUB[aa_oi] <- log((codCnt_HEG[codons_oi[2]]/codCnt_HEG[codons_oi[1]])/(codCnt_rest[codons_oi[2]]/codCnt_rest[codons_oi[1]]))
        }

        aas_oi <- c("Phe", "His", "Tyr", "Asn")
        S_weights <- sapply(aas_oi, function(x) sum(codCnt_HEG[codon_table[codon_table$aa==x, "codon"]][1:2]))
        S_CUB["avg"] <- weighted.mean(x = S_CUB[aas_oi], w = S_weights, na.rm = T)

        aawise_S_all[genome_oi,] <- round(S_CUB, digits = 4)
      }, error=function(e){
                print(e)
      })
    } #end of loop around all genomes in this dataset
    write.table(aawise_S_all, file = paste0(jumpDir, "CUB_Sharp.txt"))
  }##finish loop over jumps in one order level clade
  write.table(jumpClade_details, file = paste0(jump_common_dir, OLclade_oi, "_jumpSummaries4.tab"), quote = F)
}##finish looping over order level clades
```

#### Get some other genome features for comparison
The genome features include average GC content, genome size, and coding density.
```{r get_genomeFeatures, eval=F }
for(OLclade_n in 1:length(OLclades_oi)){
  OLclade_oi <- OLclades_oi[OLclade_n]
  print(paste0("Getting genome features for jumps within ", OLclade_oi))
  pp_cutoff <- pp_cutoffs[OLclade_n]
  alpha_oi <- alphas_oi[OLclade_n]
  p_folder <- p_folders[OLclade_n]
  res_dir <- out_dir <-  paste0(levolution_out_dir ,p_folder,"_GTDB_14k_levolution/")

  tree_pp <- read.newick(file = paste0(res_dir,OLclade_oi,"_GTDB_14k_GC_config", config_oi, "_alpha", alpha_oi,".post"))
  tree_in <- read.newick(paste0(res_dir,OLclade_oi,"_GTDB_14k_tree.newick"))
  OLclade_nTips <- length(tree_in$tip.label) ##OL=OrderLevel
  
  OLclade_jumpBranches <- which(tree_pp$edge.length>pp_cutoff)
  jumpClade_details <- read.table(file = paste0(jump_common_dir, OLclade_oi, "_jumpSummaries4.tab"), stringsAsFactors = F)
  
  for(jumpNum_oi in which(jumpClade_details$codon_counting==T)){
    print(jumpNum_oi)
    jumpBranch_oi <- OLclade_jumpBranches[jumpNum_oi]
    
    jumpDir <- paste0(jump_common_dir, OLclade_oi, "_jump", jumpNum_oi, "/")
    all_accNums <- read.table(paste0(jumpDir, "ass_fn_GCFids_found.txt"), stringsAsFactors = F)
    GTDB_accNums_map <- all_accNums$V1
    GTDB_labels <- read.table(paste0(jumpDir, "required_GCFonly_ids.txt"), stringsAsFactors = F)
    names(GTDB_accNums_map) <- GTDB_labels$V1
    
    clade_genomeFeatures <- GTDB_14k_metadata[GTDB_labels$V1, c("genome_size", "coding_density", "gc_percentage")]
    write.table(clade_genomeFeatures, paste0(jumpDir, "genome_features_GTDB.txt"))
  }
}
```


#### Actual comparison of CUB
```{r get_rRNA_tRNA_numbers }
get_RNA_nums <- function(genome_oi){
  ftcnt_fn <- paste0(gd_base,"feature_counts/",genome_oi,"_feature_count.txt")
  if(file.exists(ftcnt_fn)){
    ftcnt_lines <- readLines(ftcnt_fn)
    rRNA_linen <- which(grepl(pattern = "gene	rRNA", x = ftcnt_lines, fixed = T))
    rRNA_num <- tryCatch({
      strsplit(ftcnt_lines[rRNA_linen], split = "\t", fixed = T)[[1]][7]
      }, error=function(e){
      print(e)
      return(NA)
    })
    
    tRNA_linen <- which(grepl(pattern = "gene	tRNA", x = ftcnt_lines, fixed = T))
    tRNA_num <- tryCatch({
      strsplit(ftcnt_lines[tRNA_linen], split = "\t", fixed = T)[[1]][7]
      }, error=function(e){
      return(NA)
    })
    
    return(c(rRNA_num, tRNA_num))
  }else{
    return(c(NA,NA))
  }
}
```

```{r compare_CUB }
for(OLclade_n in 1:length(OLclades_oi)){
  OLclade_oi <- OLclades_oi[OLclade_n]
  print(paste0("Analysing jumps within ", OLclade_oi))

  jumpClade_details <- read.table(file = paste0(jump_common_dir, OLclade_oi, "_jumpSummaries4.tab"), stringsAsFactors = F)
  
  colNames_CUB_details <- c("focal_CUB", "sis_CUB", "focal_GC", "sis_GC", "nFocal_valid", "nRelated_valid", "focal_rRNA_avg", "sis_rRNA_avg", "focal_tRNA_avg", "sis_tRNA_avg")
  CUB_details <- as.data.frame(matrix(NA, nrow = nrow(jumpClade_details), ncol = length(colNames_CUB_details), dimnames = list(paste0(OLclade_oi,"_jump",1:nrow(jumpClade_details)), colNames_CUB_details)))
  
  for(jumpNum_oi in which(jumpClade_details$codon_counting==T)){#only for those jumps which are analyzable
    print(paste0("Counting codons for jump number ", jumpNum_oi))
    
    jumpDir <- paste0(jump_common_dir, OLclade_oi, "_jump", jumpNum_oi, "/")
    all_accNums <- read.table(paste0(jumpDir, "ass_fn_GCFids_found.txt"), stringsAsFactors = F)
    GTDB_accNums_map <- all_accNums$V1
    GTDB_labels <- read.table(paste0(jumpDir, "required_GCFonly_ids.txt"), stringsAsFactors = F)
    names(GTDB_accNums_map) <- GTDB_labels$V1
    
    all_cdsAccNums <- read.table(file = paste0(jumpDir,"cds_fn_GCFids_found.txt"),stringsAsFactors = F, header = F)
    all_cdsAccNums <- all_cdsAccNums$V1
    
    focal_GTDBLabels <- read.table(paste0(jumpDir, OLclade_oi,"_jump",jumpNum_oi,"_focalGTDBLabels.tab"), stringsAsFactors = F, header = F)
    focal_accNums <- sapply(focal_GTDBLabels$V1, function(x) GTDB_accNums_map[x], simplify = T, USE.NAMES = F)
    sis_GTDBLabels <- read.table(paste0(jumpDir,OLclade_oi,"_jump",jumpNum_oi,"_sisGTDBLabels.tab"), stringsAsFactors = F, header = F)
    sis_accNums <- sapply(sis_GTDBLabels$V1, function(x) GTDB_accNums_map[x], simplify = T, USE.NAMES = F)
    out1_GTDBLabels <- read.table(paste0(jumpDir,OLclade_oi,"_jump",jumpNum_oi,"_out1GTDBLabels.tab"), stringsAsFactors = F, header = F)
    out1_accNums <- sapply(out1_GTDBLabels$V1, function(x) GTDB_accNums_map[x], simplify = T, USE.NAMES = F)
       
    nested_focalTip_labels <- tryCatch({
      read.table(file = paste0(jumpDir, "nestedFocalTips.tab"), stringsAsFactors = F)$V1
    }, error=function(e){
      print(e)
      return(NA)
    })
    
    valid_focal_accNums <- setdiff(setdiff(focal_accNums, GTDB_accNums_map[nested_focalTip_labels]), NA)
    valid_sis_accNums <- setdiff(setdiff(sis_accNums, GTDB_accNums_map[nested_focalTip_labels]), NA)
    valid_out1_accNums <- setdiff(setdiff(out1_accNums, GTDB_accNums_map[nested_focalTip_labels]), NA)
    
    aawise_S_all <- read.table(file = paste0(jumpDir, "CUB_Sharp.txt"), stringsAsFactors = F)
    clade_genomeFeatures <- read.table(paste0(jumpDir, "genome_features_GTDB.txt"), stringsAsFactors = F)
    all_RNA_nums <- sapply(c(valid_focal_accNums, valid_sis_accNums, valid_out1_accNums), function(x) get_RNA_nums(x))
    all_RNA_nums <- t(all_RNA_nums)
    
    CUB_details[paste0(OLclade_oi,"_jump", jumpNum_oi),"focal_CUB"] <- median(aawise_S_all[valid_focal_accNums,"S_avg"], na.rm = T)
    CUB_details[paste0(OLclade_oi,"_jump", jumpNum_oi),"sis_CUB"] <- median(aawise_S_all[c(valid_sis_accNums),"S_avg"], na.rm = T) #,valid_out1_accNums
    CUB_details[paste0(OLclade_oi,"_jump", jumpNum_oi),"focal_GC"] <- median(clade_genomeFeatures[GTDB_labels$V1[which(all_accNums$V1 %in% valid_focal_accNums)],"gc_percentage"], na.rm = T)
    CUB_details[paste0(OLclade_oi,"_jump", jumpNum_oi),"sis_GC"] <- median(clade_genomeFeatures[GTDB_labels$V1[which(all_accNums$V1 %in% c(valid_sis_accNums))],"gc_percentage"], na.rm = T) #, valid_out1_accNums
    CUB_details[paste0(OLclade_oi,"_jump", jumpNum_oi),"nFocal_valid"] <-
      length(which(!is.na(aawise_S_all[valid_focal_accNums,"S_avg"])))
    CUB_details[paste0(OLclade_oi,"_jump", jumpNum_oi),"nRelated_valid"] <- length(which(!is.na(aawise_S_all[c(valid_sis_accNums),"S_avg"]))) #,valid_out1_accNums
    CUB_details[paste0(OLclade_oi,"_jump", jumpNum_oi),"focal_rRNA_avg"] <- median(as.numeric(all_RNA_nums[valid_focal_accNums,1]), na.rm = T)/3
    CUB_details[paste0(OLclade_oi,"_jump", jumpNum_oi),"sis_rRNA_avg"] <- median(as.numeric(all_RNA_nums[c(valid_sis_accNums),1]), na.rm = T)/3 #,valid_out1_accNums
    CUB_details[paste0(OLclade_oi,"_jump", jumpNum_oi),"focal_tRNA_avg"] <- median(as.numeric(all_RNA_nums[valid_focal_accNums,2]), na.rm = T)
    CUB_details[paste0(OLclade_oi,"_jump", jumpNum_oi),"sis_tRNA_avg"] <- median(as.numeric(all_RNA_nums[c(valid_sis_accNums),2]), na.rm = T)#,valid_out1_accNums
  }##finish loop over jumps in this order lvel clade
  write.table(CUB_details, file = paste0(jump_common_dir, OLclade_oi, "_CUB_details.tab"), quote = F)
}##finish loops over order level clade
```

#### Plotting of CUB comparisons
```{r write_all_CUB }
all_CUB_comparisons <- do.call(rbind, lapply(OLclades_oi, function(x) read.table(paste0(jump_common_dir,x,"_CUB_details.tab"), stringsAsFactors = F, header = T, row.names = 1)))
all_CUB_comparisons$CUBratio <- all_CUB_comparisons$focal_CUB/all_CUB_comparisons$sis_CUB
write.table(all_CUB_comparisons, file = paste0(jump_common_dir, "levolution_jumps_CUB_comparison.tab"), quote = F)
```

```{r plot_CUB_comparison }
all_CUB_comparisons <- read.table(file = paste0(jump_common_dir,"levolution_jumps_CUB_comparison.tab"), stringsAsFactors = F)

plt1 <- ggplot(data = all_CUB_comparisons, aes(x=sis_CUB, y=focal_CUB)) + geom_point(shape=21, color="black") + scale_x_continuous(name = "sister/outgroup CUB") + scale_y_continuous(name = "focal CUB") + geom_abline(slope = 1, intercept = 0, color="grey")

CUBratio_plt <- ggplot(data = all_CUB_comparisons, aes(x=focal_GC - sis_GC, y=CUBratio)) + geom_point(shape=21, color="black") + scale_x_continuous(name = expression(Delta*"GC (focal clade - sister clade)")) + scale_y_continuous(name = "focal CUB/\n related CUB", trans = "log2") + geom_hline(yintercept = 1, color="grey") + geom_hline(yintercept = 0.75, color="grey", linetype="dashed") + geom_hline(yintercept = 1.5, color="grey", linetype="dashed")

CUB_plt <- plot_grid(plotlist = list(plt1, CUBratio_plt), align = "hv", nrow = 1, ncol = 2, labels = "AUTO", label_size = 20)
save_plot(filename = paste0(jump_common_dir,"CUB_comparison.svg"), CUB_plt, ncol = 2, base_height = 3.6, base_width = 4.5)
CUB_plt
```

### 16s analysis: comparison of 16s rRNA evolutionary rate

#### Extract 16s gene sequences
```{r extract_16s }
for(OLclade_n in 1:length(OLclades_oi)){
  OLclade_oi <- OLclades_oi[OLclade_n]
  print(paste0("Analysing jumps within ", OLclade_oi))

  jumpClade_details <- read.table(file = paste0(jump_common_dir, OLclade_oi, "_jumpSummaries4.tab"), stringsAsFactors = F)
  jumpClade_details$rna_analysis <- NA
  
  for(jumpNum_oi in which(jumpClade_details$depth %in% c("outgroup1", "outgroup2") & jumpClade_details$genomes_downloaded==T)){#only for those jumps which are analyzable
    print(paste0("Comparing 16s evolution rate for jump number ", jumpNum_oi))
    
    jumpDir <- paste0(jump_common_dir, OLclade_oi, "_jump", jumpNum_oi, "/")
    
    all_accNums <- read.table(paste0(jumpDir, "ass_fn_GCFids_found.txt"), stringsAsFactors = F)
    GTDB_accNums_map <- all_accNums$V1
    GTDB_labels <- read.table(paste0(jumpDir, "required_GCFonly_ids.txt"), stringsAsFactors = F)
    names(GTDB_accNums_map) <- GTDB_labels$V1
    
    nested_focalTip_labels <- tryCatch({
      read.table(file = paste0(jumpDir, "nestedFocalTips.tab"), stringsAsFactors = F)$V1
    }, error=function(e){
      print(e)
      return(NA)
    })
  
    focal_GTDBLabels <- read.table(paste0(jumpDir, OLclade_oi,"_jump",jumpNum_oi,"_focalGTDBLabels.tab"), stringsAsFactors = F, header = F)
    focal_GTDBLabels_valid <- setdiff(focal_GTDBLabels$V1, nested_focalTip_labels)
    focal_accNums <- sapply(focal_GTDBLabels_valid, function(x) GTDB_accNums_map[x], simplify = T, USE.NAMES = F)
    
    sis_GTDBLabels <- read.table(paste0(jumpDir,OLclade_oi,"_jump",jumpNum_oi,"_sisGTDBLabels.tab"), stringsAsFactors = F, header = F)
    sis_GTDBLabels_valid <- setdiff(sis_GTDBLabels$V1, nested_focalTip_labels)
    sis_accNums <- sapply(sis_GTDBLabels_valid, function(x) GTDB_accNums_map[x], simplify = T, USE.NAMES = F)
    
    out1_GTDBLabels <- read.table(paste0(jumpDir,OLclade_oi,"_jump",jumpNum_oi,"_out1GTDBLabels.tab"), stringsAsFactors = F, header = F)
    out1_GTDBLabels_valid <- setdiff(out1_GTDBLabels$V1, nested_focalTip_labels)
    out1_accNums <- sapply(out1_GTDBLabels_valid, function(x) GTDB_accNums_map[x], simplify = T, USE.NAMES = F)
    
    out2_accNums <- NA
    if(jumpClade_details[paste0("jump",jumpNum_oi), "depth"]=="outgroup2"){
      out2_GTDBLabels <- read.table(paste0(jumpDir,OLclade_oi,"_jump",jumpNum_oi,"_out2GTDBLabels.tab"), stringsAsFactors = F, header = F)
      out2_GTDBLabels_valid <- setdiff(out2_GTDBLabels$V1, nested_focalTip_labels)
      out2_accNums <- sapply(out2_GTDBLabels_valid, function(x) GTDB_accNums_map[x], simplify = T, USE.NAMES = F)
    }
    
    rna_accNums_available <- read.table(paste0(jumpDir, "rna_fn_GCFids_found.txt"), stringsAsFactors = F)
    rna_accNums_available <- setdiff(rna_accNums_available$V1, NA)
    
    if(!any(focal_accNums %in% rna_accNums_available) || !any(sis_accNums %in% rna_accNums_available) || !any(c(out1_accNums, out2_accNums) %in% rna_accNums_available)){
      print("Cannot compare 16s evolutionary rates, RNA files not available")
      jumpClade_details[paste0("jump",jumpNum_oi),"rna_analysis"] <- "rna_files_unav"
      next
    }

    SixteenS_focal_found <- grepl(pattern = "16S ribosomal RNA|small Subunit ribosomal RNA", x = sapply(intersect(focal_accNums, rna_accNums_available), function(x) read_file(paste0(gd_base,"rna/",x,"_rna_from_genomic.fna"))))
    SixteenS_sister_found<- grepl(pattern = "16S ribosomal RNA|small Subunit ribosomal RNA", x = sapply(intersect(sis_accNums, rna_accNums_available), function(x) read_file(paste0(gd_base,"rna/",x,"_rna_from_genomic.fna"))))
    SixteenS_out_found<- grepl(pattern = "16S ribosomal RNA|small Subunit ribosomal RNA", x = sapply(intersect(c(out1_accNums, out2_accNums), rna_accNums_available), function(x) read_file(paste0(gd_base,"rna/",x,"_rna_from_genomic.fna"))))

    if(!any(SixteenS_focal_found) || !any(SixteenS_sister_found) || !any(SixteenS_out_found)){
      print("Cannot compare 16s evolutionary rates, could not find 16s sequences")
      jumpClade_details[paste0("jump",jumpNum_oi),"rna_analysis"] <- "16s_seqs_unav"
      next
    }

    SixteenS_fn <- paste0(jumpDir,"16s_seqs.fa")
    if(file.exists(SixteenS_fn)) system2(command = "rm", args = c(SixteenS_fn))
    
    #extract and compile 16s sequences
    for(set_oi in list(focal_accNums, sis_accNums, c(out1_accNums, out2_accNums))){
      any_found <- F
      for(accNum_oi in intersect(set_oi,rna_accNums_available)){
        rna_seqs <- read.fasta(paste0(gd_base,"rna/",accNum_oi,"_rna_from_genomic.fna"))
        SixteenS_seqs <- lapply(which(grepl(pattern = "16S ribosomal RNA|small Subunit ribosomal RNA", x = getAnnot(rna_seqs))), function(x) rna_seqs[[x]])
        if(length(SixteenS_seqs)==0){
          # print(paste0("No 16S sequence found in this genome ", accNum_oi))
        }else if(length(SixteenS_seqs)==1){
          if(length(SixteenS_seqs[[1]])>1200){
            any_found <- T
            write.fasta(SixteenS_seqs,names = accNum_oi, file.out = SixteenS_fn, open = "a")
          } 
        }else if(length(SixteenS_seqs)>1){
          print(paste0("Multiple 16s sequences found in this genome ", accNum_oi))
          valid_length_16S <- which(sapply(SixteenS_seqs, function(x) getLength(x))>1200 & sapply(SixteenS_seqs, function(x) getLength(x))<1700)
          if(identical(valid_length_16S, integer(0))){
            print("No 16s of valid length")
            next
          }else if(length(valid_length_16S)==1){
            any_found <- T
            write.fasta(SixteenS_seqs[valid_length_16S],names = accNum_oi, file.out = SixteenS_fn, open = "a")
          }else{
            any_found <- T
            write.fasta(SixteenS_seqs[valid_length_16S], names = paste0("16s_",valid_length_16S), file.out = paste0(jumpDir,accNum_oi,"_16s.fa"))
            system2(command = "muscle", args = c("-in", paste0(jumpDir,accNum_oi,"_16s.fa"), "-out", paste0(jumpDir,accNum_oi,"_16s.aln")), stdout = F, stderr = F)
            aln_oi <- read.alignment(file = paste0(jumpDir,accNum_oi,"_16s.aln"), format = "fasta")
            dist_aln <- dist.alignment(aln_oi)
            if(max(dist_aln, na.rm = T)^2<0.02){ #if maximum difference is 1%
              chosen_16s_n <- sample(valid_length_16S, size = 1)
              write.fasta(SixteenS_seqs[chosen_16s_n],names = accNum_oi, file.out = SixteenS_fn, open = "a")
            }else{
              print(paste0("Divergent 16s sequences found in ", accNum_oi, " of jump number ", jumpNum_oi))
              print(max(dist_aln, na.rm = T)^2)
            }#check 16s distances
          }##end if, is there any sequence more than cutoff length
        }##when there are multiple 16s RNAs
      }##iterating over accession numbers within set
      if(!any_found){
        print("No 16s of valid length in either focal, sister, or outgroup clade")
        jumpClade_details[paste0("jump",jumpNum_oi),"rna_analysis"] <- "valid_seqs_unav"
        break
      }else{
        jumpClade_details[paste0("jump",jumpNum_oi),"rna_analysis"] <- "yes"
      }
    }##iterating over focal, sister, and out sets
    if(!any_found){
        print("No 16s of valid length in either focal, sister, or outgroup clade")
        jumpClade_details[paste0("jump",jumpNum_oi),"rna_analysis"] <- "valid_seqs_unav"
        next
      }else{
        jumpClade_details[paste0("jump",jumpNum_oi),"rna_analysis"] <- "yes"
      }
  }##iterating over jumps
  write.table(jumpClade_details, file = paste0(jump_common_dir, OLclade_oi, "_jumpSummaries4.tab"), quote = F)
}##iterating over order level clades
```

#### Compare the branch lengths of focal vs sister clade in a 16s tree
```{r compare_16s_branch_lengths_iqtree }
for(OLclade_n in 1:length(OLclades_oi)){
  OLclade_oi <- OLclades_oi[OLclade_n]
  print(paste0("Analysing jumps within ", OLclade_oi))

  jumpClade_details <- read.table(file = paste0(jump_common_dir, OLclade_oi, "_jumpSummaries4.tab"), stringsAsFactors = F)
  
  colNames_16s_details <- c("rna_focal_bLength", "rna_sis_bLength", "rna_focal_GC", "rna_sis_GC", "focal_GC", "sis_GC")
  SixteenS_details <- as.data.frame(matrix(data = NA, nrow = length(which(jumpClade_details$rna_analysis=="yes")), ncol = length(colNames_16s_details), dimnames = list(paste0(OLclade_oi, "_jump", which(jumpClade_details$rna_analysis=="yes")), colNames_16s_details)))

  if(!any(jumpClade_details$rna_analysis=="yes")){
    print("No jumps that can be analyzed for 16s branch lengths here")
    write(x = NULL, file = paste0(jump_common_dir, OLclade_oi,"_16s_details.tab"))
    next
  } 

  for(jumpNum_oi in which(jumpClade_details$rna_analysis=="yes")){#only for those jumps which are analyzable
    print(paste0("Comparing 16s evolution rate for jump number ", jumpNum_oi))
    
    jumpDir <- paste0(jump_common_dir, OLclade_oi, "_jump", jumpNum_oi, "/")
    
    all_accNums <- read.table(paste0(jumpDir, "ass_fn_GCFids_found.txt"), stringsAsFactors = F)
    GTDB_accNums_map <- all_accNums$V1
    GTDB_labels <- read.table(paste0(jumpDir, "required_GCFonly_ids.txt"), stringsAsFactors = F)
    names(GTDB_accNums_map) <- GTDB_labels$V1
    
    nested_focalTip_labels <- tryCatch({
      read.table(file = paste0(jumpDir, "nestedFocalTips.tab"), stringsAsFactors = F)$V1
    }, error=function(e){
      print(e)
      return(NA)
    })
    
    focal_GTDBLabels <- read.table(paste0(jumpDir, OLclade_oi,"_jump",jumpNum_oi,"_focalGTDBLabels.tab"), stringsAsFactors = F, header = F)
    focal_GTDBLabels_valid <- setdiff(focal_GTDBLabels$V1, nested_focalTip_labels)
    focal_accNums <- sapply(focal_GTDBLabels_valid, function(x) GTDB_accNums_map[x], simplify = T, USE.NAMES = F)
    
    sis_GTDBLabels <- read.table(paste0(jumpDir,OLclade_oi,"_jump",jumpNum_oi,"_sisGTDBLabels.tab"), stringsAsFactors = F, header = F)
    sis_GTDBLabels_valid <- setdiff(sis_GTDBLabels$V1, nested_focalTip_labels)
    sis_accNums <- sapply(sis_GTDBLabels_valid, function(x) GTDB_accNums_map[x], simplify = T, USE.NAMES = F)
    
    out1_GTDBLabels <- read.table(paste0(jumpDir,OLclade_oi,"_jump",jumpNum_oi,"_out1GTDBLabels.tab"), stringsAsFactors = F, header = F)
    out1_GTDBLabels_valid <- setdiff(out1_GTDBLabels$V1, nested_focalTip_labels)
    out1_accNums <- sapply(out1_GTDBLabels_valid, function(x) GTDB_accNums_map[x], simplify = T, USE.NAMES = F)
    
    out2_accNums <- NA
    if(jumpClade_details[paste0("jump",jumpNum_oi), "depth"]=="outgroup2"){
      out2_GTDBLabels <- read.table(paste0(jumpDir,OLclade_oi,"_jump",jumpNum_oi,"_out2GTDBLabels.tab"), stringsAsFactors = F, header = F)
      out2_GTDBLabels_valid <- setdiff(out2_GTDBLabels$V1, nested_focalTip_labels)
      out2_accNums <- sapply(out2_GTDBLabels_valid, function(x) GTDB_accNums_map[x], simplify = T, USE.NAMES = F)
    }
    
    SixteenS_fn <- paste0(jumpDir,"16s_seqs.fa")
    SixteenS_aln_fn <- paste0(jumpDir,"16s_seqs.aln")
    system2(command = "muscle", args = c("-in", SixteenS_fn, "-out", SixteenS_aln_fn), stdout = F, stderr = F)

    SixteenS_aln <- read.alignment(SixteenS_aln_fn, format = "fasta")
    av_SixteenS <- SixteenS_aln$nam


    jumpClade_tree_GCF <- focal_clade_tree <- read.tree(paste0(jumpDir, OLclade_oi, "_jump", jumpNum_oi, ".newick"))
    jumpClade_tree_GCF$tip.label <- sapply(focal_clade_tree$tip.label, function(x) GTDB_accNums_map[x])
    jumpClade_subtree_GCF <- keep.tip(jumpClade_tree_GCF, av_SixteenS)
    jumpClade_subtree_GCF_fn <- paste0(jumpDir,"16s_constraint.tre")
    write.tree(unroot(jumpClade_subtree_GCF), jumpClade_subtree_GCF_fn)
    if(length(jumpClade_subtree_GCF$tip.label)<4){ #no constraint with 3 nodes for unrooted tree
      system2(command = "iqtree", args = c("-redo", "-s", SixteenS_aln_fn), stdout = F, stderr = F)
    }else{
      system2(command = "iqtree", args = c("-redo", "-s", SixteenS_aln_fn, "-g", jumpClade_subtree_GCF_fn), stdout = F, stderr = F)
    }

    SixteenS_iqtree <- read.tree(paste0(SixteenS_aln_fn,".treefile"))
    # plot(jumpClade_subtree_GCF)
    if(any(out2_accNums %in% SixteenS_iqtree$tip.label)){
      SixteenS_iqtree_rooted <- root(phy = SixteenS_iqtree, outgroup = intersect(out2_accNums, SixteenS_iqtree$tip.label), resolve.root = T)
    }else if(any(out1_accNums %in% SixteenS_iqtree$tip.label)){
      SixteenS_iqtree_rooted <- root(phy = SixteenS_iqtree, outgroup = intersect(out1_accNums, SixteenS_iqtree$tip.label), resolve.root = T)
    }else{
      print("No outgroup found. This is strange!")
    }
    rna_GC <- sapply(SixteenS_iqtree_rooted$tip.label, function(x) GC(s2c(SixteenS_aln$seq[which(SixteenS_aln$nam==x)][[1]])))

    rna_focal_accNums <- intersect(focal_accNums, SixteenS_iqtree_rooted$tip.label)
    rna_sis_accNums <- intersect(sis_accNums, SixteenS_iqtree_rooted$tip.label)

    bar_colors <- rep("grey", length(SixteenS_iqtree_rooted$tip.label))
    bar_colors[which(SixteenS_iqtree_rooted$tip.label %in% rna_focal_accNums)] <- "orange"
    bar_colors[which(SixteenS_iqtree_rooted$tip.label %in% rna_sis_accNums)] <- "purple"
    names(bar_colors) <- SixteenS_iqtree_rooted$tip.label

    svg(filename = paste0(jumpDir, "16s_tree.svg"), width = 15)
    SixteenS_plt <- plotTree.barplot(tree = SixteenS_iqtree_rooted, x = rna_GC, args.barplot = list(xlab="GC % (16S)"))
    dev.off()

    SixteenS_iqtree_rooted_labeled <- SixteenS_iqtree_rooted
    SixteenS_iqtree_rooted_labeled$tip.label <- sapply(SixteenS_iqtree_rooted$tip.label, function(x) get_genera_sp_names(full_GTDB_taxonomy = GTDB_14k_metadata[names(GTDB_accNums_map)[which(GTDB_accNums_map==x)],"gtdb_taxonomy"]))
    svg(filename = paste0(jumpDir, "16s_tree_colored.svg"), width = 15)
    SixteenS_plt <- plot(SixteenS_iqtree_rooted_labeled,tip.color = bar_colors)
    add.scale.bar()
    dev.off()

    rna_anc_node <- getMRCA(phy = SixteenS_iqtree_rooted, tip = c(rna_focal_accNums, rna_sis_accNums))

    rna_focal_bLength <- median(sapply(rna_focal_accNums, function(x) nodeheight(SixteenS_iqtree_rooted,which(SixteenS_iqtree_rooted$tip.label==x)))) - nodeheight(tree = SixteenS_iqtree_rooted, node =rna_anc_node)
    rna_sis_bLength <- median(sapply(rna_sis_accNums, function(x) nodeheight(SixteenS_iqtree_rooted,which(SixteenS_iqtree_rooted$tip.label==x)))) - nodeheight(tree = SixteenS_iqtree_rooted, node = rna_anc_node)
    rna_focal_GC <- median(rna_GC[rna_focal_accNums])
    rna_sis_GC <- median(rna_GC[rna_sis_accNums])

    GTDB_GC <- GTDB_14k_metadata[names(GTDB_accNums_map)[GTDB_accNums_map %in% SixteenS_iqtree_rooted$tip.label],"gc_percentage"]
    names(GTDB_GC) <- GTDB_accNums_map[GTDB_accNums_map %in% SixteenS_iqtree_rooted$tip.label]
    focal_gGC <- median(GTDB_GC[rna_focal_accNums])
    sis_gGC <- median(GTDB_GC[rna_sis_accNums])

    SixteenS_details[paste0(OLclade_oi, "_jump",jumpNum_oi), "rna_focal_bLength"] <- rna_focal_bLength
    SixteenS_details[paste0(OLclade_oi, "_jump",jumpNum_oi), "rna_sis_bLength"] <- rna_sis_bLength
    SixteenS_details[paste0(OLclade_oi, "_jump",jumpNum_oi), "rna_focal_GC"] <- rna_focal_GC
    SixteenS_details[paste0(OLclade_oi, "_jump",jumpNum_oi), "rna_sis_GC"] <- rna_sis_GC
    SixteenS_details[paste0(OLclade_oi, "_jump",jumpNum_oi), "focal_GC"] <- focal_gGC
    SixteenS_details[paste0(OLclade_oi, "_jump",jumpNum_oi), "sis_GC"] <- sis_gGC

  }##end of loop over jumps within a clade
  write.table(SixteenS_details, file = paste0(jump_common_dir, OLclade_oi, "_16s_details.tab"), quote = F)
}##end of loop over order-lvel clades

```

#### Plot comparison of 16s evolutionary rates
```{r collate_16s_comparisons }
all_16s_comparisons <- do.call(rbind, lapply(OLclades_oi, function(x) if(file.exists(paste0(jump_common_dir,x,"_16s_details.tab"))) read.table(paste0(jump_common_dir,x,"_16s_details.tab"), stringsAsFactors = F, header = T, row.names = 1)))
all_16s_comparisons$SixteenSratio <- all_16s_comparisons$rna_focal_bLength/all_16s_comparisons$rna_sis_bLength
write.table(all_16s_comparisons, file = paste0(jump_common_dir, "levolution_jumps_16s_comparison.tab"), quote = F)
```

```{r plot_16s_bLengths }
all_16s_comparisons <- read.table(file = paste0(jump_common_dir,"levolution_jumps_16s_comparison.tab"), stringsAsFactors = F)
plt1 <- ggplot(data = all_16s_comparisons, aes(x=rna_sis_bLength, y=rna_focal_bLength)) + geom_point(col="darkblue", shape=21) + scale_x_continuous(name = "sister branch length (16S)") + scale_y_continuous(name = "focal branch length (16s)") + geom_abline(slope = 1, intercept = 0, col="grey")

plt2 <- ggplot(data = all_16s_comparisons, aes(x=focal_GC - sis_GC, y=100*(rna_focal_GC -  rna_sis_GC))) + geom_point(col="darkblue", shape=21) + geom_smooth(method = "lm", col="grey") + scale_x_continuous(name = expression(Delta*"GC (genomic)")) + scale_y_continuous(name = expression(Delta*"GC (16s)"))

SixteenSratio_rnaGC_plt <- ggplot(data = all_16s_comparisons, aes(x=focal_GC - sis_GC, y=rna_focal_bLength/rna_sis_bLength)) + geom_point(col="darkblue", shape=21) + scale_x_continuous(name = expression(Delta*"GC (genomic)")) + scale_y_continuous(name = "focal 16s branch length/\n sister 16s branch length", trans = "log2", limits = c(1/(2^6),2^5), breaks = c(1/(2^6), 1/(2^5), 1/(2^4), 1/(2^2), 1/(2^0), 2^2, 2^4, 2^5)) + geom_hline(yintercept = 1, col="grey")

SixteenSratio_GC_plt <- ggplot(data = all_16s_comparisons, aes(x=rna_focal_GC - rna_sis_GC, y=rna_focal_bLength/rna_sis_bLength)) + geom_point(col="darkblue", shape=21) + scale_x_continuous(name = expression(Delta*"GC (16s)")) + scale_y_continuous(name = "focal 16s branch length/\n sister 16s branch length", trans = "log2", limits = c(1/(2^6),2^5), breaks = c(1/(2^6), 1/(2^5), 1/(2^4), 1/(2^2), 1/(2^0), 2^2, 2^4, 2^5)) + geom_hline(yintercept = 1, col="grey")

bLength_plt <- plot_grid(plotlist = list(plt1, plt2, SixteenSratio_rnaGC_plt, SixteenSratio_GC_plt), align = "hv", nrow = 2, ncol = 2, labels = "AUTO", label_size = 20)
save_plot(filename = paste0(jump_common_dir,"16s_comparison.svg"), bLength_plt, ncol = 2, base_height = 7, base_width = 4.5)
bLength_plt
```

### Comparison of intergenic region distribution

#### Function to extract IGR lengths
```{r get_igr_lengths_fn }
get_igr_lengths <- function(accNum_oi){
  ft_table <- read.table(paste0(gd_base,"feature_tables/",accNum_oi,"_feature_table.txt"), stringsAsFactors = F, sep = "\t", quote = "")
  igr_lengths <- sapply(which(ft_table$V1=="gene"), function(x) ifelse(ft_table[x+2,"V8"]>ft_table[x,"V9"], ft_table[x+2,"V8"] - ft_table[x,"V9"], NA))
  median_oi <- mean(igr_lengths, na.rm = T) #median(igr_lengths, na.rm = T)
  quantiles_oi <- quantile(igr_lengths, probs = c(0.25, 0.75), na.rm = T)
  return(c(median_oi, quantiles_oi[1], quantiles_oi[2]))
}
```

#### Extract intergenic region lengths from feature table
```{r extract_IGR_lengths }
for(OLclade_n in 1:length(OLclades_oi)){
  OLclade_oi <- OLclades_oi[OLclade_n]
  print(paste0("Analysing jumps within ", OLclade_oi))

  jumpClade_details <- read.table(file = paste0(jump_common_dir, OLclade_oi, "_jumpSummaries4.tab"), stringsAsFactors = F)
  jumpClade_details$igr_analysis <- NA
  
  colNames_IGR_details <- c("focal_igrLength", "sis_igrLength", "focal_igrQ1", "sis_igrQ1", "focal_igrQ2", "sis_igrQ2", "focal_GC", "sis_GC")
  IGR_details <- as.data.frame(matrix(NA, nrow = nrow(jumpClade_details), ncol = length(colNames_IGR_details), dimnames = list(paste0(OLclade_oi,"_jump",1:nrow(jumpClade_details)), colNames_IGR_details)))

  for(jumpNum_oi in which(jumpClade_details$depth %in% c("sister", "outgroup1", "outgroup2") & jumpClade_details$genomes_downloaded==T)){#only for those jumps which are analyzable
    print(paste0("Comparing intergenic region lengths for jump number ", jumpNum_oi))
    
    jumpDir <- paste0(jump_common_dir, OLclade_oi, "_jump", jumpNum_oi, "/")
    
    all_accNums <- read.table(paste0(jumpDir, "ass_fn_GCFids_found.txt"), stringsAsFactors = F)
    GTDB_accNums_map <- all_accNums$V1
    GTDB_labels <- read.table(paste0(jumpDir, "required_GCFonly_ids.txt"), stringsAsFactors = F)
    names(GTDB_accNums_map) <- GTDB_labels$V1
    
    nested_focalTip_labels <- tryCatch({
      read.table(file = paste0(jumpDir, "nestedFocalTips.tab"), stringsAsFactors = F)$V1
    }, error=function(e){
      print(e)
      return(NA)
    })
    
    focal_GTDBLabels <- read.table(paste0(jumpDir, OLclade_oi,"_jump",jumpNum_oi,"_focalGTDBLabels.tab"), stringsAsFactors = F, header = F)
    focal_GTDBLabels_valid <- setdiff(focal_GTDBLabels$V1, nested_focalTip_labels)
    focal_accNums <- sapply(focal_GTDBLabels_valid, function(x) GTDB_accNums_map[x], simplify = T, USE.NAMES = F)
    
    sis_GTDBLabels <- read.table(paste0(jumpDir,OLclade_oi,"_jump",jumpNum_oi,"_sisGTDBLabels.tab"), stringsAsFactors = F, header = F)
    sis_GTDBLabels_valid <- setdiff(sis_GTDBLabels$V1, nested_focalTip_labels)
    sis_accNums <- sapply(sis_GTDBLabels_valid, function(x) GTDB_accNums_map[x], simplify = T, USE.NAMES = F)
    
    out1_accNums <- NA
    if(jumpClade_details[paste0("jump",jumpNum_oi), "depth"] %in% c("outgroup1", "outgroup2")){
      out1_GTDBLabels <- read.table(paste0(jumpDir,OLclade_oi,"_jump",jumpNum_oi,"_out1GTDBLabels.tab"), stringsAsFactors = F, header = F)
      out1_GTDBLabels_valid <- setdiff(out1_GTDBLabels$V1, nested_focalTip_labels)
      out1_accNums <- sapply(out1_GTDBLabels_valid, function(x) GTDB_accNums_map[x], simplify = T, USE.NAMES = F)
    }
    
    out2_accNums <- NA
    if(jumpClade_details[paste0("jump",jumpNum_oi), "depth"]=="outgroup2"){
      out2_GTDBLabels <- read.table(paste0(jumpDir,OLclade_oi,"_jump",jumpNum_oi,"_out2GTDBLabels.tab"), stringsAsFactors = F, header = F)
      out2_GTDBLabels_valid <- setdiff(out2_GTDBLabels$V1, nested_focalTip_labels)
      out2_accNums <- sapply(out2_GTDBLabels_valid, function(x) GTDB_accNums_map[x], simplify = T, USE.NAMES = F)
    }
    
    ft_accNums_available <- read.table(paste0(jumpDir, "ft_fn_GCFids_found.txt"), stringsAsFactors = F)
    ft_accNums_available <- setdiff(ft_accNums_available$V1, NA)
    
    if(!any(focal_accNums %in% ft_accNums_available) || !any(c(sis_accNums) %in% ft_accNums_available)){ #, out1_accNums
      print("Cannot compare IGR lengths, feature tables not available")
      jumpClade_details[paste0("jump",jumpNum_oi),"igr_analysis"] <- "feat_tables_unav"
      next
    }
    
    median_igrLengths <- t(sapply(ft_accNums_available, function(x) get_igr_lengths(x)))
    
    median_igrFocal <- median(median_igrLengths[intersect(focal_accNums, ft_accNums_available),1], na.rm = T)
    median_igrSis <- median(median_igrLengths[intersect(sis_accNums, ft_accNums_available),1], na.rm = T)
    q1_igrFocal <- median(median_igrLengths[intersect(focal_accNums, ft_accNums_available),2], na.rm = T)
    q1_igrSis <- median(median_igrLengths[intersect(sis_accNums, ft_accNums_available),2], na.rm = T)
    q2_igrFocal <- median(median_igrLengths[intersect(focal_accNums, ft_accNums_available),3], na.rm = T)
    q2_igrSis <- median(median_igrLengths[intersect(sis_accNums, ft_accNums_available),3], na.rm = T)
    
    GTDB_GC <- GTDB_14k_metadata[names(GTDB_accNums_map)[GTDB_accNums_map %in% ft_accNums_available],"gc_percentage"]
    names(GTDB_GC) <- GTDB_accNums_map[GTDB_accNums_map %in% ft_accNums_available]
    focal_gGC <- median(GTDB_GC[intersect(focal_accNums, ft_accNums_available)])
    sis_gGC <- median(GTDB_GC[intersect(sis_accNums, ft_accNums_available)])
    
    IGR_details[paste0(OLclade_oi, "_jump",jumpNum_oi),] <- c(median_igrFocal, median_igrSis, q1_igrFocal, q1_igrSis, q2_igrFocal, q2_igrSis, focal_gGC, sis_gGC)
  }##end loop over jumps in a clade
  write.table(IGR_details, file = paste0(jump_common_dir, OLclade_oi, "_IGR_details.tab"), quote = F)
}##end loop over order lvel clades
```

#### Plot IGR comparisons
```{r collate_IGR_comparisons }
all_IGR_comparisons <- do.call(rbind, lapply(OLclades_oi, function(x) read.table(paste0(jump_common_dir,x,"_IGR_details.tab"), stringsAsFactors = F, header = T, row.names = 1)))
all_IGR_comparisons$IGRratio <- all_IGR_comparisons$focal_igrLength/all_IGR_comparisons$sis_igrLength
write.table(all_IGR_comparisons, file = paste0(jump_common_dir, "levolution_jumps_IGR_comparison.tab"), quote = F)
```

```{r plot_IGR }
all_IGR_comparisons <- read.table(file = paste0(jump_common_dir,"levolution_jumps_IGR_comparison.tab"), stringsAsFactors = F)

plt1 <- ggplot(data = all_IGR_comparisons, aes(x=sis_igrLength, y=focal_igrLength)) + geom_point(shape=21, color="darkblue") + scale_x_continuous(name = "sister/outgroup IGR length") + scale_y_continuous(name = "focal IGR length") + geom_abline(slope = 1, intercept = 0, color="grey")
IGRratio_plt <- ggplot(data = all_IGR_comparisons, aes(x=focal_GC - sis_GC, y=IGRratio)) + geom_point(shape=21, color="darkblue") + scale_x_continuous(name = expression(Delta*"GC (focal clade - sister clade)")) + scale_y_continuous(name = "focal IGR length/\n related IGR length", trans = "log2") + geom_hline(yintercept = 1, color="grey") + geom_hline(yintercept = 0.75, color="grey", linetype="dashed") + geom_hline(yintercept = 1.5, color="grey", linetype="dashed")

IGR_plt <- plot_grid(plotlist = list(plt1, IGRratio_plt), align = "hv", nrow = 1, ncol = 2, labels = "AUTO", label_size = 20)
save_plot(filename = paste0(jump_common_dir,"IGR_comparison.svg"), IGR_plt, ncol = 2, base_height = 3.2, base_width = 4)
IGR_plt
```
### Compare number of pseudogenes
#### Get number of pseudogenes
```{r get_pseudogene_numbers }
get_pseudogene_nums <- function(genome_oi){
  ftcnt_fn <- paste0(gd_base,"feature_counts/",genome_oi,"_feature_count.txt")
  if(file.exists(ftcnt_fn)){
    ftcnt_lines <- readLines(ftcnt_fn)
    pseudo_linen <- which(grepl(pattern = "gene	pseudogene", x = ftcnt_lines, fixed = T))
    pseudo_num <- tryCatch({
      strsplit(ftcnt_lines[pseudo_linen], split = "\t", fixed = T)[[1]][7]
      }, error=function(e){
      print(e)
      return(NA)
    })
    
    prot_linen <- which(grepl(pattern = "gene	protein_coding", x = ftcnt_lines, fixed = T))
    prot_num <- tryCatch({
      strsplit(ftcnt_lines[prot_linen], split = "\t", fixed = T)[[1]][7]
      }, error=function(e){
      print(e)
      return(NA)
    })
    
    return(c(pseudo_num, prot_num))
  }else{
    return(c(NA,NA))
  }
}
```

#### Compare pseudogene
```{r compare_pseudogene_fractions }
for(OLclade_n in 1:length(OLclades_oi)){
  OLclade_oi <- OLclades_oi[OLclade_n]
  print(paste0("Analysing jumps within ", OLclade_oi))

  jumpClade_details <- read.table(file = paste0(jump_common_dir, OLclade_oi, "_jumpSummaries4.tab"), stringsAsFactors = F)
  
  colNames_pseudogene_details <- c("focal_pseudoN", "sis_pseudoN", "focal_protN", "sis_protN", "focal_pseudoFrac", "sis_pseudoFrac")
  pseudogene_details <- as.data.frame(matrix(NA, nrow = nrow(jumpClade_details), ncol = length(colNames_pseudogene_details), dimnames = list(paste0(OLclade_oi,"_jump",1:nrow(jumpClade_details)), colNames_pseudogene_details)))
  
  for(jumpNum_oi in which(jumpClade_details$genomes_downloaded==T)){#only for those jumps which are analyzable
    print(paste0("Counting pseudogenes for jump number ", jumpNum_oi))
    
    jumpDir <- paste0(jump_common_dir, OLclade_oi, "_jump", jumpNum_oi, "/")
    all_accNums <- read.table(paste0(jumpDir, "ass_fn_GCFids_found.txt"), stringsAsFactors = F)
    GTDB_accNums_map <- all_accNums$V1
    GTDB_labels <- read.table(paste0(jumpDir, "required_GCFonly_ids.txt"), stringsAsFactors = F)
    names(GTDB_accNums_map) <- GTDB_labels$V1
    
    all_cdsAccNums <- read.table(file = paste0(jumpDir,"cds_fn_GCFids_found.txt"),stringsAsFactors = F, header = F)
    all_cdsAccNums <- all_cdsAccNums$V1
    
    focal_GTDBLabels <- read.table(paste0(jumpDir, OLclade_oi,"_jump",jumpNum_oi,"_focalGTDBLabels.tab"), stringsAsFactors = F, header = F)
    focal_accNums <- sapply(focal_GTDBLabels$V1, function(x) GTDB_accNums_map[x], simplify = T, USE.NAMES = F)
    sis_GTDBLabels <- read.table(paste0(jumpDir,OLclade_oi,"_jump",jumpNum_oi,"_sisGTDBLabels.tab"), stringsAsFactors = F, header = F)
    sis_accNums <- sapply(sis_GTDBLabels$V1, function(x) GTDB_accNums_map[x], simplify = T, USE.NAMES = F)
    out1_GTDBLabels <- read.table(paste0(jumpDir,OLclade_oi,"_jump",jumpNum_oi,"_out1GTDBLabels.tab"), stringsAsFactors = F, header = F)
    out1_accNums <- sapply(out1_GTDBLabels$V1, function(x) GTDB_accNums_map[x], simplify = T, USE.NAMES = F)
       
    nested_focalTip_labels <- tryCatch({
      read.table(file = paste0(jumpDir, "nestedFocalTips.tab"), stringsAsFactors = F)$V1
    }, error=function(e){
      print(e)
      return(NA)
    })
    
    valid_focal_accNums <- setdiff(setdiff(focal_accNums, GTDB_accNums_map[nested_focalTip_labels]), NA)
    valid_sis_accNums <- setdiff(setdiff(sis_accNums, GTDB_accNums_map[nested_focalTip_labels]), NA)
    valid_out1_accNums <- setdiff(setdiff(out1_accNums, GTDB_accNums_map[nested_focalTip_labels]), NA)
    
    all_pseudogene_nums <- sapply(c(valid_focal_accNums, valid_sis_accNums, valid_out1_accNums), function(x) get_pseudogene_nums(x))
    all_pseudogene_nums <- t(all_pseudogene_nums)
    
    pseudogene_details[paste0(OLclade_oi,"_jump", jumpNum_oi),"focal_pseudoN"] <- median(as.numeric(all_pseudogene_nums[valid_focal_accNums,1]), na.rm = T)
    pseudogene_details[paste0(OLclade_oi,"_jump", jumpNum_oi),"sis_pseudoN"] <- median(as.numeric(all_pseudogene_nums[c(valid_sis_accNums),1]), na.rm = T)#,valid_out1_accNums
    pseudogene_details[paste0(OLclade_oi,"_jump", jumpNum_oi),"focal_protN"] <- median(as.numeric(all_pseudogene_nums[valid_focal_accNums,2]), na.rm = T)
    pseudogene_details[paste0(OLclade_oi,"_jump", jumpNum_oi),"sis_protN"] <- median(as.numeric(all_pseudogene_nums[c(valid_sis_accNums),2]), na.rm = T)#,valid_out1_accNums
    pseudogene_details[paste0(OLclade_oi,"_jump", jumpNum_oi),"focal_pseudoFrac"] <- median(as.numeric(all_pseudogene_nums[valid_focal_accNums,1])/as.numeric(all_pseudogene_nums[valid_focal_accNums,2]), na.rm = T)
    pseudogene_details[paste0(OLclade_oi,"_jump", jumpNum_oi),"sis_pseudoFrac"] <- median(as.numeric(all_pseudogene_nums[c(valid_sis_accNums),1])/as.numeric(all_pseudogene_nums[c(valid_sis_accNums),2]), na.rm = T) #,valid_out1_accNums
    write.table(all_pseudogene_nums, file = paste0(jumpDir, "pseudogene_numbers.tab"), quote = F)
  }
  pseudogene_details$focal_pseudoFrac <- round(pseudogene_details$focal_pseudoFrac,3)
  pseudogene_details$sis_pseudoFrac <- round(pseudogene_details$sis_pseudoFrac,3)
  write.table(pseudogene_details, file = paste0(jump_common_dir, OLclade_oi, "_pseudogene_details.tab"), quote = F)
}
```


#### Plot pseudogene comparisons
```{r collate_pseudogene_comparisons }
all_pseudogene_comparisons <- do.call(rbind, lapply(OLclades_oi, function(x) read.table(paste0(jump_common_dir,x,"_pseudogene_details.tab"), stringsAsFactors = F, header = T, row.names = 1)))
all_pseudogene_comparisons$pseudogeneRatio <- all_pseudogene_comparisons$focal_pseudoFrac/all_pseudogene_comparisons$sis_pseudoFrac
all_pseudogene_comparisons$pseudogeneDiff <- 100*(all_pseudogene_comparisons$focal_pseudoFrac - all_pseudogene_comparisons$sis_pseudoFrac)

write.table(all_pseudogene_comparisons, file = paste0(jump_common_dir, "levolution_jumps_pseudogene_comparison.tab"), quote = F)
```

```{r plot_pseudogene }
all_pseudogene_comparisons <- read.table(file = paste0(jump_common_dir,"levolution_jumps_pseudogene_comparison.tab"), stringsAsFactors = F)

pseudoFrac_focal_vs_sis_plt <- ggplot(data = all_pseudogene_comparisons, aes(x=sis_pseudoFrac, y=focal_pseudoFrac)) + geom_point(shape=21, color="darkblue") + scale_x_continuous(name = "related clade pseudogene fraction") + scale_y_continuous(name = "focal clade fraction") + geom_abline(slope = 1, intercept = 0, color="grey")

save_plot(filename = paste0(jump_common_dir,"pseudogeneFrac_comparison.svg"), pseudoFrac_focal_vs_sis_plt, ncol = 1, base_height = 3.2, base_width = 4)
pseudoFrac_focal_vs_sis_plt
```

### Genome properties analysis
Here, I will analyze the changes in GC content and genome size, coding density, and protein branch lengths.

#### Extract genome properties
```{r extract_genome_properties }
for(OLclade_n in 1:length(OLclades_oi)){
  OLclade_oi <- OLclades_oi[OLclade_n]
  print(paste0("Analysing jumps within ", OLclade_oi))
  pp_cutoff <- pp_cutoffs[OLclade_n]
  alpha_oi <- alphas_oi[OLclade_n]
  p_folder <- p_folders[OLclade_n]
  res_dir <- out_dir <-  paste0(levolution_out_dir ,p_folder,"_GTDB_14k_levolution/")

  tree_pp <- read.newick(file = paste0(res_dir,OLclade_oi,"_GTDB_14k_GC_config", config_oi, "_alpha", alpha_oi,".post"))
  tree_in <- read.newick(paste0(res_dir,OLclade_oi,"_GTDB_14k_tree.newick"))
  OLclade_nTips <- length(tree_in$tip.label) ##OL=OrderLevel
  OLclade_jumpBranches <- which(tree_pp$edge.length>pp_cutoff)
  
  jumpClade_details <- read.table(file = paste0(jump_common_dir, OLclade_oi, "_jumpSummaries4.tab"), stringsAsFactors = F)
  
  colNames_genome_details <- c("focal_GC", "sis_GC", "focal_gSize", "sis_gSize", "focal_checkM_gSize", "sis_checkM_gSize", "focal_cDensity", "sis_cDensity", "focal_bLength", "sis_bLength")
  genome_details <- as.data.frame(matrix(data = NA, nrow = nrow(jumpClade_details), ncol = length(colNames_genome_details), dimnames = list(paste0(OLclade_oi, "_jump", 1:nrow(jumpClade_details)), colNames_genome_details)))
  
  for(jumpNum_oi in which(jumpClade_details$depth %in% c("sister", "outgroup1", "outgroup2") & jumpClade_details$genomes_downloaded==T)){#only for those jumps which are analyzable
    print(paste0("Extracting genome information for jump number ", jumpNum_oi))
    
    jumpDir <- paste0(jump_common_dir, OLclade_oi, "_jump", jumpNum_oi, "/")
    
    all_accNums <- read.table(paste0(jumpDir, "ass_fn_GCFids_found.txt"), stringsAsFactors = F)
    GTDB_accNums_map <- all_accNums$V1
    GTDB_labels <- read.table(paste0(jumpDir, "required_GCFonly_ids.txt"), stringsAsFactors = F)
    names(GTDB_accNums_map) <- GTDB_labels$V1
    
    nested_focalTip_labels <- tryCatch({
      read.table(file = paste0(jumpDir, "nestedFocalTips.tab"), stringsAsFactors = F)$V1
    }, error=function(e){
      print(e)
      return(NA)
    })
    
    focal_GTDBLabels <- read.table(paste0(jumpDir, OLclade_oi,"_jump",jumpNum_oi,"_focalGTDBLabels.tab"), stringsAsFactors = F, header = F)
    focal_GTDBLabels_valid <- setdiff(focal_GTDBLabels$V1, nested_focalTip_labels)
    focal_accNums <- sapply(focal_GTDBLabels_valid, function(x) GTDB_accNums_map[x], simplify = T, USE.NAMES = F)
    
    sis_GTDBLabels <- read.table(paste0(jumpDir,OLclade_oi,"_jump",jumpNum_oi,"_sisGTDBLabels.tab"), stringsAsFactors = F, header = F)
    sis_GTDBLabels_valid <- setdiff(sis_GTDBLabels$V1, nested_focalTip_labels)
    sis_accNums <- sapply(sis_GTDBLabels_valid, function(x) GTDB_accNums_map[x], simplify = T, USE.NAMES = F)
    
    out1_accNums <- NA
    if(jumpClade_details[paste0("jump",jumpNum_oi), "depth"] %in% c("outgroup1", "outgroup2")){
      out1_GTDBLabels <- read.table(paste0(jumpDir,OLclade_oi,"_jump",jumpNum_oi,"_out1GTDBLabels.tab"), stringsAsFactors = F, header = F)
      out1_GTDBLabels_valid <- setdiff(out1_GTDBLabels$V1, nested_focalTip_labels)
      out1_accNums <- sapply(out1_GTDBLabels_valid, function(x) GTDB_accNums_map[x], simplify = T, USE.NAMES = F)
    }
    
    out2_accNums <- NA
    if(jumpClade_details[paste0("jump",jumpNum_oi), "depth"]=="outgroup2"){
      out2_GTDBLabels <- read.table(paste0(jumpDir,OLclade_oi,"_jump",jumpNum_oi,"_out2GTDBLabels.tab"), stringsAsFactors = F, header = F)
      out2_GTDBLabels_valid <- setdiff(out2_GTDBLabels$V1, nested_focalTip_labels)
      out2_accNums <- sapply(out2_GTDBLabels_valid, function(x) GTDB_accNums_map[x], simplify = T, USE.NAMES = F)
    }
    
    focal_GC <- median(GTDB_14k_metadata[focal_GTDBLabels_valid,"gc_percentage"])
    sis_GC <- median(GTDB_14k_metadata[sis_GTDBLabels_valid,"gc_percentage"])
    focal_gSize <- median(GTDB_14k_metadata[focal_GTDBLabels_valid,"genome_size"])
    sis_gSize <- median(GTDB_14k_metadata[sis_GTDBLabels_valid,"genome_size"])
    focal_checkM_gSize <- median(100*GTDB_14k_metadata[focal_GTDBLabels_valid,"genome_size"]/GTDB_14k_metadata[focal_GTDBLabels_valid,"checkm_completeness"])
    sis_checkM_gSize <- median(100*GTDB_14k_metadata[sis_GTDBLabels_valid,"genome_size"]/GTDB_14k_metadata[sis_GTDBLabels_valid,"checkm_completeness"])
    
    focal_cDensity <- median(GTDB_14k_metadata[focal_GTDBLabels_valid,"coding_density"])
    sis_cDensity <- median(GTDB_14k_metadata[sis_GTDBLabels_valid,"coding_density"])
    
    ###branch lengths
    focal_clade_tree <- read.tree(file = paste0(jumpDir, OLclade_oi, "_jump", jumpNum_oi, ".newick"))
    anc_node <- getMRCA(phy = focal_clade_tree, tip = c(focal_GTDBLabels_valid, sis_GTDBLabels_valid))
    
    focal_clade_median_length <- median(sapply(intersect(focal_GTDBLabels_valid, focal_clade_tree$tip.label), function(x) nodeheight(focal_clade_tree,which(focal_clade_tree$tip.label==x))) - nodeheight(focal_clade_tree, anc_node))
    tryCatch({
      sister_clade_median_length <- median(sapply(intersect(sis_GTDBLabels_valid, focal_clade_tree$tip.label), function(x) nodeheight(focal_clade_tree, which(focal_clade_tree$tip.label==x))) - nodeheight(focal_clade_tree, anc_node))
    }, error=function(e){
      print(e)
      return(NA)
    })
    
    genome_details[paste0(OLclade_oi, "_jump", jumpNum_oi),] <- c(focal_GC, sis_GC, focal_gSize, sis_gSize,focal_checkM_gSize, sis_checkM_gSize, focal_cDensity, sis_cDensity, focal_clade_median_length, sister_clade_median_length)
  }##end of loop over jumps in a clade
  write.table(genome_details, file = paste0(jump_common_dir, OLclade_oi, "_genome_details.tab"), quote = F)
} ##end of loop over order level clades
```


#### Plotting of genome comparisons
```{r write_all_genome }
all_genome_comparisons <- do.call(rbind, lapply(OLclades_oi, function(x) read.table(paste0(jump_common_dir,x,"_genome_details.tab"), stringsAsFactors = F, header = T, row.names = 1)))
all_genome_comparisons$gSizeRatio <- as.numeric(all_genome_comparisons$focal_gSize)/as.numeric(all_genome_comparisons$sis_gSize)
all_genome_comparisons$checkM_gSizeRatio <- as.numeric(all_genome_comparisons$focal_checkM_gSize)/as.numeric(all_genome_comparisons$sis_checkM_gSize)
write.table(all_genome_comparisons, file = paste0(jump_common_dir, "levolution_jumps_genome_comparison.tab"), quote = F)
```

```{r plot_genome_comparison }
all_genome_comparisons <- read.table(file = paste0(jump_common_dir,"levolution_jumps_genome_comparison.tab"), stringsAsFactors = F)

gSize_GC_plt <- ggplot(data = all_genome_comparisons, aes(x=focal_GC - sis_GC, y=focal_gSize/sis_gSize)) + geom_point(shape=21, color="darkblue") + scale_x_continuous(name = expression(Delta*"GC (genomic)")) + scale_y_continuous(name = "change in genome size\n(focal/sister)", trans = "log2")# + geom_abline(slope = 1, intercept = 0, color="grey")

cDensity_GC_plt <- ggplot(data = all_genome_comparisons, aes(x=focal_GC - sis_GC, y=focal_cDensity/sis_cDensity)) + geom_point(shape=21, color="darkblue") + scale_x_continuous(name = expression(Delta*"GC (genomic)")) + scale_y_continuous(name = "change in coding density\n(focal/sister)", trans = "log2", breaks = c(0.5, 0.75,1,1.5,2), labels = c(0.5, 0.75,1,1.5,2))# + geom_abline(slope = 1, intercept = 0, color="grey")

bLength_GC_plt <- ggplot(data = all_genome_comparisons, aes(x=focal_GC - sis_GC, y=focal_bLength/sis_bLength)) + geom_point(shape=21, color="darkblue") + scale_x_continuous(name = expression(Delta*"GC (genomic)")) + scale_y_continuous(name = "change in branch lengths\n(focal/sister)", trans = "log2")# + geom_abline(slope = 1, intercept = 0, color="grey")

genome_plt <- plot_grid(plotlist = list(gSize_GC_plt, cDensity_GC_plt, bLength_GC_plt), align = "hv", nrow = 1, ncol = 3, labels = "AUTO", label_size = 20)
save_plot(filename = paste0(jump_common_dir,"genome_comparison.svg"), genome_plt, ncol = 3, base_height = 3.2, base_width = 4)
genome_plt
```

### Lifestyle properties analysis
Recently an integrated dataset of lifestyle properties has become available. It integrates data from various sources, but at the species level. Lets see if we can extract some features from it. I will first focus on habitat- soil vs aquatic, and metabolism- aerobic vs rest.

#### Extract lifestyle properties
```{r extract_lifestyle_properties }
taxmap_gtdb <- read.csv(paste0(input_dir, "trait_data/taxmap_gtdb.csv"), stringsAsFactors=F, row.names = 1)
trait_data_gtdb <- read.csv(paste0(input_dir, "trait_data/condensed_species_GTDB.csv"), stringsAsFactors=F, row.names = 1)
trait_data_ncbi <- read.csv(paste0(input_dir, "trait_data/condensed_species_NCBI.csv"), stringsAsFactors=F, row.names = 1)

for(OLclade_n in 1:length(OLclades_oi)){
  OLclade_oi <- OLclades_oi[OLclade_n]
  print(paste0("Analysing jumps within ", OLclade_oi))
  pp_cutoff <- pp_cutoffs[OLclade_n]
  alpha_oi <- alphas_oi[OLclade_n]
  p_folder <- p_folders[OLclade_n]
  res_dir <- out_dir <-  paste0(levolution_out_dir ,p_folder,"_GTDB_14k_levolution/")

  tree_pp <- read.newick(file = paste0(res_dir,OLclade_oi,"_GTDB_14k_GC_config", config_oi, "_alpha", alpha_oi,".post"))
  tree_in <- read.newick(paste0(res_dir,OLclade_oi,"_GTDB_14k_tree.newick"))
  OLclade_nTips <- length(tree_in$tip.label) ##OL=OrderLevel
  OLclade_jumpBranches <- which(tree_pp$edge.length>pp_cutoff)
  
  jumpClade_details <- read.table(file = paste0(jump_common_dir, OLclade_oi, "_jumpSummaries4.tab"), stringsAsFactors = F)
  
  colNames_lifestyle_details <- c("focal_habitat", "sis_habitat", "focal_metabolism", "sis_metabolism")
  lifestyle_details <- as.data.frame(matrix(data = NA, nrow = nrow(jumpClade_details), ncol = length(colNames_lifestyle_details), dimnames = list(paste0(OLclade_oi, "_jump", 1:nrow(jumpClade_details)), colNames_lifestyle_details)))
  
  for(jumpNum_oi in which(jumpClade_details$depth %in% c("sister", "outgroup1", "outgroup2") & jumpClade_details$genomes_downloaded==T)){#only for those jumps which are analyzable
    print(paste0("Extracting genome information for jump number ", jumpNum_oi))
    
    jumpDir <- paste0(jump_common_dir, OLclade_oi, "_jump", jumpNum_oi, "/")
    
    all_accNums <- read.table(paste0(jumpDir, "ass_fn_GCFids_found.txt"), stringsAsFactors = F)
    GTDB_accNums_map <- all_accNums$V1
    GTDB_labels <- read.table(paste0(jumpDir, "required_GCFonly_ids.txt"), stringsAsFactors = F)
    names(GTDB_accNums_map) <- GTDB_labels$V1
    
    nested_focalTip_labels <- tryCatch({
      read.table(file = paste0(jumpDir, "nestedFocalTips.tab"), stringsAsFactors = F)$V1
    }, error=function(e){
      print(e)
      return(NA)
    })
    
    focal_GTDBLabels <- read.table(paste0(jumpDir, OLclade_oi,"_jump",jumpNum_oi,"_focalGTDBLabels.tab"), stringsAsFactors = F, header = F)
    focal_GTDBLabels_valid <- setdiff(focal_GTDBLabels$V1, nested_focalTip_labels)
    focal_accNums <- sapply(focal_GTDBLabels_valid, function(x) GTDB_accNums_map[x], simplify = T, USE.NAMES = F)
    
    sis_GTDBLabels <- read.table(paste0(jumpDir,OLclade_oi,"_jump",jumpNum_oi,"_sisGTDBLabels.tab"), stringsAsFactors = F, header = F)
    sis_GTDBLabels_valid <- setdiff(sis_GTDBLabels$V1, nested_focalTip_labels)
    sis_accNums <- sapply(sis_GTDBLabels_valid, function(x) GTDB_accNums_map[x], simplify = T, USE.NAMES = F)
    
    out1_accNums <- NA
    if(jumpClade_details[paste0("jump",jumpNum_oi), "depth"] %in% c("outgroup1", "outgroup2")){
      out1_GTDBLabels <- read.table(paste0(jumpDir,OLclade_oi,"_jump",jumpNum_oi,"_out1GTDBLabels.tab"), stringsAsFactors = F, header = F)
      out1_GTDBLabels_valid <- setdiff(out1_GTDBLabels$V1, nested_focalTip_labels)
      out1_accNums <- sapply(out1_GTDBLabels_valid, function(x) GTDB_accNums_map[x], simplify = T, USE.NAMES = F)
    }
    
    out2_accNums <- NA
    if(jumpClade_details[paste0("jump",jumpNum_oi), "depth"]=="outgroup2"){
      out2_GTDBLabels <- read.table(paste0(jumpDir,OLclade_oi,"_jump",jumpNum_oi,"_out2GTDBLabels.tab"), stringsAsFactors = F, header = F)
      out2_GTDBLabels_valid <- setdiff(out2_GTDBLabels$V1, nested_focalTip_labels)
      out2_accNums <- sapply(out2_GTDBLabels_valid, function(x) GTDB_accNums_map[x], simplify = T, USE.NAMES = F)
    }
    
    focal_metabolisms <- paste0(sapply(sapply(focal_GTDBLabels_valid, function(x) taxmap_gtdb[x, "ncbi_species_taxid"]), function(y) trait_data_gtdb[as.character(y), "metabolism"]), collapse = "|")
    sis_metabolisms <- paste0(sapply(sapply(sis_GTDBLabels_valid, function(x) taxmap_gtdb[x, "ncbi_species_taxid"]), function(y) trait_data_gtdb[as.character(y), "metabolism"]),collapse = "|")
    focal_habitat <- paste0(sapply(sapply(focal_GTDBLabels_valid, function(x) taxmap_gtdb[x, "ncbi_species_taxid"]), function(y) strsplit(trait_data_gtdb[as.character(y), "isolation_source"], "_")[[1]][1]), collapse = "|")
    sis_habitat <- paste0(sapply(sapply(sis_GTDBLabels_valid, function(x) taxmap_gtdb[x, "ncbi_species_taxid"]), function(y) strsplit(trait_data_gtdb[as.character(y), "isolation_source"], "_")[[1]][1]),collapse = "|")
    
    lifestyle_details[paste0(OLclade_oi, "_jump", jumpNum_oi),] <- c(focal_habitat, sis_habitat, focal_metabolisms, sis_metabolisms)
  }##end of loop over jumps in a clade
  write.csv(lifestyle_details, file = paste0(jump_common_dir, OLclade_oi, "_lifestyle_details.csv"), quote = F)
} ##end of loop over order level clades
```

#### Collating of lifestyle comparisons
```{r write_all_lifestyle }
all_lifestyle_comparisons <- do.call(rbind, lapply(OLclades_oi, function(x) read.csv(paste0(jump_common_dir,x,"_lifestyle_details.csv"), stringsAsFactors = F, header = T, row.names = 1)))
write.csv(all_lifestyle_comparisons, file = paste0(jump_common_dir, "levolution_jumps_lifestyle_comparison.csv"), quote = F)
```

### Collate all data
```{r collate_summary_data }
all_summaries <- do.call(rbind, lapply(OLclades_oi, function(x) read.table(paste0(jump_common_dir,x,"_jumpSummaries4.tab"), stringsAsFactors = F, header = T, row.names = paste0(x,"_jump",1:nrow(read.table(paste0(jump_common_dir,x,"_jumpSummaries4.tab"))) ))))
write.table(all_summaries, file = paste0(jump_common_dir, "levolution_jumps_feature_summaries4.tab"), quote = F)
```


### Comparing 16s using raxml
```{r compare_16s_branch_lengths_raxml }
for(OLclade_n in 1:length(OLclades_oi)){
  OLclade_oi <- OLclades_oi[OLclade_n]
  print(paste0("Analysing jumps within ", OLclade_oi))

  jumpClade_details <- read.table(file = paste0(jump_common_dir, OLclade_oi, "_jumpSummaries4.tab"), stringsAsFactors = F)
  
  colNames_16s_details <- c("rna_focal_bLength", "rna_sis_bLength", "rna_focal_GC", "rna_sis_GC", "focal_GC", "sis_GC")
  SixteenS_details <- as.data.frame(matrix(data = NA, nrow = length(which(jumpClade_details$rna_analysis=="yes")), ncol = length(colNames_16s_details), dimnames = list(paste0(OLclade_oi, "_jump", which(jumpClade_details$rna_analysis=="yes")), colNames_16s_details)))

  if(!any(jumpClade_details$rna_analysis=="yes")){
    print("No jumps that can be analyzed for 16s branch lengths here")
    write(x = NULL, file = paste0(jump_common_dir, OLclade_oi,"_16s_details.tab"))
    next
  } 

  for(jumpNum_oi in which(jumpClade_details$rna_analysis=="yes")){#only for those jumps which are analyzable
    print(paste0("Comparing 16s evolution rate for jump number ", jumpNum_oi))
    
    jumpDir <- paste0(jump_common_dir, OLclade_oi, "_jump", jumpNum_oi, "/")
    
    all_accNums <- read.table(paste0(jumpDir, "ass_fn_GCFids_found.txt"), stringsAsFactors = F)
    GTDB_accNums_map <- all_accNums$V1
    GTDB_labels <- read.table(paste0(jumpDir, "required_GCFonly_ids.txt"), stringsAsFactors = F)
    names(GTDB_accNums_map) <- GTDB_labels$V1
    
    nested_focalTip_labels <- tryCatch({
      read.table(file = paste0(jumpDir, "nestedFocalTips.tab"), stringsAsFactors = F)$V1
    }, error=function(e){
      print(e)
      return(NA)
    })
    
    focal_GTDBLabels <- read.table(paste0(jumpDir, OLclade_oi,"_jump",jumpNum_oi,"_focalGTDBLabels.tab"), stringsAsFactors = F, header = F)
    focal_GTDBLabels_valid <- setdiff(focal_GTDBLabels$V1, nested_focalTip_labels)
    focal_accNums <- sapply(focal_GTDBLabels_valid, function(x) GTDB_accNums_map[x], simplify = T, USE.NAMES = F)
    
    sis_GTDBLabels <- read.table(paste0(jumpDir,OLclade_oi,"_jump",jumpNum_oi,"_sisGTDBLabels.tab"), stringsAsFactors = F, header = F)
    sis_GTDBLabels_valid <- setdiff(sis_GTDBLabels$V1, nested_focalTip_labels)
    sis_accNums <- sapply(sis_GTDBLabels_valid, function(x) GTDB_accNums_map[x], simplify = T, USE.NAMES = F)
    
    out1_GTDBLabels <- read.table(paste0(jumpDir,OLclade_oi,"_jump",jumpNum_oi,"_out1GTDBLabels.tab"), stringsAsFactors = F, header = F)
    out1_GTDBLabels_valid <- setdiff(out1_GTDBLabels$V1, nested_focalTip_labels)
    out1_accNums <- sapply(out1_GTDBLabels_valid, function(x) GTDB_accNums_map[x], simplify = T, USE.NAMES = F)
    
    out2_accNums <- NA
    if(jumpClade_details[paste0("jump",jumpNum_oi), "depth"]=="outgroup2"){
      out2_GTDBLabels <- read.table(paste0(jumpDir,OLclade_oi,"_jump",jumpNum_oi,"_out2GTDBLabels.tab"), stringsAsFactors = F, header = F)
      out2_GTDBLabels_valid <- setdiff(out2_GTDBLabels$V1, nested_focalTip_labels)
      out2_accNums <- sapply(out2_GTDBLabels_valid, function(x) GTDB_accNums_map[x], simplify = T, USE.NAMES = F)
    }
    
    SixteenS_fn <- paste0(jumpDir,"16s_seqs.fa")
    SixteenS_aln_fn <- paste0(jumpDir,"16s_seqs.aln")
    # system2(command = "muscle", args = c("-in", SixteenS_fn, "-out", SixteenS_aln_fn), stdout = F, stderr = F)
    
    SixteenS_aln <- read.alignment(SixteenS_aln_fn, format = "fasta")
    av_SixteenS <- SixteenS_aln$nam
    
    jumpClade_tree_GCF <- focal_clade_tree <- read.tree(paste0(jumpDir, OLclade_oi, "_jump", jumpNum_oi, ".newick"))
    jumpClade_tree_GCF$tip.label <- sapply(focal_clade_tree$tip.label, function(x) GTDB_accNums_map[x])
    jumpClade_subtree_GCF <- keep.tip(jumpClade_tree_GCF, av_SixteenS)
    if(length(jumpClade_subtree_GCF$tip.label)<4) {
      print("ONly 3 tips, cant optimize branch lengths")
      next
    }
    jumpClade_subtree_GCF_fn <- paste0(jumpDir,"16s_constraint.tre")
    write.tree(unroot(jumpClade_subtree_GCF), jumpClade_subtree_GCF_fn)
    system2(command = "raxml-ng", args = c("--redo", "--evaluate", "--msa", SixteenS_aln_fn, "--tree", jumpClade_subtree_GCF_fn, "--model", "GTR+G+I", "--brlen", "unlinked"), stdout = F, stderr = F)

    SixteenS_raxtree <- read.tree(paste0(jumpDir, "16s_seqs.aln.raxml.bestTree"))
    # plot(jumpClade_subtree_GCF)
    if(any(out2_accNums %in% SixteenS_raxtree$tip.label)){
      SixteenS_raxtree_rooted <- root(phy = SixteenS_raxtree, outgroup = intersect(out2_accNums, SixteenS_raxtree$tip.label), resolve.root = T)
    }else if(any(out1_accNums %in% SixteenS_raxtree$tip.label)){
      SixteenS_raxtree_rooted <- root(phy = SixteenS_raxtree, outgroup = intersect(out1_accNums, SixteenS_raxtree$tip.label), resolve.root = T)
    }else{
      print("No outgroup found. This is strange!")
    }
    rna_GC <- sapply(SixteenS_raxtree_rooted$tip.label, function(x) GC(s2c(SixteenS_aln$seq[which(SixteenS_aln$nam==x)][[1]])))
    
    rna_focal_accNums <- intersect(focal_accNums, SixteenS_raxtree_rooted$tip.label)
    rna_sis_accNums <- intersect(sis_accNums, SixteenS_raxtree_rooted$tip.label)
    
    bar_colors <- rep("grey", length(SixteenS_raxtree_rooted$tip.label))
    bar_colors[which(SixteenS_raxtree_rooted$tip.label %in% rna_focal_accNums)] <- "orange"
    bar_colors[which(SixteenS_raxtree_rooted$tip.label %in% rna_sis_accNums)] <- "purple"
    names(bar_colors) <- SixteenS_raxtree_rooted$tip.label
    
    svg(filename = paste0(jumpDir, "16s_raxTree.svg"), width = 15)
    SixteenS_plt <- plotTree.barplot(tree = SixteenS_raxtree_rooted, x = rna_GC, args.barplot = list(xlab="GC % (16S)"))
    dev.off()
    
    SixteenS_raxtree_rooted_labeled <- SixteenS_raxtree_rooted
    SixteenS_raxtree_rooted_labeled$tip.label <- sapply(SixteenS_raxtree_rooted$tip.label, function(x) get_genera_sp_names(full_GTDB_taxonomy = GTDB_14k_metadata[names(GTDB_accNums_map)[which(GTDB_accNums_map==x)],"gtdb_taxonomy"]))
    svg(filename = paste0(jumpDir, "16s_raxTree_colored.svg"), width = 15)
    SixteenS_plt <- plot(SixteenS_raxtree_rooted_labeled,tip.color = bar_colors)
    add.scale.bar()
    dev.off()
    
    rna_anc_node <- getMRCA(phy = SixteenS_raxtree_rooted, tip = c(rna_focal_accNums, rna_sis_accNums))
    
    rna_focal_bLength <- median(sapply(rna_focal_accNums, function(x) nodeheight(SixteenS_raxtree_rooted,which(SixteenS_raxtree_rooted$tip.label==x)))) - nodeheight(tree = SixteenS_raxtree_rooted, node =rna_anc_node)
    rna_sis_bLength <- median(sapply(rna_sis_accNums, function(x) nodeheight(SixteenS_raxtree_rooted,which(SixteenS_raxtree_rooted$tip.label==x)))) - nodeheight(tree = SixteenS_raxtree_rooted, node = rna_anc_node)
    rna_focal_GC <- median(rna_GC[rna_focal_accNums])
    rna_sis_GC <- median(rna_GC[rna_sis_accNums])
    
    GTDB_GC <- GTDB_14k_metadata[names(GTDB_accNums_map)[GTDB_accNums_map %in% SixteenS_raxtree_rooted$tip.label],"gc_percentage"]
    names(GTDB_GC) <- GTDB_accNums_map[GTDB_accNums_map %in% SixteenS_raxtree_rooted$tip.label]
    focal_gGC <- median(GTDB_GC[rna_focal_accNums])
    sis_gGC <- median(GTDB_GC[rna_sis_accNums])
    
    SixteenS_details[paste0(OLclade_oi, "_jump",jumpNum_oi), "rna_focal_bLength"] <- rna_focal_bLength
    SixteenS_details[paste0(OLclade_oi, "_jump",jumpNum_oi), "rna_sis_bLength"] <- rna_sis_bLength
    SixteenS_details[paste0(OLclade_oi, "_jump",jumpNum_oi), "rna_focal_GC"] <- rna_focal_GC
    SixteenS_details[paste0(OLclade_oi, "_jump",jumpNum_oi), "rna_sis_GC"] <- rna_sis_GC
    SixteenS_details[paste0(OLclade_oi, "_jump",jumpNum_oi), "focal_GC"] <- focal_gGC
    SixteenS_details[paste0(OLclade_oi, "_jump",jumpNum_oi), "sis_GC"] <- sis_gGC
    
  }##end of loop over jumps within a clade
  write.table(SixteenS_details, file = paste0(jump_common_dir, OLclade_oi, "_16s_raxml_details.tab"), quote = F)
}##end of loop over order-lvel clades

```

```{r collate_raxml_16s_comparisons }
all_raxml_16s_comparisons <- do.call(rbind, lapply(OLclades_oi, function(x) read.table(paste0(jump_common_dir,x,"_16s_raxml_details.tab"), stringsAsFactors = F, header = T, row.names = 1)))
all_raxml_16s_comparisons$SixteenSratio <- all_raxml_16s_comparisons$rna_focal_bLength/all_raxml_16s_comparisons$rna_sis_bLength
write.table(all_raxml_16s_comparisons, file = paste0(jump_common_dir, "levolution_jumps_raxml_16s_comparison.tab"), quote = F)
```

```{r plot_raxml_16s_bLengths }
all_raxml_16s_comparisons <- read.table(file = paste0(jump_common_dir,"levolution_jumps_raxml_16s_comparison.tab"), stringsAsFactors = F)
plt1 <- ggplot(data = all_raxml_16s_comparisons, aes(x=rna_sis_bLength, y=rna_focal_bLength)) + geom_point(col="darkblue", shape=21) + scale_x_continuous(name = "sister branch length (16S)") + scale_y_continuous(name = "focal branch length (16s)") + geom_abline(slope = 1, intercept = 0, col="grey")

plt2 <- ggplot(data = all_raxml_16s_comparisons, aes(x=focal_GC - sis_GC, y=100*(rna_focal_GC -  rna_sis_GC))) + geom_point(col="darkblue", shape=21) + geom_smooth(method = "lm", col="grey") + scale_x_continuous(name = expression(Delta*"GC (genomic)")) + scale_y_continuous(name = expression(Delta*"GC (16s)"))

SixteenSratio_rnaGC_plt <- ggplot(data = all_raxml_16s_comparisons, aes(x=focal_GC - sis_GC, y=rna_focal_bLength/rna_sis_bLength)) + geom_point(col="darkblue", shape=21) + scale_x_continuous(name = expression(Delta*"GC (genomic)")) + scale_y_continuous(name = "focal 16s branch length/\n sister 16s branch length", trans = "log2", limits = c(1/(2^5),2^5), breaks = c(1/(2^4), 1/(2^2), 1/(2^0), 2^2, 2^4)) + geom_hline(yintercept = 1, col="grey")

SixteenSratio_GC_plt <- ggplot(data = all_raxml_16s_comparisons, aes(x=rna_focal_GC - rna_sis_GC, y=rna_focal_bLength/rna_sis_bLength)) + geom_point(col="darkblue", shape=21) + scale_x_continuous(name = expression(Delta*"GC (16s)")) + scale_y_continuous(name = "focal 16s branch length/\n sister 16s branch length", trans = "log2", limits = c(1/(2^5),2^5), breaks = c(1/(2^4), 1/(2^2), 1/(2^0), 2^2, 2^4)) + geom_hline(yintercept = 1, col="grey")

bLength_plt <- plot_grid(plotlist = list(plt1, plt2, SixteenSratio_rnaGC_plt, SixteenSratio_GC_plt), align = "hv", nrow = 2, ncol = 2, labels = "AUTO", label_size = 20)
save_plot(filename = paste0(jump_common_dir,"raxml_16s_comparison.svg"), bLength_plt, ncol = 2, base_height = 7, base_width = 4.5)
bLength_plt
```

### Analysis for thesis

#### Collate and write data
```{r write_all_features_data }
### Average GC
all_genome_comparisons <- read.table(file = paste0(jump_common_dir,"levolution_jumps_genome_comparison.tab"), stringsAsFactors = F)
all_features_df <- round(all_genome_comparisons[,c("focal_GC", "sis_GC")], 2)

all_genome_comparisons2 <- read.table(file = paste0(jump_common_dir,"levolution_jumps_genome_extraComparisons.tab"), stringsAsFactors = F)
for(row_oi in rownames(all_features_df)){
  if(is.na(all_features_df[row_oi, "focal_GC"])){
    all_features_df[row_oi,] <- round(all_genome_comparisons2[row_oi,c("focal_GC", "sis_GC")], 2)
  }
}
all_features_df$deltaGC <- all_features_df$focal_GC - all_features_df$sis_GC
all_features_df$signGC <- as.factor(sign(all_features_df$deltaGC))

### Average dNdS from YN00
HEG_mean_dNdS <- read.table(file = paste0(jump_common_dir,"levolution_jumps_yn00_dS_corrected_comparison2.tab"), stringsAsFactors = F)
HEG_mean_dNdS2 <- read.table(file = paste0(jump_common_dir,"levolution_jumps_yn00_dS_corrected_extraComparison2.tab"), stringsAsFactors = F)

all_features_df$median_dNdS_related <- all_features_df$median_dNdS_focal <- all_features_df$median_dS_related <- all_features_df$median_dS_focal <- all_features_df$median_dNdS_ratios <- all_features_df$median_focal_GC3 <- all_features_df$median_related_GC3 <- NA
for(row_oi in rownames(HEG_mean_dNdS)){
  # print(row_oi)
  all_features_df[row_oi,c("median_dNdS_focal", "median_dNdS_related","median_dS_focal", "median_dS_related","median_dNdS_ratios", "median_focal_GC3", "median_related_GC3")] <- round(HEG_mean_dNdS[row_oi, c("valid_median_dNdS_focal", "valid_median_dNdS_related","valid_median_dS_focal", "valid_median_dS_related", "valid_median_dNdS_ratios", "valid_median_gc3_focal", "valid_median_gc3_related")], 3)
}

for(row_oi in intersect(rownames(HEG_mean_dNdS2), rownames(all_features_df))){
  # print(row_oi)
  all_features_df[row_oi,c("median_dNdS_focal", "median_dNdS_related","median_dS_focal", "median_dS_related","median_dNdS_ratios", "median_focal_GC3", "median_related_GC3")] <- round(HEG_mean_dNdS2[row_oi, c("valid_median_dNdS_focal_extra", "valid_median_dNdS_related_extra","valid_median_dS_focal_extra", "valid_median_dS_related_extra", "valid_median_dNdS_ratios_extra", "valid_median_gc3_focal_extra", "valid_median_gc3_related_extra")], 3)
}

all_features_df$deltaGC3 <- all_features_df$`median_focal_GC3` - all_features_df$`median_related_GC3`

### 16s branch length
all_features_df$rna_sis_bLength <- all_features_df$rna_focal_bLength <- NA

all_16s_comparisons <- read.table(file = paste0(jump_common_dir,"levolution_jumps_16s_comparison.tab"), stringsAsFactors = F)
all_16s_comparisons2 <- read.table(file = paste0(jump_common_dir,"levolution_jumps_16s_extraComparisons.tab"), stringsAsFactors = F)
all_features_df[rownames(all_16s_comparisons),c("rna_focal_bLength", "rna_sis_bLength")] <- round(all_16s_comparisons[, c("rna_focal_bLength", "rna_sis_bLength")],3)
for(row_oi in rownames(all_features_df)){
  if(is.na(all_features_df[row_oi, "rna_focal_bLength"])){
    all_features_df[row_oi,c("rna_focal_bLength", "rna_sis_bLength")] <- round(all_16s_comparisons2[row_oi,c("rna_focal_bLength", "rna_sis_bLength")], 2)
  }
}
all_features_df$rna_focal_bLength[which(is.infinite(all_features_df$rna_focal_bLength))] <- NA
all_features_df$rna_sis_bLength[which(is.infinite(all_features_df$rna_sis_bLength))] <- NA

##to prevent problems in taking ratios and log later, I will set negative CUB to zero, this simply indicates no CUB or reversal of CUB which can happen due to noise in codon counts
all_features_df$rna_focal_bLength[which(all_features_df$rna_focal_bLength<=0)] <- min(all_features_df$rna_focal_bLength[which(all_features_df$rna_focal_bLength>0)],na.rm = T)
all_features_df$rna_sis_bLength[which(all_features_df$rna_sis_bLength<=0)] <- min(all_features_df$rna_sis_bLength[which(all_features_df$rna_sis_bLength>0)],na.rm = T)

### CUB
all_features_df$sis_CUB <- all_features_df$focal_CUB <- NA

all_CUB_comparisons <- read.table(file = paste0(jump_common_dir,"levolution_jumps_CUB_comparison.tab"), stringsAsFactors = F)
all_features_df[rownames(all_CUB_comparisons), c("focal_CUB", "sis_CUB")] <- round(all_CUB_comparisons[, c("focal_CUB", "sis_CUB")], 2)
all_CUB_comparisons2 <- read.table(file = paste0(jump_common_dir,"levolution_extraJumps_CUB_comparison.tab"), stringsAsFactors = F)
for(row_oi in rownames(all_features_df)){
  if(is.na(all_features_df[row_oi, "focal_CUB"])){
    all_features_df[row_oi,c("focal_CUB", "sis_CUB")] <- round(all_CUB_comparisons2[row_oi,c("focal_CUB", "sis_CUB")], 2)
  }
}
all_features_df$focal_CUB[which(is.infinite(all_features_df$focal_CUB))] <- NA
all_features_df$sis_CUB[which(is.infinite(all_features_df$sis_CUB))] <- NA

##to prevent problems in taking ratios and log later, I will set negative CUB to zero, this simply indicates no CUB or reversal of CUB which can happen due to noise in codon counts
all_features_df$focal_CUB[which(all_features_df$focal_CUB<=0)] <- min(all_features_df$focal_CUB[which(all_features_df$focal_CUB>0)],na.rm = T)
all_features_df$sis_CUB[which(all_features_df$sis_CUB<=0)] <- min(all_features_df$sis_CUB[which(all_features_df$sis_CUB>0)],na.rm = T)

### Genome size
all_features_df$sis_gSize <- all_features_df$focal_gSize <-NA
all_features_df[rownames(all_genome_comparisons), c("focal_gSize", "sis_gSize")] <- round(all_genome_comparisons[, c("focal_gSize", "sis_gSize")]/(10^6), 2)

for(row_oi in rownames(all_features_df)){
  if(is.na(all_features_df[row_oi, "focal_gSize"])){
    all_features_df[row_oi,c("focal_gSize", "sis_gSize")] <- round(all_genome_comparisons2[row_oi,c("focal_gSize", "sis_gSize")]/(10^6), 2)
  }
}
all_features_df$gSizeRatio <- round(all_features_df$focal_gSize/all_features_df$sis_gSize, 2)

all_features_df$sis_checkM_gSize <- all_features_df$focal_checkM_gSize <-NA
all_features_df[rownames(all_genome_comparisons), c("focal_checkM_gSize", "sis_checkM_gSize")] <- round(all_genome_comparisons[, c("focal_checkM_gSize", "sis_checkM_gSize")]/(10^6), 2)

for(row_oi in rownames(all_features_df)){
  if(is.na(all_features_df[row_oi, "focal_checkM_gSize"])){
    all_features_df[row_oi,c("focal_checkM_gSize", "sis_checkM_gSize")] <- round(all_genome_comparisons2[row_oi,c("focal_checkM_gSize", "sis_checkM_gSize")]/(10^6), 2)
  }
}
all_features_df$checkM_gSizeRatio <- round(all_features_df$focal_checkM_gSize/all_features_df$sis_checkM_gSize, 2)

all_features_df$sis_cDensity <- all_features_df$focal_cDensity <-  NA
all_features_df[rownames(all_genome_comparisons), c("focal_cDensity", "sis_cDensity")] <- round(all_genome_comparisons[, c("focal_cDensity", "sis_cDensity")], 0)
for(row_oi in rownames(all_features_df)){
  if(is.na(all_features_df[row_oi, "focal_cDensity"])){
    all_features_df[row_oi,c("focal_cDensity", "sis_cDensity")] <- round(all_genome_comparisons2[row_oi,c("focal_cDensity", "sis_cDensity")], 2)
  }
}

all_features_df$sis_pseudoFrac <- all_features_df$focal_pseudoFrac <-  NA
all_pseudogene_comparisons <- read.table(file = paste0(jump_common_dir,"levolution_jumps_pseudogene_comparison.tab"), stringsAsFactors = F)
all_features_df[rownames(all_genome_comparisons), c("focal_pseudoFrac", "sis_pseudoFrac")] <- round(100*all_pseudogene_comparisons[, c("focal_pseudoFrac", "sis_pseudoFrac")], 2)
all_pseudogene_comparisons2 <- read.table(file = paste0(jump_common_dir,"levolution_jumps_pseudogene_extraComparisons.tab"), stringsAsFactors = F)
for(row_oi in rownames(all_features_df)){
  if(is.na(all_features_df[row_oi, "focal_pseudoFrac"])){
    all_features_df[row_oi,c("focal_pseudoFrac", "sis_pseudoFrac")] <- round(100*all_pseudogene_comparisons2[row_oi,c("focal_pseudoFrac", "sis_pseudoFrac")], 2)
  }
}

###IGR length
all_features_df$sis_igrLength <- all_features_df$focal_igrLength <-  NA
all_IGR_comparisons <- read.table(file = paste0(jump_common_dir,"levolution_jumps_IGR_comparison.tab"), stringsAsFactors = F)
all_features_df[rownames(all_IGR_comparisons), c("focal_igrLength", "sis_igrLength")] <- round(all_IGR_comparisons[, c("focal_igrLength", "sis_igrLength")], 0)
all_IGR_comparisons2 <- read.table(file = paste0(jump_common_dir,"levolution_jumps_IGR_extraComparisons.tab"), stringsAsFactors = F)
for(row_oi in rownames(all_features_df)){
  if(is.na(all_features_df[row_oi, "focal_igrLength"])){
    all_features_df[row_oi,c("focal_igrLength", "sis_igrLength")] <- round(all_IGR_comparisons2[row_oi,c("focal_igrLength", "sis_igrLength")], 0)
  }
}

###habitat and metabolism
all_features_df$sis_Isolation <- all_features_df$focal_Isolation <- all_features_df$sis_Metabolism <- all_features_df$focal_Metabolism <-  NA
all_lifestyle_comparisons <- read.csv(file = paste0(jump_common_dir,"levolution_jumps_lifestyle_comparison.csv"), stringsAsFactors = F, row.names = 1)
all_features_df[, c("focal_Isolation", "sis_Isolation", "focal_Metabolism", "sis_Metabolism")] <- all_lifestyle_comparisons[, c("focal_habitat", "sis_habitat", "focal_metabolism", "sis_metabolism")]

isolation_not_covered <- sapply(rownames(all_features_df), function(x) all(is.na(all_features_df[x, c("focal_Isolation", "sis_Isolation")])) )
metabolism_not_covered <- sapply(rownames(all_features_df), function(x) all(is.na(all_features_df[x, c("focal_Metabolism", "sis_Metabolism")])) )

all_lifestyle_comparisons2 <- read.csv(file = paste0(jump_common_dir,"levolution_jumps_lifestyle_extraComparisons.csv"), stringsAsFactors = F, row.names = 1)
for(row_oi in rownames(all_features_df)[isolation_not_covered]){
  all_features_df[row_oi, c("focal_Isolation", "sis_Isolation")] <- all_lifestyle_comparisons2[row_oi, c("focal_habitat", "sis_habitat")]
}
for(row_oi in rownames(all_features_df)[metabolism_not_covered]){
  all_features_df[row_oi, c("focal_Metabolism", "sis_Metabolism")] <- all_lifestyle_comparisons2[row_oi, c("focal_metabolism", "sis_metabolism")]
}

endosymbiont_status <- read.csv(paste0(input_dir, "all_jumps_endosymbionts_status.csv"), row.names = 1, stringsAsFactors = F)
all_features_df$endoSymb_status <- sapply(rownames(all_features_df), function(ds) ifelse(endosymbiont_status[ds, "status"]=="Endosymbiont" | endosymbiont_status[ds, "status"]=="Obligate pathogen", 1, 0))
all_features_df$endoSymb_status <- factor(all_features_df$endoSymb_status)

colnames(all_features_df) <- c("focal_GC", "sister_GC", "delta_GC", "directionGC", "related_GC3", "focal_GC3","change_in_dNdS", "focal_dS", "related_dS", "focal_dNdS", "related_dNdS", "deltaGC3","focal_branch_length", "sister_branch_length", "focal_CUB", "sister_CUB", "focal_genome_size", "sister_genome_size", "genome_size_ratio", "focal_genome_size_checkM", "sister_genome_size_checkM", "genome_size_ratio_checkM", "focal_coding_density", "sister_coding_density", "focal_pseudogenes", "sister_pseudogenes", "focal_IGR_length", "sister_IGR_length", "focal_metabolism", "sister_metabolism", "focal_isolation", "sister_isolation", "endoSymb_status")

write.table(all_features_df, file = paste0(jump_common_dir, "all_features_comparison_table.tab"))
View(all_features_df)

all_features_df_sel <- all_features_df[setdiff(rownames(all_features_df), symbiont_clades),]
all_features_df_sel <- all_features_df_sel[order(all_features_df_sel$`focal_GC3` - all_features_df_sel$`related_GC3`),]
write.table(all_features_df_sel, file = paste0(jump_common_dir, "all_features_comparison_table_ordered.tab"))
View(all_features_df_sel)
```

Collate and write data for genomic changes
```{r create_df_for_changes }
all_features_df <- read.table(file = paste0(jump_common_dir, "all_features_comparison_table.tab"), stringsAsFactors = F)

invalid_datasets <- c("Flavobacteriale_jump23", "Flavobacteriale_jump24", "Enterobacterales_jump15")
all_features_df_valid <- all_features_df[setdiff(rownames(all_features_df), invalid_datasets),]

all_feature_changes<- cbind.data.frame(all_features_df_valid[,"delta_GC"], all_features_df_valid[,"focal_CUB"]/all_features_df_valid[,"sister_CUB"],all_features_df_valid[,"focal_branch_length"]/all_features_df_valid[,"sister_branch_length"],all_features_df_valid[,"genome_size_ratio_checkM"], all_features_df_valid[,"focal_IGR_length"]-all_features_df_valid[,"sister_IGR_length"],round((all_features_df_valid[,"focal_IGR_length"]-all_features_df_valid[,"sister_IGR_length"])*100/all_features_df_valid[,"sister_IGR_length"],0), all_features_df_valid[,"focal_pseudogenes"]/all_features_df_valid[,"sister_pseudogenes"], all_features_df_valid[,"focal_coding_density"]-all_features_df_valid[,"sister_coding_density"], all_features_df_valid$directionGC, all_features_df_valid$endoSymb_status)

colnames(all_feature_changes)  <- colNames_changes <- c("GC", "CUB", "rna_bLength", "gSize", "IGRlength","IGRpercentage", "pseudoFrac", "cDensity", "directionGC", "endoSymb_status")
rownames(all_feature_changes) <- rownames(all_features_df_valid)

all_feature_changes$IGRpercentage[which(all_feature_changes$IGRlength>200)] <- max(all_feature_changes$IGRpercentage[which(all_feature_changes$IGRlength<200)])
all_feature_changes$IGRlength[which(all_feature_changes$IGRlength>200)] <- max(all_feature_changes$IGRlength[which(all_feature_changes$IGRlength<200)])
all_feature_changes$CUB_log2 <- log2(all_feature_changes$CUB)
all_feature_changes$rna_bLength_log2 <- log2(all_feature_changes$rna_bLength)
all_feature_changes$gSize_log2 <- log2(all_feature_changes$gSize)
all_feature_changes$pseudoFrac_log2 <- log2(all_feature_changes$pseudoFrac)

write.table(all_feature_changes, file = paste0(jump_common_dir, "all_features_changes_table.tab"))
```

#### Read back all files
So that we dont have to collate them every time for subsequent analysis.
```{r read_meta_files }
all_features_df <- read.table(file = paste0(jump_common_dir, "all_features_comparison_table.tab"), stringsAsFactors = F)
all_feature_changes <- read.table(file = paste0(jump_common_dir, "all_features_changes_table.tab"), stringsAsFactors = F)
all_features_df_sel <- read.table(file = paste0(jump_common_dir, "all_features_comparison_table_ordered.tab"), stringsAsFactors = F, check.names = F)

invalid_datasets <- c("Flavobacteriale_jump23", "Flavobacteriale_jump24", "Enterobacterales_jump15")
all_features_df_valid <- all_features_df[setdiff(rownames(all_features_df), invalid_datasets),]
```

#### Plot all variables against GC
```{r plot_all_vs_GC }
p_corr = function(corr_oi){ ##a function that returns linear equation with r-squared values
  eq <- substitute(italic(r)~"="~r2,l)    
  as.character(as.expression(eq));                 
}

col_labels <- c(paste0("\U0394", "GC3 (in % points)"), "fold change in CUB", "fold change in\t\n16s branch lengths", "fold change in\ngenome size", paste0("\U0394", "(IGR length)"), "fold change in\t\npseudogene fraction", paste0("\U0394", "(coding density)"))
names(col_labels) <- colnames(all_feature_changes)[1:7]

all_feature_changes$endoSymb_status <- factor(all_feature_changes$endoSymb_status)
all_feature_changes$directionGC <- factor(all_feature_changes$directionGC)

gSize_vs_GC <- ggplot(data = all_feature_changes, mapping = aes(x = GC, y = gSize, color=directionGC, shape=endoSymb_status)) + geom_point() + theme_cowplot()
gSize_vs_GC <- gSize_vs_GC + geom_hline(yintercept = 1, linetype="dashed", size=1.05, color="grey")
gSize_vs_GC <- gSize_vs_GC + scale_shape_manual(values = c(21,19)) + scale_x_continuous(breaks = seq(-40,20,10), labels = c("-40", "-30", "-20", "-10", "0", "10", "20"), limits=c(-40,25), name = col_labels["GC"]) + scale_y_continuous(name = "fold change in\ngenome size", limits = c(0.15, 3.5), trans = "log2") + scale_color_manual(values = cbPalette[c(7,6)])
gSize_vs_GC <- gSize_vs_GC + rremove("legend")
gSize_vs_GC <- gSize_vs_GC + stat_cor(data = filter(all_feature_changes, endoSymb_status==0), method = "spearman", mapping = aes(color = directionGC), label.y.npc = 1, label.x.npc = 0, cor.coef.name = "rho")
gSize_vs_GC

gSize_vs_jumps <- ggplot(data = all_feature_changes[!is.na(all_feature_changes$directionGC),], mapping = aes(x = directionGC, y = gSize, color=directionGC, fill=endoSymb_status))
gSize_vs_jumps <- gSize_vs_jumps + geom_dotplot(dotsize=0.5, mapping = aes(fill=endoSymb_status), binaxis='y', stackdir='center') + theme_cowplot()
gSize_vs_jumps <- gSize_vs_jumps + scale_x_discrete(name="", labels = c("downward", "upward")) + scale_y_continuous(name = "fold change in\ngenome size", trans="log2", limits = c(0.15, 3.5)) + scale_color_manual(values = cbPalette[c(7,6)], guide=F) + scale_fill_manual(values = c(NA, cbPalette[7]), guide=F)
gSize_vs_jumps <- gSize_vs_jumps + geom_hline(yintercept = 1, linetype="dashed", size=1.05, color="darkgrey")
gSize_vs_jumps <- gSize_vs_jumps + stat_summary(fun.data="median_hilow", fun.args = list(conf.int=0.5), geom="crossbar", size=0.5, inherit.aes = F, mapping = aes(x = directionGC, y = gSize, color=directionGC))
gSize_vs_jumps <- gSize_vs_jumps

gSize_wilcox_down <- wilcox.test(x = all_feature_changes$gSize[all_feature_changes$endoSymb_status==0 & all_feature_changes$directionGC==-1], mu = 1, conf.int=T)
gSize_wilcox_up <- wilcox.test(x = all_feature_changes$gSize[all_feature_changes$endoSymb_status==0 & all_feature_changes$directionGC==1], mu = 1, conf.int=T)
gSize_vs_jumps <- gSize_vs_jumps + annotate("text", x = 1, y = 3, label=paste("p = ", format(gSize_wilcox_down$p.value, digits = 3, scientific = ifelse(gSize_wilcox_down$p.value<0.001,T,F))), color=cbPalette[7])
gSize_vs_jumps <- gSize_vs_jumps + annotate("text", x = 2, y = 3, label=paste("p = ", format(gSize_wilcox_up$p.value, digits = 3, scientific = ifelse(gSize_wilcox_up$p.value<0.001,T,F))), color=cbPalette[6])
gSize_vs_jumps

CUB_vs_GC <- ggplot(data = filter(all_feature_changes, is.finite(CUB_log2)), mapping = aes(x = GC, y = CUB, color=directionGC, shape=endoSymb_status)) + geom_point() + theme_cowplot()
CUB_vs_GC <- CUB_vs_GC + scale_shape_manual(values = c(21,19)) + scale_x_continuous(name = col_labels["GC"], breaks = seq(-40,20,10), labels = c("-40", "-30", "-20", "-10", "0", "10", "20"), limits=c(-40,25)) + scale_y_continuous(name = "fold change in CUB", trans = "log2", limits = c(1/32, 8)) + scale_color_manual(values = c(cbPalette[c(7,6)]))
CUB_vs_GC <- CUB_vs_GC + geom_hline(yintercept = 1, linetype="dashed", size=1.05, color="grey")
CUB_vs_GC <- CUB_vs_GC + rremove("legend")
CUB_vs_GC <- CUB_vs_GC + stat_cor(data = filter(all_feature_changes, endoSymb_status==0), method = "spearman", mapping = aes(color = directionGC), label.y.npc = 1, label.x.npc = 0, cor.coef.name = "rho")
CUB_vs_GC

CUB_vs_jumps <- ggplot(data = all_feature_changes[!is.na(all_feature_changes$directionGC),], mapping = aes(x = directionGC, y = CUB, color=directionGC, fill=endoSymb_status))
CUB_vs_jumps <- CUB_vs_jumps + geom_dotplot(dotsize=0.5, mapping = aes(fill=endoSymb_status), binaxis='y', stackdir='center') + theme_cowplot()
CUB_vs_jumps <- CUB_vs_jumps + scale_x_discrete(name="", labels = c("downward", "upward")) + scale_y_continuous(name = "fold change in CUB", trans="log2", limits = c(1/32, 8)) + scale_color_manual(values = cbPalette[c(7,6)], guide=F) + scale_fill_manual(values = c(NA, cbPalette[7]), guide=F)
CUB_vs_jumps <- CUB_vs_jumps + geom_hline(yintercept = 1, linetype="dashed", size=1.05, color="darkgrey")
CUB_vs_jumps <- CUB_vs_jumps + stat_summary(data = filter(all_feature_changes, endoSymb_status==0), fun.data="median_hilow", fun.args = list(conf.int=0.5), geom="crossbar", size=0.5, inherit.aes = F, mapping = aes(x = directionGC, y = CUB, color=directionGC))
CUB_vs_jumps <- CUB_vs_jumps

CUB_wilcox_down <- wilcox.test(x = all_feature_changes$CUB[all_feature_changes$endoSymb_status==0 & all_feature_changes$directionGC==-1], mu = 1, conf.int=T)
CUB_wilcox_up <- wilcox.test(x = all_feature_changes$CUB[all_feature_changes$endoSymb_status==0 & all_feature_changes$directionGC==1], mu = 1, conf.int=T)
CUB_vs_jumps <- CUB_vs_jumps + annotate("text", x = 1, y = 6, label=paste("p = ", format(CUB_wilcox_down$p.value, digits = 3, scientific = ifelse(CUB_wilcox_down$p.value<0.001,T,F))), color=cbPalette[7])
CUB_vs_jumps <- CUB_vs_jumps + annotate("text", x = 2, y = 6, label=paste("p = ", format(CUB_wilcox_up$p.value, digits = 3, scientific = ifelse(CUB_wilcox_up$p.value<0.001,T,F))), color=cbPalette[6])
CUB_vs_jumps

rna_bLength_vs_GC <- ggplot(filter(all_feature_changes, is.finite(rna_bLength_log2)), mapping = aes(x = GC, y = rna_bLength, color=directionGC, shape=endoSymb_status)) + geom_point() + theme_cowplot()
rna_bLength_vs_GC <- rna_bLength_vs_GC + geom_hline(yintercept = 1, linetype="dashed", size=1.05, color="grey")
rna_bLength_vs_GC <- rna_bLength_vs_GC + scale_shape_manual(values = c(21,19)) + scale_x_continuous(name = col_labels["GC"], breaks = seq(-40,20,10), labels = c("-40", "-30", "-20", "-10", "0", "10", "20"), limits=c(-40,25)) + scale_y_continuous(name = "fold change in\n16s rRNA branch length", trans = "log2", limits = c(1/16,32), breaks = c(0.25,1,4,16), labels = c("0.25", "1", "4", "16")) + scale_color_manual(values = cbPalette[c(7,6)])
rna_bLength_vs_GC <- rna_bLength_vs_GC + rremove("legend")
rna_bLength_vs_GC <- rna_bLength_vs_GC + stat_cor(data = filter(all_feature_changes, endoSymb_status==0), method = "spearman", mapping = aes(color = directionGC), label.y.npc = 1, label.x.npc = 0, cor.coef.name = "rho")
rna_bLength_vs_GC

rna_bLength_vs_jumps <- ggplot(data = all_feature_changes[!is.na(all_feature_changes$directionGC),], mapping = aes(x = directionGC, y = rna_bLength, color=directionGC, fill=endoSymb_status))
rna_bLength_vs_jumps <- rna_bLength_vs_jumps + geom_dotplot(dotsize=0.5, mapping = aes(fill=endoSymb_status), binaxis='y', stackdir='center') + theme_cowplot()
rna_bLength_vs_jumps <- rna_bLength_vs_jumps + scale_x_discrete(name="", labels = c("downward", "upward")) + scale_y_continuous(name = "fold change in\n16s rRNA branch length", trans="log2", limits = c(1/16, 32)) + scale_color_manual(values = cbPalette[c(7,6)], guide=F) + scale_fill_manual(values = c(NA, cbPalette[7]), guide=F)
rna_bLength_vs_jumps <- rna_bLength_vs_jumps + geom_hline(yintercept = 1, linetype="dashed", size=1.05, color="darkgrey")
rna_bLength_vs_jumps <- rna_bLength_vs_jumps + stat_summary(data = filter(all_feature_changes, endoSymb_status==0), fun.data="median_hilow", fun.args = list(conf.int=0.5), geom="crossbar", size=0.5, inherit.aes = F, mapping = aes(x = directionGC, y = rna_bLength, color=directionGC))
rna_bLength_vs_jumps <- rna_bLength_vs_jumps

rna_bLength_wilcox_down <- wilcox.test(x = all_feature_changes$rna_bLength[all_feature_changes$endoSymb_status==0 & all_feature_changes$directionGC==-1], mu = 1, conf.int=T)
rna_bLength_wilcox_up <- wilcox.test(x = all_feature_changes$rna_bLength[all_feature_changes$endoSymb_status==0 & all_feature_changes$directionGC==1], mu = 1, conf.int=T)
rna_bLength_vs_jumps <- rna_bLength_vs_jumps + annotate("text", x = 1, y = 24, label=paste("p = ", format(rna_bLength_wilcox_down$p.value, digits = 3, scientific = ifelse(rna_bLength_wilcox_down$p.value<0.001,T,F))), color=cbPalette[7])
rna_bLength_vs_jumps <- rna_bLength_vs_jumps + annotate("text", x = 2, y = 24, label=paste("p = ", format(rna_bLength_wilcox_up$p.value, digits = 3, scientific = ifelse(rna_bLength_wilcox_up$p.value<0.001,T,F))), color=cbPalette[6])
rna_bLength_vs_jumps

cDensity_vs_GC <- ggplot(data = all_feature_changes, mapping = aes(x = GC, y = cDensity, color=directionGC, shape=endoSymb_status)) + geom_point() + theme_cowplot()
cDensity_vs_GC <- cDensity_vs_GC + geom_hline(yintercept = 0, linetype="dashed", size=1.05, color="grey")
cDensity_vs_GC <- cDensity_vs_GC + scale_shape_manual(values = c(21,19)) + scale_x_continuous(name= col_labels["GC"], breaks = seq(-40,20,10), labels = c("-40", "-30", "-20", "-10", "0", "10", "20"), limits = c(-40,25)) + scale_y_continuous(name=paste0("\U0394", "coding density\n(in % points)"), limits = c(-35,20)) + scale_color_manual(values = cbPalette[c(7,6)])
cDensity_vs_GC <- cDensity_vs_GC + rremove("legend")
cDensity_vs_GC <- cDensity_vs_GC + stat_cor(data = filter(all_feature_changes, endoSymb_status==0), method = "spearman", mapping = aes(color = directionGC), label.y.npc = 1, label.x.npc = 0, cor.coef.name = "rho")
cDensity_vs_GC

cDensity_vs_jumps <- ggplot(data = all_feature_changes[!is.na(all_feature_changes$directionGC),], mapping = aes(x = directionGC, y = cDensity, color=directionGC, fill=endoSymb_status))
cDensity_vs_jumps <- cDensity_vs_jumps + geom_dotplot(dotsize=0.5, mapping = aes(fill=endoSymb_status), binaxis='y', stackdir='center') + theme_cowplot()
cDensity_vs_jumps <- cDensity_vs_jumps + scale_x_discrete(name="", labels = c("downward", "upward")) + scale_y_continuous(name=paste0("\U0394","coding density\n(in % points)"), limits = c(-35,20)) + scale_color_manual(values = cbPalette[c(7,6)], guide=F) + scale_fill_manual(values = c(NA, cbPalette[7]), guide=F)
cDensity_vs_jumps <- cDensity_vs_jumps + geom_hline(yintercept = 1, linetype="dashed", size=1.05, color="darkgrey")
cDensity_vs_jumps <- cDensity_vs_jumps + stat_summary(data = filter(all_feature_changes, endoSymb_status==0), fun.data="median_hilow", fun.args = list(conf.int=0.5), geom="crossbar", size=0.5, inherit.aes = F, mapping = aes(x = directionGC, y = cDensity, color=directionGC))
cDensity_vs_jumps <- cDensity_vs_jumps

cDensity_wilcox_down <- wilcox.test(x = all_feature_changes$cDensity[all_feature_changes$endoSymb_status==0 & all_feature_changes$directionGC==-1], mu = 0, conf.int=T)
cDensity_wilcox_up <- wilcox.test(x = all_feature_changes$cDensity[all_feature_changes$endoSymb_status==0 & all_feature_changes$directionGC==1], mu = 0, conf.int=T)
cDensity_vs_jumps <- cDensity_vs_jumps + annotate("text", x = 1, y = 15, label=paste("p = ", format(cDensity_wilcox_down$p.value, digits = 3, scientific = ifelse(cDensity_wilcox_down$p.value<0.001,T,F))), color=cbPalette[7])
cDensity_vs_jumps <- cDensity_vs_jumps + annotate("text", x = 2, y = 15, label=paste("p = ", format(cDensity_wilcox_up$p.value, digits = 3, scientific = ifelse(cDensity_wilcox_up$p.value<0.001,T,F))), color=cbPalette[6])
cDensity_vs_jumps

IGRlength_vs_GC <- ggplot(data = all_feature_changes, mapping = aes(x = GC, y = IGRlength, color=directionGC, shape=endoSymb_status)) + geom_point() + theme_cowplot()
IGRlength_vs_GC <- IGRlength_vs_GC + geom_hline(yintercept = 0, linetype="dashed", size=1.05, color="grey")
IGRlength_vs_GC <- IGRlength_vs_GC + scale_shape_manual(values = c(21,19)) + scale_x_continuous(name= col_labels["GC"], breaks = seq(-40,20,10), labels = c("-40", "-30", "-20", "-10", "0", "10", "20"), limits=c(-40,25)) + scale_y_continuous(name=paste0("\U0394","IGR length\n(in basepairs)")) + scale_color_manual(values = c(cbPalette[c(7,6)]))
IGRlength_vs_GC <- IGRlength_vs_GC + rremove("legend")
IGRlength_vs_GC <- IGRlength_vs_GC + stat_cor(data = filter(all_feature_changes, endoSymb_status==0), method = "spearman", mapping = aes(color = directionGC), label.y.npc = 1, label.x.npc = 0, cor.coef.name = "rho")
IGRlength_vs_GC

IGRlength_vs_jumps <- ggplot(data = all_feature_changes[!is.na(all_feature_changes$directionGC),], mapping = aes(x = directionGC, y = IGRlength, color=directionGC, fill=endoSymb_status))
IGRlength_vs_jumps <- IGRlength_vs_jumps + geom_dotplot(dotsize=0.5, mapping = aes(fill=endoSymb_status), binaxis='y', stackdir='center') + theme_cowplot()
IGRlength_vs_jumps <- IGRlength_vs_jumps + scale_x_discrete(name="", labels = c("downward", "upward")) + scale_y_continuous(name=paste0("\U0394","IGR length\n(in basepairs)")) + scale_color_manual(values = cbPalette[c(7,6)], guide=F) + scale_fill_manual(values = c(NA, cbPalette[7]), guide=F)
IGRlength_vs_jumps <- IGRlength_vs_jumps + geom_hline(yintercept = 1, linetype="dashed", size=1.05, color="darkgrey")
IGRlength_vs_jumps <- IGRlength_vs_jumps + stat_summary(data = filter(all_feature_changes, endoSymb_status==0), fun.data="median_hilow", fun.args = list(conf.int=0.5), geom="crossbar", size=0.5, inherit.aes = F, mapping = aes(x = directionGC, y = IGRlength, color=directionGC))
IGRlength_vs_jumps <- IGRlength_vs_jumps

IGRlength_wilcox_down <- wilcox.test(x = all_feature_changes$IGRlength[all_feature_changes$endoSymb_status==0 & all_feature_changes$directionGC==-1], mu = 0, conf.int=T)
IGRlength_wilcox_up <- wilcox.test(x = all_feature_changes$IGRlength[all_feature_changes$endoSymb_status==0 & all_feature_changes$directionGC==1], mu = 0, conf.int=T)
IGRlength_vs_jumps <- IGRlength_vs_jumps + annotate("text", x = 1, y = 120, label=paste("p = ", format(IGRlength_wilcox_down$p.value, digits = 3, scientific = ifelse(IGRlength_wilcox_down$p.value<0.001,T,F))), color=cbPalette[7])
IGRlength_vs_jumps <- IGRlength_vs_jumps + annotate("text", x = 2, y = 120, label=paste("p = ", format(IGRlength_wilcox_up$p.value, digits = 3, scientific = ifelse(IGRlength_wilcox_up$p.value<0.001,T,F))), color=cbPalette[6])
IGRlength_vs_jumps

pseudoFrac_vs_GC <- ggplot(data = filter(all_feature_changes, is.finite(pseudoFrac_log2)), mapping = aes(x = GC, y = pseudoFrac, color=directionGC, shape=endoSymb_status)) + geom_point() + theme_cowplot()
pseudoFrac_vs_GC <- pseudoFrac_vs_GC + geom_hline(yintercept = 1, linetype="dashed", size=1.05, color="grey")
pseudoFrac_vs_GC <- pseudoFrac_vs_GC + scale_shape_manual(values = c(21,19)) + scale_x_continuous(name =  col_labels["GC"], breaks = seq(-40,20,10), labels = c("-40", "-30", "-20", "-10", "0", "10", "20"), limits = c(-40,25)) + scale_y_continuous(name = "fold change in\npseudogene fraction", trans = "log2", breaks = c(0.25,1,4,16), labels = c("0.25", "1", "4", "16")) + scale_color_manual(values = cbPalette[c(7,6)])
pseudoFrac_vs_GC <- pseudoFrac_vs_GC + rremove("legend")
pseudoFrac_vs_GC <- pseudoFrac_vs_GC + stat_cor(data = filter(all_feature_changes, endoSymb_status==0), method = "spearman", mapping = aes(color = directionGC), label.y.npc = 1, label.x.npc = 0, cor.coef.name = "rho")
# pseudoFrac_vs_GC <- pseudoFrac_vs_GC + coord_cartesian(ylim = c(min(all_feature_changes$pseudoFrac, na.rm = T), 1.1*(max(all_feature_changes$pseudoFrac, na.rm = T))))
# pseudoFrac_vs_GC

pseudoFrac_vs_jumps <- ggplot(data = all_feature_changes[!is.na(all_feature_changes$directionGC),], mapping = aes(x = directionGC, y = pseudoFrac, color=directionGC, fill=endoSymb_status))
pseudoFrac_vs_jumps <- pseudoFrac_vs_jumps + geom_dotplot(dotsize=0.5, mapping = aes(fill=endoSymb_status), binaxis='y', stackdir='center') + theme_cowplot()
pseudoFrac_vs_jumps <- pseudoFrac_vs_jumps + scale_x_discrete(name="", labels = c("downward", "upward")) + scale_y_continuous(name = "fold change in\npseudogene fraction", trans="log2", breaks = c(0.25,1,4,16), labels = c("0.25", "1", "4", "16")) + scale_color_manual(values = cbPalette[c(7,6)], guide=F) + scale_fill_manual(values = c(NA, cbPalette[7]), guide=F)
pseudoFrac_vs_jumps <- pseudoFrac_vs_jumps + geom_hline(yintercept = 1, linetype="dashed", size=1.05, color="darkgrey")
pseudoFrac_vs_jumps <- pseudoFrac_vs_jumps + stat_summary(data = filter(all_feature_changes, endoSymb_status==0), fun.data="median_hilow", fun.args = list(conf.int=0.5), geom="crossbar", size=0.5, inherit.aes = F, mapping = aes(x = directionGC, y = pseudoFrac, color=directionGC), alpha=0.7)#, color="darkgrey", alpha=0.5)
pseudoFrac_vs_jumps <- pseudoFrac_vs_jumps

pseudoFrac_wilcox_down <- wilcox.test(x = all_feature_changes$pseudoFrac[all_feature_changes$endoSymb_status==0 & all_feature_changes$directionGC==-1], mu = 1, conf.int=T)
pseudoFrac_wilcox_up <- wilcox.test(x = all_feature_changes$pseudoFrac[all_feature_changes$endoSymb_status==0 & all_feature_changes$directionGC==1], mu = 1, conf.int=T)
pseudoFrac_vs_jumps <- pseudoFrac_vs_jumps + annotate("text", x = 1, y = 26, label=paste("p = ", format(pseudoFrac_wilcox_down$p.value, digits = 3, scientific = ifelse(pseudoFrac_wilcox_down$p.value<0.001,T,F))), color=cbPalette[7])
pseudoFrac_vs_jumps <- pseudoFrac_vs_jumps + annotate("text", x = 2, y = 26, label=paste("p = ", format(pseudoFrac_wilcox_up$p.value, digits = 3, scientific = ifelse(pseudoFrac_wilcox_up$p.value<0.001,T,F))), color=cbPalette[6])
pseudoFrac_vs_jumps

all_vars_vs_GC <- gSize_vs_GC + CUB_vs_GC + rna_bLength_vs_GC + cDensity_vs_GC  + IGRlength_vs_GC + pseudoFrac_vs_GC + plot_annotation(tag_levels = "A") + plot_layout(ncol = 2) 

#plot_grid(plotlist = list(gSize_vs_GC, gSize_vs_jumps, CUB_vs_GC, CUB_vs_jumps, rna_bLength_vs_GC, rna_bLength_vs_jumps, cDensity_vs_GC, cDensity_vs_jumps, IGRlength_vs_GC, IGRlength_vs_jumps, pseudoFrac_vs_GC, pseudoFrac_vs_jumps), align = "hv", ncol = 4, label_size = 18,  rel_widths = c(2,1,2,1)) + scale_shape_manual(values = c(21,19))

ggsave(filename = paste0(jump_common_dir, "all_feature_vs_GC_corrs.svg"), plot = all_vars_vs_GC, width = 8, height=9, units = "in", dpi = 300)
ggsave(filename = paste0(jump_common_dir, "all_feature_vs_GC_corrs.png"), plot = all_vars_vs_GC, width = 8, height=9, units = "in", dpi = 300)

all_vars_vs_jumps <- gSize_vs_jumps + CUB_vs_jumps + rna_bLength_vs_jumps + cDensity_vs_jumps  + IGRlength_vs_jumps + pseudoFrac_vs_jumps + plot_annotation(tag_levels = "A") + plot_layout(ncol = 2)

ggsave(filename = paste0(jump_common_dir, "all_feature_vs_jumps.svg"), plot = all_vars_vs_jumps, width = 7, height=9, units = "in", dpi = 300)
ggsave(filename = paste0(jump_common_dir, "all_feature_vs_jumps.png"), plot = all_vars_vs_jumps, width = 7, height=9, units = "in", dpi = 300)
```

### Transformation, scaling, and clustering
First, I will cluster only those datasets where full data is available. I will use consensus clustering which summarizes clustering across multiple methods and parameters.
```{r summarize_transform_changes }
all_feature_changes_transformed <- all_feature_changes[,c("GC", "CUB_log2", "rna_bLength_log2", "gSize_log2", "IGRpercentage", "pseudoFrac_log2", "cDensity")]

scale_factors <- sapply(1:ncol(all_feature_changes_transformed), function(x) sd(all_feature_changes_transformed[,x], na.rm=T))
all_feature_changes_scaled <- scale(x = all_feature_changes_transformed, center = F, scale = scale_factors)
all_feature_changes_scaled <- all_feature_changes_scaled[!apply(is.na(all_feature_changes_scaled), 1, all),] #remove rows with only NAs
all_feature_changes_scaled <- as.data.frame(all_feature_changes_scaled)

all_feature_changes_scaled_noNA <- 
all_feature_changes_scaled[!apply(is.na(all_feature_changes_scaled), 1, any),] #remove rows with any NAs
all_feature_changes_scaled_noNA <- as.data.frame(all_feature_changes_scaled_noNA)
```

#### Clustering
Since data for some genomic features is missing for many datasets, I will use two approaches. One, cluster only those datasets for which data for all genomic features is available. Two, cluster all datasets using only GC, genome size, and coding density, whose data is available for all datasets. I will first cluster only those datasets where data is available for all genomic features.

There are many clustering methods and algorithms with different assumptions. Consensus clustering combines all and possibly filters noise from each clustering method and gives a robust clustering. 

In general, I will use two approaches, one is based on proportion of ambiguous clusters (PAC) and another on MonteCarlo sampling based comparison with null hypothesis (no clusters).
##### Complete data, based on PAC using diceR
```{r clustering_noNA_data_diceR_PAC_k2_all }
library(diceR)

all_feature_changes_scaled_noNA_cc <- consensus_cluster(data = all_feature_changes_scaled_noNA, nk = 2, p.item = 0.8, reps = 100, algorithms = c("hc", "pam", "diana", "cmeans", "km", "block", "som", "nmf"), scale = F, progress = T)
all_feature_changes_scaled_noNA_ccomb_class_indices <- consensus_evaluate(all_feature_changes_scaled_noNA, all_feature_changes_scaled_noNA_cc)

all_feature_changes_scaled_noNA_ccomb_class_indices$pac
```

I will retain only those algorithms that cluster >75% samples unambiguously across bootstraps.
```{r clustering_noNA_data_diceR_PAC_k2 }
all_feature_changes_scaled_noNA_cc <- consensus_cluster(data = all_feature_changes_scaled_noNA, nk = 2, p.item = 0.8, reps = 100, algorithms = c("hc", "diana", "km", "block", "som"), scale = F, progress = T)

all_feature_changes_scaled_noNA_ccomb_class <- consensus_combine(all_feature_changes_scaled_noNA_cc, element = "class")
all_feature_changes_scaled_noNA_ccomb_class_cm <- consensus_matrix(all_feature_changes_scaled_noNA_ccomb_class)
rownames(all_feature_changes_scaled_noNA_ccomb_class_cm) <- colnames(all_feature_changes_scaled_noNA_ccomb_class_cm) <- rownames(all_feature_changes_scaled_noNA)

all_feature_changes_scaled_noNA_ccomb_class_cm_hclust <- as.hclust(ladderize(as.dendrogram(hclust(d = dist(all_feature_changes_scaled_noNA_ccomb_class_cm), method = "ward.D2"))))
all_feature_changes_scaled_noNA_ccomb_class_cm_dend <- as.dendrogram(all_feature_changes_scaled_noNA_ccomb_class_cm_hclust)
all_feature_changes_scaled_noNA_ccomb_class_cm_dend <- as.ggdend(all_feature_changes_scaled_noNA_ccomb_class_cm_dend)
all_feature_changes_scaled_noNA_ccomb_class_cm_dend_plot <- ggplot(all_feature_changes_scaled_noNA_ccomb_class_cm_dend, horiz = T, labels = F, theme = NULL) + scale_y_reverse() + theme_cowplot() + theme(panel.border = element_blank(), plot.margin = unit(x = c(0,0,0,0), units = "in"), axis.line = element_blank(), axis.title = element_blank(), axis.ticks = element_blank(), axis.text = element_blank())

all_feature_changes_scaled_noNA_ccomb_class_cm.long <- melt(data = all_feature_changes_scaled_noNA_ccomb_class_cm, varnames=c("ds1", "ds2"))
all_feature_changes_scaled_noNA_ccomb_class_cm.long$ds1 <-  factor(all_feature_changes_scaled_noNA_ccomb_class_cm.long$ds1, levels = all_feature_changes_scaled_noNA_ccomb_class_cm_hclust$labels[all_feature_changes_scaled_noNA_ccomb_class_cm_hclust$order], ordered = T)
all_feature_changes_scaled_noNA_ccomb_class_cm.long$ds2 <- factor(all_feature_changes_scaled_noNA_ccomb_class_cm.long$ds2, levels = all_feature_changes_scaled_noNA_ccomb_class_cm_hclust$labels[all_feature_changes_scaled_noNA_ccomb_class_cm_hclust$order], ordered = T)

# all_feature_changes_scaled_noNA_ccomb_class_cm.long$ds1 <-  factor(all_feature_changes_scaled_noNA_ccomb_class_cm.long$ds1, levels = all_hclust_ladder$labels[all_hclust_ladder$order])
# all_feature_changes_scaled_noNA_ccomb_class_cm.long$ds2 <- factor(all_feature_changes_scaled_noNA_ccomb_class_cm.long$ds2, levels = all_hclust_ladder$labels[all_hclust_ladder$order])

all_feature_changes_scaled_noNA_ccomb_class_cm_plot <- ggplot(data=all_feature_changes_scaled_noNA_ccomb_class_cm.long, mapping=aes(x = ds1, y = ds2, fill=value)) + geom_tile()
all_feature_changes_scaled_noNA_ccomb_class_cm_plot <- all_feature_changes_scaled_noNA_ccomb_class_cm_plot + scale_x_discrete(name = "datasets") + scale_y_discrete(name = "datasets") + scale_fill_gradient(low = "lightgreen", high = "darkgreen", guide = guide_colorbar(title = "co-clustering frequency", title.position = "top", label.position = "bottom", title.vjust = 0.5, title.hjust = 0))# + scale_fill_viridis(direction = -1, guide = guide_colorbar(title = "co-clustering frequency", title.position = "top", label.position = "bottom", title.vjust = 0.5, title.hjust = 0))
all_feature_changes_scaled_noNA_ccomb_class_cm_plot <- all_feature_changes_scaled_noNA_ccomb_class_cm_plot + theme(axis.text.x = element_text(angle = 90, size = 6, hjust = 1, vjust = 0.5, face =  ifelse(all_feature_changes[all_feature_changes_scaled_noNA_ccomb_class_cm_hclust$labels[all_feature_changes_scaled_noNA_ccomb_class_cm_hclust$order], "endoSymb_status"]==1, "bold", "plain")),axis.text.y = element_text(size = 6, face = ifelse(all_feature_changes[all_feature_changes_scaled_noNA_ccomb_class_cm_hclust$labels[all_feature_changes_scaled_noNA_ccomb_class_cm_hclust$order], "endoSymb_status"]==1, "bold", "plain")), legend.position = "top", legend.direction = "horizontal", axis.title.y = element_blank())

# Heatmap of changes
##############################################
all_feature_changes_scaled_noNA.long <- melt(data = as.matrix(all_feature_changes_scaled_noNA), varnames = c("datasets", "features"))
all_feature_changes_scaled_noNA.long$datasets <- factor(x = all_feature_changes_scaled_noNA.long$datasets, levels = all_feature_changes_scaled_noNA_ccomb_class_cm_hclust$labels[all_feature_changes_scaled_noNA_ccomb_class_cm_hclust$order], ordered = T)
all_feature_changes_scaled_noNA.long$features <- factor(x = all_feature_changes_scaled_noNA.long$features, levels = c("GC", "gSize_log2", "cDensity", "CUB_log2", "rna_bLength_log2", "IGRpercentage", "pseudoFrac_log2"), ordered = T)

all_feature_changes_scaled_noNA_heatmap_plot <- ggplot(data = all_feature_changes_scaled_noNA.long, mapping = aes(x = features, y = datasets, fill=value)) + geom_tile()
all_feature_changes_scaled_noNA_heatmap_plot <- all_feature_changes_scaled_noNA_heatmap_plot + scale_color_gradient2(aesthetics = "fill", low = cbPalette[7], mid = "white", high = cbPalette[6], guide = guide_colorbar(title = "normalized differences", title.position = "top", label.position = "bottom", title.vjust = 0.75, title.hjust = 0.5)) + scale_x_discrete(labels=c("GC", "log2(genome size)", "coding density", "log2(CUB)", "log2(16s branch length)", "IGR length", "log2(pseudogene fraction)"))
all_feature_changes_scaled_noNA_heatmap_plot <- all_feature_changes_scaled_noNA_heatmap_plot + theme(axis.text.x = element_text(angle = 90, hjust=1, vjust=0.5), axis.text.y = element_blank(), legend.position = "top", legend.direction = "horizontal", axis.title.y = element_blank(), axis.ticks.y = element_blank(), axis.text.x.bottom = element_text(size = 8))

dendro_cc_changes_combined_plot <- all_feature_changes_scaled_noNA_ccomb_class_cm_dend_plot + all_feature_changes_scaled_noNA_ccomb_class_cm_plot + all_feature_changes_scaled_noNA_heatmap_plot + plot_layout(widths = c(3,10,2))

ggsave(filename = "all_feature_changes_noNA_diceR_k2_cm.svg", path = jump_common_dir, plot = dendro_cc_changes_combined_plot, device = "svg", width = 12, height = 9, units = "in", dpi = 300)
ggsave(filename ="all_feature_changes_noNA_diceR_k2_cm.png", path = jump_common_dir, plot = dendro_cc_changes_combined_plot, device = "png", width = 12, height = 9, units = "in", dpi = 300)
```
Consensus indices show that the endosymbionts cluster together most times, and other datasets cluster together most times, although there is some possible sub-structure. E.g. One of the endosymbionts does not always cluster with rest, and some non-endosymbionts dont always cluster with remaining datasets. However, broadly, there are 2 clusters, one with endosymbionts and another with non-endosymbionts.

Now let us try the MonteCarlo sampling method to identify the number of clusters.
##### Complete data, based on MC sub-sampling
```{r clustering_noNA_data_M3C_k28_all }
library(M3C)
all_feature_changes_scaled_noNA_M3C_k28 <- M3C(mydata = t(all_feature_changes_scaled_noNA), cores = 3, iters = 100, maxK = 8)
```

We find that the optimal k=2 based on PAC, RCSI, and p-values.

```{r clustering_noNA_data_M3C_k28 }
all_feature_changes_scaled_noNA_M3C_k2_cm <- all_feature_changes_scaled_noNA_M3C_k28$realdataresults[[2]]$consensus_matrix
rownames(all_feature_changes_scaled_noNA_M3C_k2_cm) <- rownames(all_feature_changes_scaled_noNA)[as.integer(colnames(all_feature_changes_scaled_noNA_M3C_k2_cm))]
colnames(all_feature_changes_scaled_noNA_M3C_k2_cm) <- rownames(all_feature_changes_scaled_noNA)[as.integer(colnames(all_feature_changes_scaled_noNA_M3C_k2_cm))]

all_feature_changes_scaled_noNA_M3C_k2_cm_hclust <- as.hclust(ladderize(as.dendrogram(hclust(d = dist(all_feature_changes_scaled_noNA_M3C_k2_cm), method = "ward.D2"))))
all_feature_changes_scaled_noNA_M3C_k2_cm_dend <- as.dendrogram(all_feature_changes_scaled_noNA_M3C_k2_cm_hclust)
all_feature_changes_scaled_noNA_M3C_k2_cm_dend <- as.ggdend(all_feature_changes_scaled_noNA_M3C_k2_cm_dend)
all_feature_changes_scaled_noNA_M3C_k2_cm_dend_plot <- ggplot(all_feature_changes_scaled_noNA_M3C_k2_cm_dend, horiz = T, labels = F, theme = NULL) + scale_y_reverse() + theme_cowplot() + theme(panel.border = element_blank(), plot.margin = unit(x = c(0,0,0,0), units = "in"), axis.line = element_blank(), axis.title = element_blank(), axis.ticks = element_blank(), axis.text = element_blank())

all_feature_changes_scaled_noNA_M3C_k2_cm.long <- melt(data = all_feature_changes_scaled_noNA_M3C_k2_cm, varnames=c("ds1", "ds2"))
all_feature_changes_scaled_noNA_M3C_k2_cm.long$ds1 <-  factor(all_feature_changes_scaled_noNA_M3C_k2_cm.long$ds1, levels = all_feature_changes_scaled_noNA_M3C_k2_cm_hclust$labels[all_feature_changes_scaled_noNA_M3C_k2_cm_hclust$order], ordered = T)
all_feature_changes_scaled_noNA_M3C_k2_cm.long$ds2 <- factor(all_feature_changes_scaled_noNA_M3C_k2_cm.long$ds2, levels = all_feature_changes_scaled_noNA_M3C_k2_cm_hclust$labels[all_feature_changes_scaled_noNA_M3C_k2_cm_hclust$order], ordered = T)

# all_feature_changes_scaled_noNA_M3C_k2_cm.long$ds1 <-  factor(all_feature_changes_scaled_noNA_M3C_k2_cm.long$ds1, levels = all_hclust_ladder$labels[all_hclust_ladder$order])
# all_feature_changes_scaled_noNA_M3C_k2_cm.long$ds2 <- factor(all_feature_changes_scaled_noNA_M3C_k2_cm.long$ds2, levels = all_hclust_ladder$labels[all_hclust_ladder$order])

all_feature_changes_scaled_noNA_M3C_k2_cm_plot <- ggplot(data=all_feature_changes_scaled_noNA_M3C_k2_cm.long, mapping=aes(x = ds1, y = ds2, fill=value)) + geom_tile()
all_feature_changes_scaled_noNA_M3C_k2_cm_plot <- all_feature_changes_scaled_noNA_M3C_k2_cm_plot + scale_x_discrete(name = "datasets") + scale_y_discrete(name = "datasets") + scale_fill_gradient(low = "lightgreen", high = "darkgreen", guide = guide_colorbar(title = "co-clustering frequency", title.position = "top", label.position = "bottom", title.vjust = 0.5, title.hjust = 0))# + scale_fill_viridis(direction = -1, guide = guide_colorbar(title = "co-clustering frequency", title.position = "top", label.position = "bottom", title.vjust = 0.5, title.hjust = 0))
face_oi <- ifelse(all_feature_changes[all_feature_changes_scaled_noNA_M3C_k2_cm_hclust$labels[all_feature_changes_scaled_noNA_M3C_k2_cm_hclust$order], "endoSymb_status"]==1, "bold", "plain")
all_feature_changes_scaled_noNA_M3C_k2_cm_plot <- all_feature_changes_scaled_noNA_M3C_k2_cm_plot + theme(axis.text.x = element_text(angle = 90, size = 6, hjust = 1, vjust = 0.5, face = face_oi),axis.text.y = element_text(size = 6, face =  face_oi), legend.position = "top", legend.direction = "horizontal", axis.title.y = element_blank())

# Heatmap of changes
##############################################
all_feature_changes_scaled_noNA.long <- melt(data = as.matrix(all_feature_changes_scaled_noNA), varnames = c("datasets", "features"))
all_feature_changes_scaled_noNA.long$datasets <- factor(x = all_feature_changes_scaled_noNA.long$datasets, levels = all_feature_changes_scaled_noNA_M3C_k2_cm_hclust$labels[all_feature_changes_scaled_noNA_M3C_k2_cm_hclust$order], ordered = T)
all_feature_changes_scaled_noNA.long$features <- factor(x = all_feature_changes_scaled_noNA.long$features, levels = c("GC", "gSize_log2", "cDensity", "CUB_log2", "rna_bLength_log2", "IGRpercentage", "pseudoFrac_log2"), ordered = T)

all_feature_changes_scaled_noNA_heatmap_plot <- ggplot(data = all_feature_changes_scaled_noNA.long, mapping = aes(x = features, y = datasets, fill=value)) + geom_tile()
all_feature_changes_scaled_noNA_heatmap_plot <- all_feature_changes_scaled_noNA_heatmap_plot + scale_color_gradient2(aesthetics = "fill", low = cbPalette[7], mid = "white", high = cbPalette[6], guide = guide_colorbar(title = "normalized differences", title.position = "top", label.position = "bottom", title.vjust = 0.75, title.hjust = 0.5)) + scale_x_discrete(labels=c("GC", "log2(genome size)", "coding density", "log2(CUB)", "log2(16s branch length)", "IGR length", "log2(pseudogene fraction)"))
all_feature_changes_scaled_noNA_heatmap_plot <- all_feature_changes_scaled_noNA_heatmap_plot + theme(axis.text.x = element_text(angle = 90, hjust=1, vjust=0.5), axis.text.y = element_blank(), legend.position = "top", legend.direction = "horizontal", axis.title.y = element_blank(), axis.ticks.y = element_blank(), axis.text.x.bottom = element_text(size = 8))

dendro_cc_changes_combined_plot <- all_feature_changes_scaled_noNA_M3C_k2_cm_dend_plot + all_feature_changes_scaled_noNA_M3C_k2_cm_plot + all_feature_changes_scaled_noNA_heatmap_plot + plot_layout(widths = c(3,10,2))

ggsave(filename = "all_feature_changes_noNA_M3C_k2_cm.svg", path = jump_common_dir, plot = dendro_cc_changes_combined_plot, device = "svg", width = 12, height = 9, units = "in", dpi = 300)
ggsave(filename ="all_feature_changes_noNA_M3C_k2_cm.png", path = jump_common_dir, plot = dendro_cc_changes_combined_plot, device = "png", width = 12, height = 9, units = "in", dpi = 300)
```
As expected, one cluster refers to endosymbionts and another to non-endosymbionts.

Deepa asks, what happens if we remove endosymbionts, does it reveal additional structure? Lets check by clustering after removing endosymbionts.
##### Complete data but no endosymbionts, based on MC sub-sampling
```{r clustering_noNA_noEndo_data_M3C_k28_all }
endo_names <- rownames(all_feature_changes)[which(all_feature_changes$endoSymb_status==1)]
all_feature_changes_scaled_noNA_noEndo <- all_feature_changes_scaled_noNA[!rownames(all_feature_changes_scaled_noNA) %in% endo_names, ]
library(M3C)
all_feature_changes_scaled_noNA_noEndo_M3C_k28 <- M3C(mydata = t(all_feature_changes_scaled_noNA_noEndo), cores = 3, iters = 100, maxK = 8)
```
There is no more structure in the datasets when endosymbionts are removed.

##### All data, based on M3C 
Now let us see what happens when we clsuter all datasets, but only based on GC, genome size, and coding density (for which we have full data).
```{r clustering_all_data_M3C_k28_all }
all_feature_changes_scaled_3Params <- all_feature_changes_scaled[,c("GC", "gSize_log2", "cDensity")]
library(M3C)
all_feature_changes_scaled_3Params_M3C_k28 <- M3C(mydata = t(all_feature_changes_scaled_3Params), cores = 3, iters = 100, maxK = 8)
```

In this data, although 3 clusters are the best supported (according to p-value), even 5 clusters are significantly supported compared to MonteCarlo simulations (p < 0.05). So, I will plot the consensus clustering statistics assuming 5 clusters.
```{r clustering_all_data_M3C_k28 }
# all_feature_changes_scaled_3Params_M3C_k5_cm <- all_feature_changes_scaled_3Params_M3C_k28$realdataresults[[5]]$consensus_matrix
# rownames(all_feature_changes_scaled_3Params_M3C_k5_cm) <- rownames(all_feature_changes_scaled_3Params)[as.integer(colnames(all_feature_changes_scaled_3Params_M3C_k5_cm))]
# colnames(all_feature_changes_scaled_3Params_M3C_k5_cm) <- rownames(all_feature_changes_scaled_3Params)[as.integer(colnames(all_feature_changes_scaled_3Params_M3C_k5_cm))]
# write.table(all_feature_changes_scaled_3Params_M3C_k5_cm, file = paste0(jump_common_dir, "all_feature_changes_scaled_3Params_M3C_k5_cm.tab"), quote = F)

all_feature_changes_scaled_3Params_M3C_k5_cm <- read.table(file = paste0(jump_common_dir, "all_feature_changes_scaled_3Params_M3C_k5_cm.tab"), stringsAsFactors = F)
all_feature_changes_scaled_3Params_M3C_k5_cm <- as.matrix(all_feature_changes_scaled_3Params_M3C_k5_cm)

# all_feature_changes_scaled_3Params_M3C_k5_cm_hclust <- hclust(d = dist(all_feature_changes_scaled_3Params_M3C_k5_cm), method = "ward.D2")
# save(all_feature_changes_scaled_3Params_M3C_k5_cm_hclust, file = paste0(jump_common_dir, "all_feature_changes_scaled_3Params_M3C_k5_cm_hclust.RData"))
load(file = paste0(jump_common_dir, "all_feature_changes_scaled_3Params_M3C_k5_cm_hclust.RData"))

all_feature_changes_scaled_3Params_M3C_k5_cm_hclust <- dendextend::rotate(x = all_feature_changes_scaled_3Params_M3C_k5_cm_hclust, order = all_feature_changes_scaled_3Params_M3C_k5_cm_hclust$order[c(85:192, 1:84)]) ##rearrange
all_feature_changes_scaled_3Params_M3C_k5_cm_hclust <- dendextend::rotate(x = all_feature_changes_scaled_3Params_M3C_k5_cm_hclust, order = all_feature_changes_scaled_3Params_M3C_k5_cm_hclust$order[c(24:31, 1:23, 32:192)]) ##rearrange
# all_feature_changes_scaled_3Params_M3C_k5_cm_hclust <- dendextend::rotate(x = all_feature_changes_scaled_3Params_M3C_k5_cm_hclust, order = all_feature_changes_scaled_3Params_M3C_k5_cm_hclust$order[c(24:31) ,c(1:32), c(1, 84))]) ##rearrange
# all_feature_changes_scaled_3Params_M3C_k5_cm_hclust <- dendextend::rotate(x = all_feature_changes_scaled_3Params_M3C_k5_cm_hclust, order = c(1:162, seq(192,163,-1))) ##rearrange
# which(all_feature_changes_scaled_3Params_M3C_k5_cm_hclust$labels[all_feature_changes_scaled_3Params_M3C_k5_cm_hclust$order]=="Betaproteo_jump5")
# all_feature_changes_scaled_3Params_M3C_k5_cm_hclust <- dendextend::rotate(x=all_feature_changes_scaled_3Params_M3C_k5_cm_hclust, order=all_feature_changes_scaled_3Params_M3C_k5_cm_hclust$order[c(1:161, seq(192,162,-1))]) ##rearrange

all_feature_changes_scaled_3Params_M3C_k5_cm_dend <- as.dendrogram(all_feature_changes_scaled_3Params_M3C_k5_cm_hclust)
# all_feature_changes_scaled_3Params_M3C_k5_cm_dend <- as.ggdend(all_feature_changes_scaled_3Params_M3C_k5_cm_dend)
all_feature_changes_scaled_3Params_M3C_k5_cm_dend_plot <- ggplot(all_feature_changes_scaled_3Params_M3C_k5_cm_dend, horiz = T, labels = F, theme = NULL, size=0.05) + scale_y_reverse() + theme_cowplot() + theme(panel.border = element_blank(), plot.margin = unit(x = c(0,0,0,0), units = "in"), axis.line = element_blank(), axis.title = element_blank(), axis.ticks = element_blank(), axis.text = element_blank())

all_feature_changes_scaled_3Params_M3C_k5_cm.long <- melt(data = all_feature_changes_scaled_3Params_M3C_k5_cm, varnames=c("ds1", "ds2"))
all_feature_changes_scaled_3Params_M3C_k5_cm.long$ds1 <-  factor(all_feature_changes_scaled_3Params_M3C_k5_cm.long$ds1, levels = all_feature_changes_scaled_3Params_M3C_k5_cm_hclust$labels[all_feature_changes_scaled_3Params_M3C_k5_cm_hclust$order], ordered = T)
all_feature_changes_scaled_3Params_M3C_k5_cm.long$ds2 <- factor(all_feature_changes_scaled_3Params_M3C_k5_cm.long$ds2, levels = all_feature_changes_scaled_3Params_M3C_k5_cm_hclust$labels[all_feature_changes_scaled_3Params_M3C_k5_cm_hclust$order], ordered = T)

# all_feature_changes_scaled_3Params_M3C_k5_cm.long$ds1 <-  factor(all_feature_changes_scaled_3Params_M3C_k5_cm.long$ds1, levels = all_hclust_ladder$labels[all_hclust_ladder$order])
# all_feature_changes_scaled_3Params_M3C_k5_cm.long$ds2 <- factor(all_feature_changes_scaled_3Params_M3C_k5_cm.long$ds2, levels = all_hclust_ladder$labels[all_hclust_ladder$order])

face_oi <- ifelse(all_feature_changes[all_feature_changes_scaled_3Params_M3C_k5_cm_hclust$labels[all_feature_changes_scaled_3Params_M3C_k5_cm_hclust$order], "endoSymb_status"]==1, "bold", "plain")
all_feature_changes_scaled_3Params_M3C_k5_cm_plot <- ggplot(data=all_feature_changes_scaled_3Params_M3C_k5_cm.long, mapping=aes(x = ds1, y = ds2, fill=value)) + geom_tile()
all_feature_changes_scaled_3Params_M3C_k5_cm_plot <- all_feature_changes_scaled_3Params_M3C_k5_cm_plot + scale_x_discrete(name = "datasets") + scale_y_discrete(name = "datasets", position = "right") + scale_fill_gradient(low = "lightgreen", high = "darkgreen", guide = guide_colorbar(title = "co-clustering frequency", title.position = "top", label.position = "bottom", title.vjust = 0.5, title.hjust = 0))# + scale_fill_viridis(direction = -1, guide = guide_colorbar(title = "co-clustering frequency", title.position = "top", label.position = "bottom", title.vjust = 0.5, title.hjust = 0))
all_feature_changes_scaled_3Params_M3C_k5_cm_plot <- all_feature_changes_scaled_3Params_M3C_k5_cm_plot + theme(axis.text.x = element_text(angle = 90, size = 4, hjust = 1, vjust = 0.5, face=face_oi),axis.text.y.right = element_text(hjust = 0.5, size = 4, face=face_oi), legend.position = "top", legend.direction = "horizontal", axis.title.y = element_blank())

# Heatmap of changes
##############################################
all_feature_changes_scaled.long <- melt(data = as.matrix(all_feature_changes_scaled), varnames = c("datasets", "features"))
all_feature_changes_scaled.long$datasets <- factor(x = all_feature_changes_scaled.long$datasets, levels = all_feature_changes_scaled_3Params_M3C_k5_cm_hclust$labels[all_feature_changes_scaled_3Params_M3C_k5_cm_hclust$order], ordered = T)
all_feature_changes_scaled.long$features <- factor(x = all_feature_changes_scaled.long$features, levels = c("GC", "gSize_log2", "cDensity", "CUB_log2", "rna_bLength_log2", "IGRpercentage", "pseudoFrac_log2"), ordered = T)

all_feature_changes_scaled_heatmap_plot <- ggplot(data = all_feature_changes_scaled.long, mapping = aes(x = features, y = datasets, fill=value)) + geom_tile()
all_feature_changes_scaled_heatmap_plot <- all_feature_changes_scaled_heatmap_plot + scale_color_gradient2(aesthetics = "fill", low = cbPalette[7], mid = "white", high = cbPalette[6], guide = guide_colorbar(title = "normalized differences", title.position = "top", label.position = "bottom", title.vjust = 0.75, title.hjust = 0.5)) + scale_x_discrete(labels=c("GC", "log2(genome size)", "coding density", "log2(CUB)", "log2(16s branch length)", "IGR length", "log2(pseudogene fraction)"))
all_feature_changes_scaled_heatmap_plot <- all_feature_changes_scaled_heatmap_plot + theme(axis.text.x = element_text(angle = 90, hjust=1, vjust=0.5), axis.text.y = element_blank(), legend.position = "top", legend.direction = "horizontal", axis.title.y = element_blank(), axis.ticks.y = element_blank(), axis.text.x.bottom = element_text(size = 8))

dendro_cc_changes_combined_plot <- all_feature_changes_scaled_3Params_M3C_k5_cm_dend_plot + all_feature_changes_scaled_3Params_M3C_k5_cm_plot + all_feature_changes_scaled_heatmap_plot + plot_layout(widths = c(3,10,2))

ggsave(filename = "all_feature_changes_3Params_M3C_k5_cm_ori.svg", path = jump_common_dir, plot = dendro_cc_changes_combined_plot, device = "svg", width = 12, height = 9, units = "in", dpi = 300)
ggsave(filename ="all_feature_changes_3Params_M3C_k5_cm_ori.png", path = jump_common_dir, plot = dendro_cc_changes_combined_plot, device = "png", width = 12, height = 9, units = "in", dpi = 300)
```

Here, we can see 5 clusters appearing with reasonable co-clustering and exclusivity. Interestingly, the endosymbionts do not co-cluster with each other in this analyses.
##### All data, based on PAC using diceR
```{r clustering_all_data_diceR_PAC_k28_all }
library(diceR)

all_feature_changes_scaled_3Params_diceR_k28 <- consensus_cluster(data = all_feature_changes_scaled_3Params, nk = 2:8, p.item = 0.8, reps = 100, algorithms = c("hc", "pam", "diana", "cmeans", "km", "block", "som", "nmf"), scale = F, progress = T)
all_feature_changes_scaled_3Params_diceR_k28_indices <- consensus_evaluate(all_feature_changes_scaled_3Params, all_feature_changes_scaled_3Params_diceR_k28)

all_feature_changes_scaled_3Params_diceR_k28_indices$pac
```

```{r clustering_all_data_diceR_PAC_k2 }
all_feature_changes_scaled_3Params_diceR_k2 <- consensus_cluster(data = all_feature_changes_scaled_3Params, nk = 2, p.item = 0.8, reps = 100, algorithms = c("hc", "pam", "cmeans", "som"), scale = F, progress = T)

all_feature_changes_scaled_3Params_diceR_k2_cc <- consensus_combine(all_feature_changes_scaled_3Params_diceR_k2, element = "class")
all_feature_changes_scaled_3Params_diceR_k2_cc_cm <- consensus_matrix(all_feature_changes_scaled_3Params_diceR_k2_cc)
rownames(all_feature_changes_scaled_3Params_diceR_k2_cc_cm) <- rownames(all_feature_changes_scaled_3Params)
colnames(all_feature_changes_scaled_3Params_diceR_k2_cc_cm) <- rownames(all_feature_changes_scaled_3Params)

all_feature_changes_scaled_3Params_diceR_k2_cc_cm_hclust <- as.hclust(ladderize(as.dendrogram(hclust(d = dist(all_feature_changes_scaled_3Params_diceR_k2_cc_cm), method = "ward.D2"))))
all_feature_changes_scaled_3Params_diceR_k2_cc_cm_dend <- as.dendrogram(all_feature_changes_scaled_3Params_diceR_k2_cc_cm_hclust)
all_feature_changes_scaled_3Params_diceR_k2_cc_cm_dend <- as.ggdend(all_feature_changes_scaled_3Params_diceR_k2_cc_cm_dend)

all_feature_changes_scaled_3Params_diceR_k2_cc_cm_dend_plot <- ggplot(all_feature_changes_scaled_3Params_diceR_k2_cc_cm_dend, horiz = T, labels = F, theme = NULL, size=0.2) + scale_y_reverse() + theme_cowplot() + theme(panel.border = element_blank(), plot.margin = unit(x = c(0,0,0,0), units = "in"), axis.line = element_blank(), axis.title = element_blank(), axis.ticks = element_blank(), axis.text = element_blank())

all_feature_changes_scaled_3Params_diceR_k2_cc_cm.long <- melt(data = all_feature_changes_scaled_3Params_diceR_k2_cc_cm, varnames=c("ds1", "ds2"))
all_feature_changes_scaled_3Params_diceR_k2_cc_cm.long$ds1 <-  factor(all_feature_changes_scaled_3Params_diceR_k2_cc_cm.long$ds1, levels = all_feature_changes_scaled_3Params_diceR_k2_cc_cm_hclust$labels[all_feature_changes_scaled_3Params_diceR_k2_cc_cm_hclust$order], ordered = T)
all_feature_changes_scaled_3Params_diceR_k2_cc_cm.long$ds2 <- factor(all_feature_changes_scaled_3Params_diceR_k2_cc_cm.long$ds2, levels = all_feature_changes_scaled_3Params_diceR_k2_cc_cm_hclust$labels[all_feature_changes_scaled_3Params_diceR_k2_cc_cm_hclust$order], ordered = T)

face_oi <- ifelse(all_feature_changes[all_feature_changes_scaled_3Params_diceR_k2_cc_cm_hclust$labels[all_feature_changes_scaled_3Params_diceR_k2_cc_cm_hclust$order], "endoSymb_status"]==1, "bold", "plain")
all_feature_changes_scaled_3Params_diceR_k2_cc_cm_plot <- ggplot(data=all_feature_changes_scaled_3Params_diceR_k2_cc_cm.long, mapping=aes(x = ds1, y = ds2, fill=value)) + geom_tile()
all_feature_changes_scaled_3Params_diceR_k2_cc_cm_plot <- all_feature_changes_scaled_3Params_diceR_k2_cc_cm_plot + scale_x_discrete(name = "datasets") + scale_y_discrete(name = "datasets") + scale_fill_gradient(low = "lightgreen", high = "darkgreen", guide = guide_colorbar(title = "co-clustering frequency", title.position = "top", label.position = "bottom", title.vjust = 0.5, title.hjust = 0))# + scale_fill_viridis(direction = -1, guide = guide_colorbar(title = "co-clustering frequency", title.position = "top", label.position = "bottom", title.vjust = 0.5, title.hjust = 0))
all_feature_changes_scaled_3Params_diceR_k2_cc_cm_plot <- all_feature_changes_scaled_3Params_diceR_k2_cc_cm_plot + theme(axis.text.x = element_text(angle = 90, size = 6, hjust = 1, vjust = 0.5, face=face_oi),axis.text.y = element_text(size = 6, face=face_oi), legend.position = "top", legend.direction = "horizontal", axis.title.y = element_blank())

# Heatmap of changes
##############################################
all_feature_changes_scaled.long <- melt(data = as.matrix(all_feature_changes_scaled), varnames = c("datasets", "features"))
all_feature_changes_scaled.long$datasets <- factor(x = all_feature_changes_scaled.long$datasets, levels = all_feature_changes_scaled_3Params_diceR_k2_cc_cm_hclust$labels[all_feature_changes_scaled_3Params_diceR_k2_cc_cm_hclust$order], ordered = T) #[all_feature_changes_scaled_3Params_diceR_k2_cc_cm_hclust$order]
all_feature_changes_scaled.long$features <- factor(x = all_feature_changes_scaled.long$features, levels = c("GC", "gSize_log2", "cDensity", "CUB_log2", "rna_bLength_log2", "IGRpercentage", "pseudoFrac_log2"), ordered = T)

all_feature_changes_scaled_heatmap_plot <- ggplot(data = all_feature_changes_scaled.long, mapping = aes(x = features, y = datasets, fill=value)) + geom_tile()
all_feature_changes_scaled_heatmap_plot <- all_feature_changes_scaled_heatmap_plot + scale_color_gradient2(aesthetics = "fill", low = cbPalette[7], mid = "white", high = cbPalette[6], guide = guide_colorbar(title = "normalized differences", title.position = "top", label.position = "bottom", title.vjust = 0.75, title.hjust = 0.5)) + scale_x_discrete(labels=c("GC", "log2(genome size)", "coding density", "log2(CUB)", "log2(16s branch length)", "IGR length", "log2(pseudogene fraction)"))
all_feature_changes_scaled_heatmap_plot <- all_feature_changes_scaled_heatmap_plot + theme(axis.text.x = element_text(angle = 90, hjust=1, vjust=0.5), axis.text.y = element_blank(), legend.position = "top", legend.direction = "horizontal", axis.title.y = element_blank(), axis.ticks.y = element_blank(), axis.text.x.bottom = element_text(size = 8))

dendro_cc_changes_combined_plot <- all_feature_changes_scaled_3Params_diceR_k2_cc_cm_dend_plot + all_feature_changes_scaled_3Params_diceR_k2_cc_cm_plot + all_feature_changes_scaled_heatmap_plot + plot_layout(widths = c(3,10,2))

ggsave(filename = "all_feature_changes_3Params_diceR_k2_cm.svg", path = jump_common_dir, plot = dendro_cc_changes_combined_plot, device = "svg", width = 12, height = 9, units = "in", dpi = 300)
ggsave(filename ="all_feature_changes_3Params_diceR_k2_cm.png", path = jump_common_dir, plot = dendro_cc_changes_combined_plot, device = "png", width = 12, height = 9, units = "in", dpi = 300)
```
 
In the image obtained, the clusters are arranged differently from the previous clusterings. Specifically, a cluster of endosymbiotns and other datasets, gets aranged in the middle. This can be fixed by rotating one cluster to make it comparable to the previous clusterings.
```{r rearrange_cluster }
all_feature_changes_scaled_3Params_diceR_k2_cc_cm_hclust <- as.hclust(ladderize(as.dendrogram(hclust(d = dist(all_feature_changes_scaled_3Params_diceR_k2_cc_cm), method = "ward.D2"))))
which(all_feature_changes_scaled_3Params_diceR_k2_cc_cm_hclust$labels[all_feature_changes_scaled_3Params_diceR_k2_cc_cm_hclust$order]=="Alpha1_jump6")

library(dendextend)
all_feature_changes_scaled_3Params_diceR_k2_cc_cm_hclust <- dendextend::rotate(x = all_feature_changes_scaled_3Params_diceR_k2_cc_cm_hclust, order = all_feature_changes_scaled_3Params_diceR_k2_cc_cm_hclust$labels[all_feature_changes_scaled_3Params_diceR_k2_cc_cm_hclust$order][c(108:192, 84:96, 1:83, 97:107)])
all_feature_changes_scaled_3Params_diceR_k2_cc_cm_dend <- as.dendrogram(all_feature_changes_scaled_3Params_diceR_k2_cc_cm_hclust) %>% set("branches_lwd", 0.1)
all_feature_changes_scaled_3Params_diceR_k2_cc_cm_dend <- as.ggdend(all_feature_changes_scaled_3Params_diceR_k2_cc_cm_dend)

all_feature_changes_scaled_3Params_diceR_k2_cc_cm_dend_plot <- ggplot(all_feature_changes_scaled_3Params_diceR_k2_cc_cm_dend, horiz = T, labels = F, theme = NULL, size=0.2) + scale_y_reverse() + theme_cowplot() + theme(panel.border = element_blank(), plot.margin = unit(x = c(0,0,0,0), units = "in"), axis.line = element_blank(), axis.title = element_blank(), axis.ticks = element_blank(), axis.text = element_blank())

all_feature_changes_scaled_3Params_diceR_k2_cc_cm.long <- melt(data = all_feature_changes_scaled_3Params_diceR_k2_cc_cm, varnames=c("ds1", "ds2"))
all_feature_changes_scaled_3Params_diceR_k2_cc_cm.long$ds1 <-  factor(all_feature_changes_scaled_3Params_diceR_k2_cc_cm.long$ds1, levels = all_feature_changes_scaled_3Params_diceR_k2_cc_cm_hclust$labels[all_feature_changes_scaled_3Params_diceR_k2_cc_cm_hclust$order], ordered = T)
all_feature_changes_scaled_3Params_diceR_k2_cc_cm.long$ds2 <- factor(all_feature_changes_scaled_3Params_diceR_k2_cc_cm.long$ds2, levels = all_feature_changes_scaled_3Params_diceR_k2_cc_cm_hclust$labels[all_feature_changes_scaled_3Params_diceR_k2_cc_cm_hclust$order], ordered = T)

face_oi <- ifelse(all_feature_changes[all_feature_changes_scaled_3Params_diceR_k2_cc_cm_hclust$labels[all_feature_changes_scaled_3Params_diceR_k2_cc_cm_hclust$order], "endoSymb_status"]==1, "bold", "plain")
all_feature_changes_scaled_3Params_diceR_k2_cc_cm_plot <- ggplot(data=all_feature_changes_scaled_3Params_diceR_k2_cc_cm.long, mapping=aes(x = ds1, y = ds2, fill=value)) + geom_tile()
all_feature_changes_scaled_3Params_diceR_k2_cc_cm_plot <- all_feature_changes_scaled_3Params_diceR_k2_cc_cm_plot + scale_x_discrete(name = "datasets") + scale_y_discrete(name = "datasets") + scale_fill_gradient(low = "lightgreen", high = "darkgreen", guide = guide_colorbar(title = "co-clustering frequency", title.position = "top", label.position = "bottom", title.vjust = 0.5, title.hjust = 0))# + scale_fill_viridis(direction = -1, guide = guide_colorbar(title = "co-clustering frequency", title.position = "top", label.position = "bottom", title.vjust = 0.5, title.hjust = 0))
all_feature_changes_scaled_3Params_diceR_k2_cc_cm_plot <- all_feature_changes_scaled_3Params_diceR_k2_cc_cm_plot + theme(axis.text.x = element_text(angle = 90, size = 4, hjust = 1, vjust = 0.5, face=face_oi),axis.text.y = element_text(size = 4, face=face_oi), legend.position = "top", legend.direction = "horizontal", axis.title.y = element_blank())

# Heatmap of changes
##############################################
all_feature_changes_scaled.long <- melt(data = as.matrix(all_feature_changes_scaled), varnames = c("datasets", "features"))
all_feature_changes_scaled.long$datasets <- factor(x = all_feature_changes_scaled.long$datasets, levels = all_feature_changes_scaled_3Params_diceR_k2_cc_cm_hclust$labels[all_feature_changes_scaled_3Params_diceR_k2_cc_cm_hclust$order], ordered = T)
all_feature_changes_scaled.long$features <- factor(x = all_feature_changes_scaled.long$features, levels = c("GC", "gSize_log2", "cDensity", "CUB_log2", "rna_bLength_log2", "IGRpercentage", "pseudoFrac_log2"), ordered = T)

all_feature_changes_scaled_heatmap_plot <- ggplot(data = all_feature_changes_scaled.long, mapping = aes(x = features, y = datasets, fill=value)) + geom_tile()
all_feature_changes_scaled_heatmap_plot <- all_feature_changes_scaled_heatmap_plot + scale_color_gradient2(aesthetics = "fill", low = cbPalette[7], mid = "white", high = cbPalette[6], guide = guide_colorbar(title = "normalized differences", title.position = "top", label.position = "bottom", title.vjust = 0.75, title.hjust = 0.5)) + scale_x_discrete(labels=c("GC", "log2(genome size)", "coding density", "log2(CUB)", "log2(16s branch length)", "IGR length", "log2(pseudogene fraction)"))
all_feature_changes_scaled_heatmap_plot <- all_feature_changes_scaled_heatmap_plot + theme(axis.text.x = element_text(angle = 90, hjust=1, vjust=0.5), axis.text.y = element_blank(), legend.position = "top", legend.direction = "horizontal", axis.title.y = element_blank(), axis.ticks.y = element_blank(), axis.text.x.bottom = element_text(size = 8))

dendro_cc_changes_combined_plot <- all_feature_changes_scaled_3Params_diceR_k2_cc_cm_dend_plot + all_feature_changes_scaled_3Params_diceR_k2_cc_cm_plot + all_feature_changes_scaled_heatmap_plot + plot_layout(widths = c(3,10,2))

ggsave(filename = "all_feature_changes_3Params_diceR_k2_cmA.svg", path = jump_common_dir, plot = dendro_cc_changes_combined_plot, device = "svg", width = 12, height = 9, units = "in", dpi = 300)
ggsave(filename ="all_feature_changes_3Params_diceR_k2_cmA.png", path = jump_common_dir, plot = dendro_cc_changes_combined_plot, device = "png", width = 12, height = 9, units = "in", dpi = 300)
```

### Discretized data clustering
```{r discretize_data }
all_feature_changes_discretized <- all_feature_changes_scaled
for(col_oi in c("GC", "gSize_log2", "cDensity", "CUB_log2", "rna_bLength_log2", "IGRpercentage", "pseudoFrac_log2")){
  lower_median <- median(all_feature_changes_scaled[all_feature_changes_scaled[,col_oi]<0,col_oi], na.rm = T)
  upper_median <- median(all_feature_changes_scaled[all_feature_changes_scaled[,col_oi]>0,col_oi], na.rm = T)
  for(row_n in 1:nrow(all_feature_changes_discretized)){
    if(is.na(all_feature_changes_scaled[row_n,col_oi])){
      all_feature_changes_discretized[row_n,col_oi] <- NA
    }else if(all_feature_changes_scaled[row_n,col_oi] < lower_median){
      all_feature_changes_discretized[row_n,col_oi] <- -1
    } else if(all_feature_changes_scaled[row_n,col_oi] > upper_median){
      all_feature_changes_discretized[row_n,col_oi] <- 1
    }else{
      all_feature_changes_discretized[row_n,col_oi] <- 0
    }
  }
}

all_feature_changes_discretized_noNA <- 
all_feature_changes_discretized[!apply(is.na(all_feature_changes_discretized), 1, any),] #remove rows with any NAs
all_feature_changes_discretized_noNA <- as.data.frame(all_feature_changes_discretized_noNA)

all_feature_changes_discretized_3Params <- all_feature_changes_discretized[,c("GC", "gSize_log2", "cDensity")]
```

##### Complete discretized data, based on PAC using diceR
```{r clustering_discretized_noNA_data_diceR_PAC_k2_all }
library(diceR)

all_feature_changes_discretized_noNA_cc <- consensus_cluster(data = all_feature_changes_discretized_noNA, nk = 2, p.item = 0.8, reps = 100, algorithms = c("hc", "pam", "diana", "cmeans", "km", "block", "som"), scale = F, progress = T, distance = "manhattan")
all_feature_changes_discretized_noNA_ccomb_class_indices <- consensus_evaluate(all_feature_changes_discretized_noNA, all_feature_changes_discretized_noNA_cc)

all_feature_changes_discretized_noNA_ccomb_class_indices$pac
```

Since we have discretized the data, 2 clusters will not be sufficient for clustering, hence the high ambiguity. Let me try to discover cluster number.

##### Complete discretized data, based on MC sub-sampling
```{r clustering_discretized_noNA_data_M3C_k28_all }
library(M3C)
all_feature_changes_discretized_noNA_M3C_k28 <- M3C(mydata = t(all_feature_changes_discretized_noNA), cores = 3, iters = 100, maxK = 8)
```

We find that the optimal k=6 based on PAC, RCSI, and p-values. As expected, one cluster refers to endosymbionts and another to non-endosymbionts.

```{r clustering_discretized_noNA_data_M3C_k28 }
all_feature_changes_discretized_noNA_M3C_k2_cm <- all_feature_changes_discretized_noNA_M3C_k28$realdataresults[[6]]$consensus_matrix
rownames(all_feature_changes_discretized_noNA_M3C_k2_cm) <- rownames(all_feature_changes_discretized_noNA)[as.integer(colnames(all_feature_changes_discretized_noNA_M3C_k2_cm))]
colnames(all_feature_changes_discretized_noNA_M3C_k2_cm) <- rownames(all_feature_changes_discretized_noNA)[as.integer(colnames(all_feature_changes_discretized_noNA_M3C_k2_cm))]

all_feature_changes_discretized_noNA_M3C_k2_cm_hclust <- as.hclust(ladderize(as.dendrogram(hclust(d = dist(all_feature_changes_discretized_noNA_M3C_k2_cm), method = "ward.D2"))))
all_feature_changes_discretized_noNA_M3C_k2_cm_dend <- as.dendrogram(all_feature_changes_discretized_noNA_M3C_k2_cm_hclust)
all_feature_changes_discretized_noNA_M3C_k2_cm_dend <- as.ggdend(all_feature_changes_discretized_noNA_M3C_k2_cm_dend)
all_feature_changes_discretized_noNA_M3C_k2_cm_dend_plot <- ggplot(all_feature_changes_discretized_noNA_M3C_k2_cm_dend, horiz = T, labels = F, theme = NULL) + scale_y_reverse() + theme_cowplot() + theme(panel.border = element_blank(), plot.margin = unit(x = c(0,0,0,0), units = "in"), axis.line = element_blank(), axis.title = element_blank(), axis.ticks = element_blank(), axis.text = element_blank())

all_feature_changes_discretized_noNA_M3C_k2_cm.long <- melt(data = all_feature_changes_discretized_noNA_M3C_k2_cm, varnames=c("ds1", "ds2"))
all_feature_changes_discretized_noNA_M3C_k2_cm.long$ds1 <-  factor(all_feature_changes_discretized_noNA_M3C_k2_cm.long$ds1, levels = all_feature_changes_discretized_noNA_M3C_k2_cm_hclust$labels[all_feature_changes_discretized_noNA_M3C_k2_cm_hclust$order], ordered = T)
all_feature_changes_discretized_noNA_M3C_k2_cm.long$ds2 <- factor(all_feature_changes_discretized_noNA_M3C_k2_cm.long$ds2, levels = all_feature_changes_discretized_noNA_M3C_k2_cm_hclust$labels[all_feature_changes_discretized_noNA_M3C_k2_cm_hclust$order], ordered = T)

# all_feature_changes_discretized_noNA_M3C_k2_cm.long$ds1 <-  factor(all_feature_changes_discretized_noNA_M3C_k2_cm.long$ds1, levels = all_hclust_ladder$labels[all_hclust_ladder$order])
# all_feature_changes_discretized_noNA_M3C_k2_cm.long$ds2 <- factor(all_feature_changes_discretized_noNA_M3C_k2_cm.long$ds2, levels = all_hclust_ladder$labels[all_hclust_ladder$order])

all_feature_changes_discretized_noNA_M3C_k2_cm_plot <- ggplot(data=all_feature_changes_discretized_noNA_M3C_k2_cm.long, mapping=aes(x = ds1, y = ds2, fill=value)) + geom_tile()
all_feature_changes_discretized_noNA_M3C_k2_cm_plot <- all_feature_changes_discretized_noNA_M3C_k2_cm_plot + scale_x_discrete(name = "datasets") + scale_y_discrete(name = "datasets") + scale_fill_gradient(low = "lightgreen", high = "darkgreen", guide = guide_colorbar(title = "co-clustering frequency", title.position = "top", label.position = "bottom", title.vjust = 0.5, title.hjust = 0))# + scale_fill_viridis(direction = -1, guide = guide_colorbar(title = "co-clustering frequency", title.position = "top", label.position = "bottom", title.vjust = 0.5, title.hjust = 0))
face_oi <- ifelse(all_feature_changes[all_feature_changes_discretized_noNA_M3C_k2_cm_hclust$labels[all_feature_changes_discretized_noNA_M3C_k2_cm_hclust$order], "endoSymb_status"]==1, "bold", "plain")
all_feature_changes_discretized_noNA_M3C_k2_cm_plot <- all_feature_changes_discretized_noNA_M3C_k2_cm_plot + theme(axis.text.x = element_text(angle = 90, size = 6, hjust = 1, vjust = 0.5, face = face_oi),axis.text.y = element_text(size = 6, face =  face_oi), legend.position = "top", legend.direction = "horizontal", axis.title.y = element_blank())

# Heatmap of changes
##############################################
all_feature_changes_discretized_noNA.long <- melt(data = as.matrix(all_feature_changes_discretized_noNA), varnames = c("datasets", "features"))
all_feature_changes_discretized_noNA.long$datasets <- factor(x = all_feature_changes_discretized_noNA.long$datasets, levels = all_feature_changes_discretized_noNA_M3C_k2_cm_hclust$labels[all_feature_changes_discretized_noNA_M3C_k2_cm_hclust$order], ordered = T)
all_feature_changes_discretized_noNA.long$features <- factor(x = all_feature_changes_discretized_noNA.long$features, levels = c("GC", "gSize_log2", "cDensity", "CUB_log2", "rna_bLength_log2", "IGRpercentage", "pseudoFrac_log2"), ordered = T)

all_feature_changes_discretized_noNA_heatmap_plot <- ggplot(data = all_feature_changes_discretized_noNA.long, mapping = aes(x = features, y = datasets, fill=value)) + geom_tile()
all_feature_changes_discretized_noNA_heatmap_plot <- all_feature_changes_discretized_noNA_heatmap_plot + scale_color_gradient2(aesthetics = "fill", low = cbPalette[7], mid = "white", high = cbPalette[6], guide = guide_colorbar(title = "normalized differences", title.position = "top", label.position = "bottom", title.vjust = 0.75, title.hjust = 0.5)) + scale_x_discrete(labels=c("GC", "log2(genome size)", "coding density", "log2(CUB)", "log2(16s branch length)", "IGR length", "log2(pseudogene fraction)"))
all_feature_changes_discretized_noNA_heatmap_plot <- all_feature_changes_discretized_noNA_heatmap_plot + theme(axis.text.x = element_text(angle = 90, hjust=1, vjust=0.5), axis.text.y = element_blank(), legend.position = "top", legend.direction = "horizontal", axis.title.y = element_blank(), axis.ticks.y = element_blank(), axis.text.x.bottom = element_text(size = 8))

dendro_cc_changes_combined_plot <- all_feature_changes_discretized_noNA_M3C_k2_cm_dend_plot + all_feature_changes_discretized_noNA_M3C_k2_cm_plot + all_feature_changes_discretized_noNA_heatmap_plot + plot_layout(widths = c(3,10,2))

ggsave(filename = "all_feature_changes_discretized_noNA_M3C_k2_cm.svg", path = jump_common_dir, plot = dendro_cc_changes_combined_plot, device = "svg", width = 12, height = 9, units = "in", dpi = 300)
ggsave(filename ="all_feature_changes_discretized_noNA_M3C_k2_cm.png", path = jump_common_dir, plot = dendro_cc_changes_combined_plot, device = "png", width = 12, height = 9, units = "in", dpi = 300)
```

```{r clustering_discretized_all_data_M3C_k28_all }
library(M3C)
all_feature_changes_discretized_3Params_M3C_k28 <- M3C(mydata = t(all_feature_changes_discretized_3Params), cores = 3, iters = 100, maxK = 8)
```

```{r clustering_discretized_all_data_M3C_k28 }
dist_all_features_scaled <- dist(all_feature_changes_scaled_noNA, method = "euclidean")
hclust_all_features_scaled <- hclust(dist_all_features_scaled, method = "ward.D2")
vdist_all_features_scaled <- dist(t(all_feature_changes_scaled_noNA), method = "euclidean")
vclust_all_features_scaled <- hclust(vdist_all_features_scaled, method = "ward.D2")

all_feature_changes_scaled_noNA.long <- melt(data = as.matrix(all_feature_changes_scaled_noNA), varnames = c("datasets", "params"))
all_feature_changes_scaled_noNA.long$datasets <- factor(x = all_feature_changes_scaled_noNA.long$datasets, levels = all_feature_changes_scaled_noNA.long$datasets[hclust_all_features_scaled$order], ordered = T)

# all_feature_changes_scaled_noNA.long$params <- factor(x = all_feature_changes_scaled_noNA.long$params, levels = vclust_all_features_scaled$labels[vclust_all_features_scaled$order], ordered = T)

all_feature_changes_scaled_noNA_heatmap_plot <- ggplot(data = all_feature_changes_scaled_noNA.long, mapping = aes(x = params, y = datasets, fill=value)) + geom_tile()
all_feature_changes_scaled_noNA_heatmap_plot <- all_feature_changes_scaled_noNA_heatmap_plot + scale_color_gradient2(aesthetics = "fill", low = cbPalette[7], mid = "white", high = cbPalette[6])
all_feature_changes_scaled_noNA_heatmap_plot <- all_feature_changes_scaled_noNA_heatmap_plot + theme(axis.text.x = element_text(angle = 90, hjust=1, vjust=0.5), axis.text.y = element_text(hjust=1, vjust=0.5, size = 8))

# all_feature_changes_scaled_noNA_heatmap_plot <- all_feature_changes_scaled_noNA_heatmap_plot + scale_x_discrete(labels=c(expression(Delta~"GC3"), "log2(genome size ratio)", "log2(CUB ratio)", expression(Delta~"(coding density)"), "log2(pseudogene fraction ratio)", "log2(16s branch lengths ratio)", expression(Delta~"(IGR length)")))

library(ggdendro)
dendro_hclust <- ggdendrogram(hclust_all_features_scaled, rotate = T, leaf_labels = F)
heatmap_dendro <- plot_grid(all_feature_changes_scaled_noNA_heatmap_plot, dendro_hclust, nrow = 1, align = "hv")

ggsave(filename = paste0(jump_common_dir, "all_feature_changes_scaled_noNA_heatmap.svg"), plot = heatmap_dendro, width = 10, height=25, units = "in", dpi = 300)
```

In this dataset, we obtain 3 major clusters: 1. datasets where GC, genome, size, and CUB and/or coding density decrease. 2. datasets where GC decreased, but there was not much change in other params including genome size (or data was missing); 3a. datasets where GC increased, but not much change in other parameters; 3b. datasets where GC increased, not much change in genome size, but CUB or coding density decreased.

I wonder whether missing data creates problems in clustering. Specifically, data for changes in 16s branch lengths is missing for 128 datasets. I will perform clustering after removing 16s branch lengths.

#### Clustering without 16s branch lengths
```{r clustering_wo_16s }
all_feature_changes_scaled2 <- all_feature_changes_scaled[,-3]

dist_all_features_scaled2 <- dist(all_feature_changes_scaled2, method = "eucl")
hclust_all_features_scaled2 <- hclust(dist_all_features_scaled2, method = "ward.D2")
vdist_all_features_scaled2 <- dist(t(all_feature_changes_scaled2), method = "eucl")
vclust_all_features_scaled2 <- hclust(vdist_all_features_scaled2, method = "ward.D2")

all_feature_changes_scaled2.long <- melt(data = as.matrix(all_feature_changes_scaled2), varnames = c("datasets", "params"))
all_feature_changes_scaled2.long$datasets <- factor(x = all_feature_changes_scaled2.long$datasets, levels = all_feature_changes_scaled2.long$datasets[hclust_all_features_scaled2$order], ordered = T)

all_feature_changes_scaled2.long$params <- factor(x =
all_feature_changes_scaled2.long$params, levels = colnames(all_feature_changes_scaled2)[vclust_all_features_scaled2$order], ordered = T)

all_feature_changes_scaled_noNA_heatmap_plot <- ggplot(data = all_feature_changes_scaled2.long, mapping = aes(x = params, y = datasets, fill=value)) + geom_tile()

all_feature_changes_scaled_noNA_heatmap_plot <- all_feature_changes_scaled_noNA_heatmap_plot + scale_color_gradient2(aesthetics = "fill")

all_feature_changes_scaled_noNA_heatmap_plot <- all_feature_changes_scaled_noNA_heatmap_plot + theme(axis.text.x = element_text(angle = 90, hjust=1, vjust=0.5), axis.text.y = element_text(hjust=1, vjust=0.5, size = 8))

# all_feature_changes_scaled_noNA_heatmap_plot <- all_feature_changes_scaled_noNA_heatmap_plot + scale_x_discrete(labels=c(expression(Delta~"GC3"), "log2(genome size ratio)", "log2(CUB ratio)", expression(Delta~"(coding density)"), "log2(pseudogene fraction ratio)", "log2(16s branch lengths ratio)", expression(Delta~"(IGR length)")))

library(ggdendro)
dendro_hclust <- ggdendrogram(hclust_all_features_scaled2, rotate = T, leaf_labels = F)
heatmap_dendro <- plot_grid(all_feature_changes_scaled_noNA_heatmap_plot, dendro_hclust, nrow = 1, align = "hv")

ggsave(filename = paste0(jump_common_dir, "all_feature_changes2a_scaled_heatmap.svg"), plot = heatmap_dendro, width = 10, height=25, units = "in", dpi = 300)
```
The clusters remain more or less the same. I will try a different strategy. I will cluster based only on those variables with complete data i.e. GC, genome size, and coding density, but will plot other variables without including them during clustering.
#### Clustering with only complete data
```{r clustering_w_complete_data }
all_feature_changes_scaled3 <- all_feature_changes_scaled[,c(1,4,7)]

dist_all_features_scaled3 <- dist(all_feature_changes_scaled3, method = "euclidean")
hclust_all_features_scaled3 <- hclust(dist_all_features_scaled3, method = "ward.D2")
vdist_all_features_scaled3 <- dist(t(all_feature_changes_scaled3), method = "euclidean")
vclust_all_features_scaled3 <- hclust(vdist_all_features_scaled3, method = "ward.D2")

all_feature_changes_scaled.long <- melt(data = as.matrix(all_feature_changes_scaled), varnames = c("datasets", "params"))
all_feature_changes_scaled.long$datasets <- factor(x = all_feature_changes_scaled.long$datasets, levels = hclust_all_features_scaled3$labels[hclust_all_features_scaled3$order], ordered = T)

all_feature_changes_scaled.long$params <- factor(x =
all_feature_changes_scaled.long$params, levels = c("GC", "gSize_log2", "cDensity", "CUB_log2", "pseudoFrac_log2", "IGRlength", "rna_bLength_log2"), ordered = T)

all_feature_changes_scaled_noNA_heatmap_plot <- ggplot(data = all_feature_changes_scaled.long, mapping = aes(x = params, y = datasets, fill=value)) + geom_tile()

all_feature_changes_scaled_noNA_heatmap_plot <- all_feature_changes_scaled_noNA_heatmap_plot + scale_color_gradient2(aesthetics = "fill")

all_feature_changes_scaled_noNA_heatmap_plot <- all_feature_changes_scaled_noNA_heatmap_plot + theme(axis.text.x = element_text(angle = 90, hjust=1, vjust=0.5), axis.text.y = element_text(hjust=1, vjust=0.5, size = 8))

all_feature_changes_scaled_noNA_heatmap_plot <- all_feature_changes_scaled_noNA_heatmap_plot + scale_x_discrete(labels=c(expression(Delta~"GC3"), "log2(genome size ratio)", expression(Delta~"(coding density)"), "log2(CUB ratio)", "log2(pseudogene fraction ratio)", expression(Delta~"(IGR length)"), "log2(16s branch ratio)"))

library(ggdendro)
dendro_hclust <- ggdendrogram(hclust_all_features_scaled3, rotate = T, leaf_labels = F)
heatmap_dendro <- plot_grid(all_feature_changes_scaled_noNA_heatmap_plot, dendro_hclust, nrow = 1, align = "hv")

ggsave(filename = paste0(jump_common_dir, "all_feature_changes3a_scaled_heatmap.svg"), plot = heatmap_dendro, width = 10, height=25, units = "in", dpi = 300)

```

I will try and plot different color scales for every parameter. This requires to separate each plot. But clustering is still based on the three complete parameters- GC, genome size, coding density.
#### Clustering with only complete data
```{r clustering_w_complete_data_sep_plots }
all_feature_changes_scaled3 <- all_feature_changes_scaled[,c(1,4,7)]

dist_all_features_scaled3 <- dist(all_feature_changes_scaled3, method = "euclidean")
hclust_all_features_scaled3 <- hclust(dist_all_features_scaled3, method = "ward.D2")
vdist_all_features_scaled3 <- dist(t(all_feature_changes_scaled3), method = "euclidean")
vclust_all_features_scaled3 <- hclust(vdist_all_features_scaled3, method = "ward.D2")

all_feature_changes_scaled.long <- melt(data = as.matrix(all_feature_changes_scaled), varnames = c("datasets", "params"))
all_feature_changes_scaled.long$datasets <- factor(x = all_feature_changes_scaled.long$datasets, levels = hclust_all_features_scaled3$labels[hclust_all_features_scaled3$order], ordered = T)
# all_feature_changes_scaled.long$datasets <- factor(x = all_feature_changes_scaled.long$datasets, levels = rownames(all_feature_changes_scaled3)[order(all_feature_changes_scaled3[, "GC"])], ordered = T)

all_feature_changes_scaled.long$params <- factor(x =
all_feature_changes_scaled.long$params, levels = c("GC", "gSize_log2", "cDensity", "CUB_log2", "pseudoFrac_log2", "IGRlength", "rna_bLength_log2"), ordered = T)

labels_oi <- c(expression(Delta~"GC3"), "log2(genome size ratio)", expression(Delta~"(coding density)"), "log2(CUB ratio)", "log2(pseudogene fraction ratio)", expression(Delta~"(IGR length)"), "log2(16s branch ratio)")
names(labels_oi) <- c("GC", "gSize_log2", "cDensity", "CUB_log2", "pseudoFrac_log2", "IGRlength", "rna_bLength_log2")

sep_plots <- lapply(names(labels_oi), function(pp){
  all_feature_changes_scaled_noNA_heatmap_plot <- ggplot(data = filter(all_feature_changes_scaled.long, params==pp), mapping = aes(x = params, y = datasets, fill=value)) + geom_tile()

  all_feature_changes_scaled_noNA_heatmap_plot <- all_feature_changes_scaled_noNA_heatmap_plot + scale_color_gradient2(aesthetics = "fill")#, guide = guide_legend(title.position = "left", title.theme = element_text(angle = 90)), name=labels_oi[pp])

  all_feature_changes_scaled_noNA_heatmap_plot <- all_feature_changes_scaled_noNA_heatmap_plot + theme(axis.text.x = element_blank(), axis.title.y = element_blank(), axis.line.y = element_blank(), axis.ticks.y = element_blank() , axis.text.y = element_blank(), axis.title.x = element_blank(), legend.position = "bottom", legend.direction = "vertical", plot.margin = unit(c(0,0,10,0), "pt"))

  # all_feature_changes_scaled_noNA_heatmap_plot <- all_feature_changes_scaled_noNA_heatmap_plot + scale_x_discrete(name="", labels=labels_oi[pp])
  all_feature_changes_scaled_noNA_heatmap_plot  
})
sep_plots[["nrow"]] <- 1

library(ggdendro)
dendro_hclust <- ggdendrogram(hclust_all_features_scaled3, rotate = T, leaf_labels = F)

# heatmap_dendro <- plot_grid(sep_plots, nrow = 1, align = "hv")

library(gridExtra)
heatmap_dendro <- do.call(grid.arrange, sep_plots)
final_plot <- plot_grid(heatmap_dendro, dendro_hclust, nrow = 1, rel_widths = c(5,5))
  
ggsave(filename = paste0(jump_common_dir, "all_feature_changes5a_scaled_heatmap.svg"), plot = final_plot, width = 20, height=25, units = "in", dpi = 300)

for(col_oi in colnames(all_feature_changes_scaled)){
  hist(all_feature_changes_scaled[,col_oi], breaks=20, main = col_oi)
  abline(v=sd(all_feature_changes_scaled[,col_oi], na.rm = T), col="red")
  abline(v=-sd(all_feature_changes_scaled[,col_oi], na.rm = T), col="red")
}
```

#### Discretizing scaled data
```{r categorizing_scale }
for(col_oi in colnames(all_feature_changes_scaled)){
  hist(all_feature_changes_scaled[,col_oi], breaks=30, main = col_oi)
  abline(v=sd(all_feature_changes_scaled[,col_oi], na.rm = T), col="red")
  abline(v=-sd(all_feature_changes_scaled[,col_oi], na.rm = T), col="red")
}

for(col_oi in colnames(all_feature_changes_scaled)){
    hist(all_feature_changes_scaled[,col_oi], breaks=20, main = col_oi)
    abline(v=median(all_feature_changes_scaled[,col_oi], na.rm = T), col="darkgreen")
    abline(v=IQR(all_feature_changes_scaled[,col_oi], na.rm = T), col="red")
    abline(v=-sd(all_feature_changes_scaled[,col_oi], na.rm = T), col="red")
    abline(v=IQR(all_feature_changes_scaled[,col_oi], na.rm = T), col="blue")
    abline(v=-IQR(all_feature_changes_scaled[,col_oi], na.rm = T), col="blue")
}

for(col_oi in colnames(all_feature_changes_scaled)){
    hist(all_feature_changes_scaled[,col_oi], breaks=20, main = col_oi)
    abline(v=median(all_feature_changes_scaled[,col_oi], na.rm = T), col="darkgreen")
    abline(v=quantile(all_feature_changes_scaled[all_feature_changes_scaled[,col_oi]<0,col_oi], probs = 0.75, na.rm = T), col="red")
    abline(v=quantile(all_feature_changes_scaled[all_feature_changes_scaled[,col_oi]>0,col_oi], probs = 0.25, na.rm = T), col="red")
}
```

#### deltaGC distribution
```{r deltaGC_distribution_plt }
downward_GCjumps_stats <- subset(all_feature_changes, `GC` < 0)
upward_GCjumps_stats <- subset(all_feature_changes,  `GC` > 0)

deltaGC_hist <- ggplot(all_feature_changes, aes(x =  `GC`, fill=as.factor( `GC`/abs(`GC`)))) + geom_histogram(binwidth = 2, na.rm = T, col="black") + ggtitle("Distribution of GC jump magnitudes")

deltaGC_hist <- deltaGC_hist + scale_x_continuous(name = expression(Delta*"GC (affected \u2013 unaffected)"), breaks = seq(-35,25,5), labels=seq(-35,25,5)) + scale_y_continuous(name="number of jumps") + scale_fill_grey(name=expression("direction"), labels=c(paste0("downward\n(n=", length(which(all_feature_changes$GC<0)), ")"), paste0("upward\n(n=", length(which(all_feature_changes$GC>0)), ")")), start = 0.2, end=0.8)

deltaGC_hist <- deltaGC_hist + theme_cowplot() + theme(legend.position = c(0.025,0.85), legend.title.align = 0.5)

deltaGC_hist <- deltaGC_hist + geom_segment(data = subset(all_feature_changes,  `GC` < 0),mapping = aes(x=median( `GC`), y=1, xend=median( `GC`), yend=0), arrow = arrow(type = "closed", length = unit(x = 1, units = "picas"), ends = "last")) + geom_segment(data = subset(all_feature_changes,  `GC` > 0),mapping = aes(x=median( `GC`), y=1, xend=median( `GC`), yend=0), arrow = arrow(type = "closed", length = unit(x = 1, units = "picas"), ends = "last"))

# deltaGC_hist <- deltaGC_hist + geom_segment(data = subset(all_genome_comparisons[setdiff(rownames(all_genome_comparisons), symbiont_clades),], focal_GC - sis_GC < 0),mapping = aes(x=quantile(focal_GC-sis_GC, 0.25), y=1, xend=quantile(focal_GC-sis_GC, 0.25), yend=0), arrow = arrow(type = "closed", length = unit(x = 1, units = "picas"), ends = "last")) + geom_segment(data = subset(all_genome_comparisons[setdiff(rownames(all_genome_comparisons), symbiont_clades),], focal_GC - sis_GC > 0),mapping = aes(x=quantile(focal_GC-sis_GC, 0.75), y=1, xend=quantile(focal_GC-sis_GC, 0.75), yend=0), arrow = arrow(type = "closed", length = unit(x = 1, units = "picas"), ends = "last"))

# save_plot(filename = paste0(jump_common_dir, "deltaGC_hist1.svg"), plot = deltaGC_hist, ncol = 1, base_width = 5.6, base_height = 4)
# save_plot(filename = paste0(jump_common_dir, "deltaGC_hist1.png"), plot = deltaGC_hist, ncol = 1, base_width = 5.6, base_height = 4)
# deltaGC_hist
# 
# deltaGC_hist <- deltaGC_hist + theme_cowplot() + scale_fill_grey(name=expression("direction"), labels=c(paste0("downward\n(n=", nrow(downward_GCjumps_stats), ")"), paste0("upward\n(n=", nrow(upward_GCjumps_stats), ")")))
# save_plot(filename = paste0(jump_common_dir, "deltaGC_hist1_grey.svg"), plot = deltaGC_hist, ncol = 1, base_width = 7.2, base_height = 4)
# save_plot(filename = paste0(jump_common_dir, "deltaGC_hist1_grey.png"), plot = deltaGC_hist, ncol = 1, base_width = 7.2, base_height = 4)
# deltaGC_hist

print("medians")
median(downward_GCjumps_stats$`GC`, na.rm = T)
median(upward_GCjumps_stats$`GC`, na.rm = T)

print("quartiles")
quantile(downward_GCjumps_stats$`GC`,0.25)
quantile(upward_GCjumps_stats$`GC`,0.75)

print("fraction above 10%")
length(which(downward_GCjumps_stats$`GC`< (-10)))/nrow(downward_GCjumps_stats)

length(which(upward_GCjumps_stats$`GC` > (10)))/nrow(upward_GCjumps_stats)

print("fraction above 15%")
length(which(downward_GCjumps_stats$`GC`< (-15)))/nrow(downward_GCjumps_stats)

length(which(upward_GCjumps_stats$`GC` > (15)))/nrow(upward_GCjumps_stats)
```

#### deltaGC vs GC
```{r deltaGC_vs_GC_plot }
bestFit_params <- read.table(paste0(levolution_data_dir, "levolution_best_fit_parameters_summary.txt"))
all_feature_changes$OLclade <- sapply(rownames(all_feature_changes), function(y) strsplit(x = y, split = "_")[[1]][1])
Jumps_num <- sapply(OLclades_oi, function(x) length(which(all_feature_changes$OLclade==x)))
downwardJumps_num <- sapply(OLclades_oi, function(x) length(which(all_feature_changes[which(all_feature_changes$OLclade==x),"GC"]<0)))
upwardJumps_num <- sapply(OLclades_oi, function(x) length(which(all_feature_changes[which(all_feature_changes$OLclade==x),"GC"]>0)))

deltaGC_vs_GC_df <- as.data.frame(matrix(data = c(bestFit_params[OLclades_oi,"root_state"], Jumps_num, upwardJumps_num,downwardJumps_num), ncol = 4, dimnames = list(OLclades_oi, c("ancestralGC", "jumps_num", "upJumps_num", "downJumps_num"))))

deltaGC_vs_GC_OLclades_plt <- ggplot(data = deltaGC_vs_GC_df, aes(x = ancestralGC, y = upJumps_num/(downJumps_num + upJumps_num))) + geom_point(color=grey(0.5)) + theme_cowplot() + ggtitle("Fraction of upward jumps \n in order-level clades") + theme(plot.title=element_text(hjust=0.5))
deltaGC_vs_GC_OLclades_plt <- deltaGC_vs_GC_OLclades_plt + scale_x_continuous(name="ancestral GC") + scale_y_continuous(name="fraction of upward jumps", limits = c(0,1))
deltaGC_vs_GC_OLclades_plt

all_features_df_valid$endoSymb_status <- factor(all_features_df_valid$endoSymb_status, levels=c(1,0))

deltaGC_vs_GC <- ggplot(data=all_features_df_valid, aes(x = sister_GC, y = delta_GC, color=endoSymb_status)) + geom_point() + theme_cowplot() + theme(legend.position=c(0.02,0.18), plot.title=element_text(hjust=0.5)) + ggtitle("Jump magnitude vs ancestral GC \n for each jump")
deltaGC_vs_GC <- deltaGC_vs_GC + scale_x_continuous(name="ancestral GC (unaffected)") + scale_y_continuous(name = expression(Delta*"GC (affected \u2013 unaffected)")) + scale_color_grey(start = 0, end = 0.5, name = "", labels=c("endosymbionts", "others"))
deltaGC_vs_GC <- deltaGC_vs_GC + geom_smooth(method = "lm", aes(x=sister_GC, y=delta_GC), data=subset(all_features_df_valid, endoSymb_status==0))
deltaGC_vs_GC <- deltaGC_vs_GC + stat_cor(data = subset(all_features_df_valid,endoSymb_status=0), label.y=c(35, 35), label.x=c(50,30)) + stat_regline_equation(data = subset(all_features_df_valid,endoSymb_status=0), label.y=c(30,30), label.x=c(50,30))

deltaGC_vs_GC_all <- deltaGC_vs_GC_OLclades_plt + deltaGC_hist + deltaGC_vs_GC + plot_annotation(tag_levels = "A") + plot_layout(ncol = 3) 
# deltaGC_vs_GC_all

ggsave2(filename = "deltaGC_hist_vs_GC_all.png", plot = deltaGC_vs_GC_all, path = jump_common_dir, width = 14, height = 4, units = "in", dpi = 300)
ggsave2(filename = "deltaGC_hist_vs_GC_all.svg", plot = deltaGC_vs_GC_all, path = jump_common_dir, width = 14, height = 4, units = "in", dpi = 300)
```

#### deltaGC distribution 2
```{r deltaGC_distribution_plt2 }
downward_GCjumps_stats <- subset(all_features_df_sel, `focal GC (%)` - `sister GC (%)` < 0)
upward_GCjumps_stats <- subset(all_features_df_sel,  `focal GC (%)` - `sister GC (%)` > 0)

deltaGC_hist <- ggplot(all_features_df_sel, aes(x =  `focal GC (%)` - `sister GC (%)`, fill=as.factor(( `focal GC (%)` - `sister GC (%)`)/abs( `focal GC (%)` - `sister GC (%)`)))) + geom_histogram(binwidth = 2, na.rm = T, col="black")

deltaGC_hist <- deltaGC_hist + scale_x_continuous(name = expression(Delta*"GC (focal - related)"), limits = c(-35,25), breaks = seq(-35,25,5), labels=seq(-35,25,5)) + scale_y_continuous(name="number of jumps") + scale_fill_discrete(name=expression("direction"), labels=c(paste0("downward\n(n=", nrow(downward_GCjumps_stats), ")"), paste0("upward\n(n=", nrow(upward_GCjumps_stats), ")")))

deltaGC_hist <- deltaGC_hist + theme(legend.position = c(0.025,0.85), legend.title.align = 0.5)

deltaGC_hist <- deltaGC_hist + geom_segment(data = subset(all_features_df_sel,  `focal GC (%)` - `sister GC (%)` < 0),mapping = aes(x=median( `focal GC (%)` - `sister GC (%)`), y=1, xend=median( `focal GC (%)` - `sister GC (%)`), yend=0), arrow = arrow(type = "closed", length = unit(x = 1, units = "picas"), ends = "last")) + geom_segment(data = subset(all_features_df_sel,  `focal GC (%)` - `sister GC (%)` > 0),mapping = aes(x=median( `focal GC (%)` - `sister GC (%)`), y=1, xend=median( `focal GC (%)` - `sister GC (%)`), yend=0), arrow = arrow(type = "closed", length = unit(x = 1, units = "picas"), ends = "last"))

# deltaGC_hist <- deltaGC_hist + geom_segment(data = subset(all_genome_comparisons[setdiff(rownames(all_genome_comparisons), symbiont_clades),], focal_GC - sis_GC < 0),mapping = aes(x=quantile(focal_GC-sis_GC, 0.25), y=1, xend=quantile(focal_GC-sis_GC, 0.25), yend=0), arrow = arrow(type = "closed", length = unit(x = 1, units = "picas"), ends = "last")) + geom_segment(data = subset(all_genome_comparisons[setdiff(rownames(all_genome_comparisons), symbiont_clades),], focal_GC - sis_GC > 0),mapping = plot_grid()aes(x=quantile(focal_GC-sis_GC, 0.75), y=1, xend=quantile(focal_GC-sis_GC, 0.75), yend=0), arrow = arrow(type = "closed", length = unit(x = 1, units = "picas"), ends = "last"))

save_plot(filename = paste0(jump_common_dir, "deltaGC_hist.svg"), plot = deltaGC_hist, ncol = 1, base_width = 5.6, base_height = 4)
save_plot(filename = paste0(jump_common_dir, "deltaGC_hist.png"), plot = deltaGC_hist, ncol = 1, base_width = 5.6, base_height = 4)
deltaGC_hist

print("medians")
median(downward_GCjumps_stats$`focal GC (%)` - downward_GCjumps_stats$`sister GC (%)`, na.rm = T)
median(upward_GCjumps_stats$`focal GC (%)` - upward_GCjumps_stats$`sister GC (%)`, na.rm = T)

print("quartiles")
quantile(downward_GCjumps_stats$ `focal GC (%)` - downward_GCjumps_stats$`sister GC (%)`,0.25)
quantile(upward_GCjumps_stats$`focal GC (%)` - upward_GCjumps_stats$`sister GC (%)`,0.75)

print("fraction above 10%")
length(which(downward_GCjumps_stats$`focal GC (%)` - downward_GCjumps_stats$`sister GC (%)`< (-10)))/nrow(downward_GCjumps_stats)

length(which(upward_GCjumps_stats$`focal GC (%)` - upward_GCjumps_stats$`sister GC (%)` > (10)))/nrow(upward_GCjumps_stats)

print("fraction above 15%")
length(which(downward_GCjumps_stats$`focal GC (%)` - downward_GCjumps_stats$`sister GC (%)`< (-15)))/nrow(downward_GCjumps_stats)

length(which(upward_GCjumps_stats$`focal GC (%)` - upward_GCjumps_stats$`sister GC (%)` > (15)))/nrow(upward_GCjumps_stats)
```

#### Focal vs sister correlation plots
```{r}
gSize_corr_plt <- ggplot(data = all_features_df_valid, mapping = aes(y = focal_genome_size_checkM, x = sister_genome_size_checkM)) + geom_point(mapping = aes(color=focal_GC-sister_GC, shape=endoSymb_status))
gSize_corr_plt <- gSize_corr_plt + scale_color_gradient2(low = "darkred", high = "darkblue") + scale_shape_manual(values = c(21,19))
gSize_corr_plt <- gSize_corr_plt + geom_abline(intercept = 0, slope = 1, color="grey", alpha=0.5)
gSize_corr_plt

cDensity_corr_plt <- ggplot(data = all_features_df_valid, mapping = aes(y = focal_coding_density, x = sister_coding_density)) + geom_point(mapping = aes(color=focal_GC-sister_GC, shape=endoSymb_status))
cDensity_corr_plt <- cDensity_corr_plt + scale_color_gradient2(low = "darkred", high = "darkblue") + scale_shape_manual(values = c(21,19))
cDensity_corr_plt <- cDensity_corr_plt + geom_abline(intercept = 0, slope = 1, color="grey", alpha=0.5)
cDensity_corr_plt

IGR_corr_plt <- ggplot(data = all_features_df_valid, mapping = aes(y = focal_IGR_length, x = sister_IGR_length)) + geom_point(mapping = aes(color=focal_GC-sister_GC, shape=endoSymb_status))
IGR_corr_plt <- IGR_corr_plt + scale_color_gradient2(low = "darkred", high = "darkblue") + scale_shape_manual(values = c(21,19)) + scale_y_continuous(limits = c(0,200))
IGR_corr_plt <- IGR_corr_plt + geom_abline(intercept = 0, slope = 1, color="grey", alpha=0.5)
IGR_corr_plt
```

#### dN/dS vs dS, focal and related clades
```{r dNdS_dS_focal_vs_sis }
all_yn00_dNdS_dS_deltaGC3 <- data.frame()
for(dataset_oi in rownames(all_features_df_sel)){
  if(abs(all_features_df_sel[dataset_oi,"deltaGC3"])<0.05) next
  print(dataset_oi)
  OLclade_oi <- strsplit(dataset_oi, "_")[[1]][1]
  jumpNum_oi <- strsplit(dataset_oi, "_jump")[[1]][2]
  
  jumpBranch_oi <- OLclade_jumpBranches[jumpNum_oi]
    
  jumpDir <- paste0(jump_common_dir, OLclade_oi, "_jump", jumpNum_oi, "/")
  cds_aln_dir <- paste0(jumpDir,"cds_alignments/")
  YN00_dir <- paste0(jumpDir,"YN00/")
    
  pairwise_yn00_est_focal <- read.table(file = paste0(YN00_dir, OLclade_oi,"_", paste0("jump", jumpNum_oi), "_HEG_YN00_estimates_focal.tab"), stringsAsFactors = F)
  pairwise_yn00_est_focal$deltaGC3 <- all_features_df_sel[dataset_oi,"deltaGC3"]  
  
  pairwise_yn00_est_related <- read.table(file = paste0(YN00_dir, OLclade_oi,"_", paste0("jump", jumpNum_oi), "_HEG_YN00_estimates_related.tab"), stringsAsFactors = F)
  pairwise_yn00_est_related$deltaGC3 <- all_features_df_sel[dataset_oi,"deltaGC3"]
  
  all_yn00_dNdS_dS_deltaGC3 <- rbind(all_yn00_dNdS_dS_deltaGC3, pairwise_yn00_est_focal, pairwise_yn00_est_related)
}

all_yn00_dNdS_dS_deltaGC3$full_dataset <- paste0(all_yn00_dNdS_dS_deltaGC3$dataset,"_", all_yn00_dNdS_dS_deltaGC3$parent)
all_yn00_dNdS_dS_deltaGC3 <- all_yn00_dNdS_dS_deltaGC3[complete.cases(all_yn00_dNdS_dS_deltaGC3),]

deltaGC3_df <- all_features_df_sel[unique(all_yn00_dNdS_dS_deltaGC3$full_dataset), c("change_in_dNdS", "deltaGC3")]
deltaGC3_df$full_dataset <- rownames(deltaGC3_df)
deltaGC3_df$figLabels <- LETTERS[1:nrow(deltaGC3_df)]

all_yn00_dNdS_dS_deltaGC3$full_dataset <- ordered(x = all_yn00_dNdS_dS_deltaGC3$full_dataset, levels = deltaGC3_df$full_dataset)
all_yn00_dNdS_dS_deltaGC3$category <- ordered(x = all_yn00_dNdS_dS_deltaGC3$category, levels = c("focal", "sister", "other"))

plt_yn00_dNdS_vs_dS <- ggplot(data = all_yn00_dNdS_dS_deltaGC3, mapping = aes(x = dS, y = dN.dS, color=category)) + geom_point(data = subset(all_yn00_dNdS_dS_deltaGC3, category=="other"), shape=21, alpha=0.7) + geom_point(data = subset(all_yn00_dNdS_dS_deltaGC3, category=="sister"), shape=21, alpha=0.7) + geom_point(data = subset(all_yn00_dNdS_dS_deltaGC3, category=="focal"), shape=21, alpha=0.9)
  
plt_yn00_dNdS_vs_dS <- plt_yn00_dNdS_vs_dS + facet_wrap(.~fct_reorder(full_dataset, deltaGC3), ncol = 3, scales = "fixed") + geom_text(data = deltaGC3_df, mapping = aes(x = 3.2, y = 0.092, label = "Delta", group=NULL), inherit.aes = F, parse = T) + geom_text(data =  deltaGC3_df, mapping = aes(x = 5, y = 0.092, label = paste("GC3 = ", deltaGC3), group=NULL), inherit.aes = F) + geom_text(data =  deltaGC3_df, mapping = aes(x = 0.2, y = 0.092, label = figLabels, group=NULL), inherit.aes = F, size=6)

plt_yn00_dNdS_vs_dS <- plt_yn00_dNdS_vs_dS + scale_y_continuous(name = "dN/dS", breaks = c(0, 0.025, 0.05, 0.075, 0.1), labels = c(0, 0.025, 0.05, 0.075, 0.1)) + scale_color_manual(values=c("orange", "black", "purple"))

plt_yn00_dNdS_vs_dS <- plt_yn00_dNdS_vs_dS + theme_bw() + theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(), panel.border = element_rect(colour = "black"), legend.position = c(0.3, 0.77), legend.justification = c(1, 0), legend.key.width = unit(x = 1, units = "line"), legend.title.align = 0, axis.title.x = element_text(size = 16), axis.title.y = element_text(size = 16), strip.text.x = element_text(size=12), legend.spacing = unit(1, "points"))

ggsave(filename = paste0(jump_common_dir, "YN00_HEG_dNdS_example_facets.svg"), plot = plt_yn00_dNdS_vs_dS, dpi = 300, width = 7.2, height = 7.2, units = "in")
ggsave(filename = paste0(jump_common_dir, "YN00_HEG_dNdS_example_facets.png"), plot = plt_yn00_dNdS_vs_dS, dpi = 300, width = 7.2, height = 7.2, units = "in")
```

#### dN/dS vs dS, focal and related clades
```{r dNdS_dS_focal_vs_sis_deltaGC3all }
all_yn00_dNdS_dS_deltaGC3 <- data.frame()
for(dataset_oi in rownames(all_features_df_sel)){
  if(abs(all_features_df_sel[dataset_oi,"deltaGC3"])>0.05 | is.na(all_features_df_sel[dataset_oi,"deltaGC3"])) next
  print(dataset_oi)
  OLclade_oi <- strsplit(dataset_oi, "_")[[1]][1]
  jumpNum_oi <- strsplit(dataset_oi, "_jump")[[1]][2]
  
  jumpBranch_oi <- OLclade_jumpBranches[jumpNum_oi]
    
  jumpDir <- paste0(jump_common_dir, OLclade_oi, "_jump", jumpNum_oi, "/")
  cds_aln_dir <- paste0(jumpDir,"cds_alignments/")
  YN00_dir <- paste0(jumpDir,"YN00/")
    
  f_out <- tryCatch({
      pairwise_yn00_est_focal <- read.table(file = paste0(YN00_dir, OLclade_oi,"_", paste0("jump", jumpNum_oi), "_HEG_YN00_estimates_focal.tab"), stringsAsFactors = F)
    }, error=function(e){
      return("notFound")
    }
  )
  if(f_out=="notFound") next
  pairwise_yn00_est_focal$deltaGC3 <- all_features_df_sel[dataset_oi,"deltaGC3"]  
  
  pairwise_yn00_est_related <- read.table(file = paste0(YN00_dir, OLclade_oi,"_", paste0("jump", jumpNum_oi), "_HEG_YN00_estimates_related.tab"), stringsAsFactors = F)
  pairwise_yn00_est_related$deltaGC3 <- all_features_df_sel[dataset_oi,"deltaGC3"]
  
  all_yn00_dNdS_dS_deltaGC3 <- rbind(all_yn00_dNdS_dS_deltaGC3, pairwise_yn00_est_focal, pairwise_yn00_est_related)
}

all_yn00_dNdS_dS_deltaGC3$full_dataset <- paste0(all_yn00_dNdS_dS_deltaGC3$dataset,"_", all_yn00_dNdS_dS_deltaGC3$parent)
all_yn00_dNdS_dS_deltaGC3 <- all_yn00_dNdS_dS_deltaGC3[complete.cases(all_yn00_dNdS_dS_deltaGC3),]

deltaGC3_df <- all_features_df_sel[unique(all_yn00_dNdS_dS_deltaGC3$full_dataset), c("change_in_dNdS", "deltaGC3")]
deltaGC3_df$full_dataset <- rownames(deltaGC3_df)
deltaGC3_df$figLabels <- LETTERS[1:nrow(deltaGC3_df)]

all_yn00_dNdS_dS_deltaGC3$full_dataset <- ordered(x = all_yn00_dNdS_dS_deltaGC3$full_dataset, levels = deltaGC3_df$full_dataset)
all_yn00_dNdS_dS_deltaGC3$category <- ordered(x = all_yn00_dNdS_dS_deltaGC3$category, levels = c("focal", "sister", "other"))

plt_yn00_dNdS_vs_dS <- ggplot(data = all_yn00_dNdS_dS_deltaGC3, mapping = aes(x = dS, y = dN.dS, color=category)) + geom_point(data = subset(all_yn00_dNdS_dS_deltaGC3, category=="other"), shape=21, alpha=0.7) + geom_point(data = subset(all_yn00_dNdS_dS_deltaGC3, category=="sister"), shape=21, alpha=0.7) + geom_point(data = subset(all_yn00_dNdS_dS_deltaGC3, category=="focal"), shape=21, alpha=0.9)
  
plt_yn00_dNdS_vs_dS <- plt_yn00_dNdS_vs_dS + facet_wrap(.~fct_reorder(full_dataset, deltaGC3), ncol = 3, scales = "fixed") + geom_text(data = deltaGC3_df, mapping = aes(x = 3.2, y = 0.085, label = "Delta", group=NULL), inherit.aes = F, parse = T) + geom_text(data =  deltaGC3_df, mapping = aes(x = 5, y = 0.085, label = paste("GC3 = ", deltaGC3), group=NULL), inherit.aes = F) + geom_text(data =  deltaGC3_df, mapping = aes(x = 0.2, y = 0.085, label = figLabels, group=NULL), inherit.aes = F, size=6)

plt_yn00_dNdS_vs_dS <- plt_yn00_dNdS_vs_dS + scale_y_continuous(name = "dN/dS", breaks = c(0, 0.025, 0.05, 0.075, 0.1), labels = c(0, 0.025, 0.05, 0.075, 0.1)) + scale_color_manual(values=c("orange", "black", "purple"))

plt_yn00_dNdS_vs_dS <- plt_yn00_dNdS_vs_dS + theme_bw() + theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(), panel.border = element_rect(colour = "black"), legend.position = c(0.3, 0.77), legend.justification = c(1, 0), legend.key.width = unit(x = 1, units = "line"), legend.title.align = 0, axis.title.x = element_text(size = 16), axis.title.y = element_text(size = 16), strip.text.x = element_text(size=12), legend.spacing = unit(1, "points"))

ggsave(filename = paste0(jump_common_dir, "YN00_HEG_dNdS_allGC3_facets.svg"), plot = plt_yn00_dNdS_vs_dS, dpi = 300, width = 7.2, height = 7.2, units = "in")
ggsave(filename = paste0(jump_common_dir, "YN00_HEG_dNdS_allGC3_facets.png"), plot = plt_yn00_dNdS_vs_dS, dpi = 300, width = 7.2, height = 7.2, units = "in")
```

#### Focal vs related dNdS, summarized changes
```{r dNdS_focal_vs_related_summarized }
# HEG_mean_dNdS$grp <- sapply(HEG_mean_dNdS$dNdSRatio, function(x) ifelse(x>1.25 || x<0.75, "A", "B"))

dNdS_focal_sis_plt <- ggplot(all_features_df_sel, aes(x=`related dN/dS`, y=`focal dN/dS`, fill=`focal GC (%)` - `sister GC (%)`)) + geom_point(size=3, stroke=0.75, col="black", shape=21, alpha=0.8)#, shape=grp)) , aes(fill=gc_focal - gc_other)
dNdS_focal_sis_plt <- dNdS_focal_sis_plt + scale_x_continuous(name = "median dN/dS (related clade)") + scale_y_continuous(name = "median dN/dS (focal clade)") + scale_fill_gradient2(name=expression(Delta*"GC3"))# + scale_shape_manual(values = c(24, 21), labels=c("moderate", "nil"), name="change")
dNdS_focal_sis_plt <- dNdS_focal_sis_plt + theme(legend.title.align = 0.5, legend.position = c(0.025, 0.75), panel.background = element_rect(color = "black", linetype = "dashed", fill=NA, inherit.blank = F))
dNdS_focal_sis_plt <- dNdS_focal_sis_plt + geom_abline(slope = 1, intercept = 0, color="grey") + geom_abline(slope = 1.25, intercept = 0, color="grey", linetype="dashed") + geom_abline(slope = 0.75, intercept = 0, color="grey", linetype="dashed")

# save_plot(filename = paste0(jump_common_dir, "HEG_dNdS_yn00_focal_vs_sis.svg"), plot = dNdS_focal_sis_plt, ncol = 1, base_width = 4, base_height = 4)
# save_plot(filename = paste0(jump_common_dir, "HEG_dNdS_yn00_focal_vs_sis.png"), plot = dNdS_focal_sis_plt, ncol = 1, base_width = 4, base_height = 4)
# dNdS_focal_sis_plt

dNdSratio_deltaGC_plt <- ggplot(all_features_df_sel, aes(x=`deltaGC3`,  y=`change_in_dNdS`, color=directionGC)) + geom_point(size=2)#, aes(shape=grp))

dNdSratio_deltaGC_plt <- dNdSratio_deltaGC_plt + scale_x_continuous(name = expression(Delta*"GC3 (focal clade - sister clade)")) + scale_y_continuous(name = "change in omega\n(focal/related)", breaks = c(0.25,0.5,1,2), limits = c(0.25,2), trans = "log2")# + scale_shape_manual(values = c(24, 21), labels=c("moderate", ""), name="change")

dNdSratio_deltaGC_plt <- dNdSratio_deltaGC_plt + scale_color_manual(breaks = c("downward", "upward"), values=c(cbPalette[7], cbPalette[6]))

dNdSratio_deltaGC_plt <- dNdSratio_deltaGC_plt + theme(plot.margin = margin(0, 0, 0, 20, "points"))

dNdSratio_deltaGC_plt <- dNdSratio_deltaGC_plt + geom_hline(yintercept = 1, color="black", linetype="dashed")
dNdSratio_deltaGC_plt

y_hist <- ggplot(data = all_features_df_sel, mapping = aes(x=change_in_dNdS, fill=directionGC)) + geom_histogram(position = "identity", alpha=0.4, binwidth = 0.5, boundary=1) + scale_x_continuous(name = "", trans = "log2", labels = NULL) + scale_fill_manual(values=c(cbPalette[7], cbPalette[6])) + rremove("legend") + rremove("y.axis") + rremove("y.ticks") + guides(fill=F) + theme(plot.margin = margin()) + theme_void()
y_hist

aligned_y_hist <- align_plots(y_hist + coord_flip(), dNdSratio_deltaGC_plt, align = "v")[[1]]

dNdSratio_deltaGC_marginal_plt <- plot_grid(dNdSratio_deltaGC_plt, aligned_y_hist, ncol=2, rel_widths = c(1,0.5), align = "hv")

save_plot(filename = paste0(jump_common_dir, "HEG_dNdS_yn00_GC3_comparisons.svg"), plot = dNdSratio_deltaGC_marginal_plt, ncol = 2, base_width = 3.5, base_height = 4)
save_plot(filename = paste0(jump_common_dir, "HEG_dNdS_yn00_GC3_comparisons.png"), plot = dNdSratio_deltaGC_marginal_plt, ncol = 2, base_width = 3.5, base_height = 4)
```

#### dNdS vs GC3
```{r dNdS_vs_GC3}
related_dNdS_GC_plt <- ggplot(all_features_df_sel, aes(x=`related GC3`,  y=`related dN/dS`)) + geom_point(size=2, color=grey(0.4))#, aes(shape=grp))

related_dNdS_GC_plt <- related_dNdS_GC_plt + scale_x_continuous(name = "GC3") + scale_y_continuous(name = "dN/dS")

related_dNdS_GC_plt <- related_dNdS_GC_plt + stat_smooth(data = all_features_df_sel, mapping = aes(x=`related GC3`,  y=`related dN/dS`), method="lm", formula=y~x+I(x^2), size=1, color=grey(0.7), se = F)
related_dNdS_GC_plt

focal_dNdS_GC_plt <- ggplot(all_features_df_sel, aes(x=`focal GC3`,  y=`focal dN/dS`)) + geom_point(size=2, color=grey(0.4))#, aes(shape=grp))

focal_dNdS_GC_plt <- focal_dNdS_GC_plt + scale_x_continuous(name = "GC3") + scale_y_continuous(name = "dN/dS")

focal_dNdS_GC_plt <- focal_dNdS_GC_plt + stat_smooth(data = all_features_df_sel, mapping = aes(x=`focal GC3`,  y=`focal dN/dS`), method="lm", formula=y~x+I(x^2), size=1, color=grey(0.7), se = F)
focal_dNdS_GC_plt

dNdS_GC_df <- cbind.data.frame(c(all_features_df_sel$`focal GC3`, all_features_df_sel$`related GC3`), c(all_features_df_sel$`focal dN/dS`, all_features_df_sel$`related dN/dS`), c(all_features_df_sel$`focal dS`, all_features_df_sel$`related dS`))
colnames(dNdS_GC_df) <- c("GC3", "dNdS", "dS")

dNdS_GC_plt <- ggplot(dNdS_GC_df, aes(x=`GC3`,  y=`dNdS`)) + geom_point(size=2, color=grey(0.4))#, aes(shape=grp))

dNdS_GC_plt <- dNdS_GC_plt + scale_x_continuous(name = "GC3") + scale_y_continuous(name = "dN/dS")

dNdS_GC_plt <- dNdS_GC_plt + stat_smooth(data = dNdS_GC_df, mapping = aes(x=`GC3`,  y=`dNdS`), method="lm", formula=y~x+I(x^2), size=1, color=grey(0.7), se = F)
dNdS_GC_plt

dNdS_dS_plt <- ggplot(dNdS_GC_df, aes(x=`dS`,  y=`dNdS`, color=`GC3`)) + geom_point(size=2)#, aes(shape=grp))

dNdS_dS_plt <- dNdS_dS_plt + scale_x_continuous(name = "dS") + scale_y_continuous(name = "dN/dS") + scale_color_gradient()

# dNdS_dS_plt <- dNdS_dS_plt + stat_smooth(data = dNdS_GC_df, mapping = aes(x=`dS`,  y=`dNdS`), method="lm", formula=y~x+I(x^2), size=1, color=grey(0.7), se = F)
dNdS_dS_plt

dS_GC3_plt <- ggplot(dNdS_GC_df, aes(x=`GC3`,  y=`dS`)) + geom_point(size=2, color=grey(0.4))#, aes(shape=grp))

dS_GC3_plt <- dS_GC3_plt + scale_x_continuous(name = "GC3") + scale_y_continuous(name = "dS")

# dNdS_dS_plt <- dNdS_dS_plt + stat_smooth(data = dNdS_GC_df, mapping = aes(x=`dS`,  y=`dNdS`), method="lm", formula=y~x+I(x^2), size=1, color=grey(0.7), se = F)
dS_GC3_plt

dNdS_GC3_all_plt <- plot_grid(plot_grid(related_dNdS_GC_plt, focal_dNdS_GC_plt, dNdS_GC_plt, align = "hv", ncol = 1, labels = "AUTO", label_size = 20))

save_plot(filename = paste0(jump_common_dir, "HEG_dNdS_dS_yn00_comparisons.svg"), plot = dNdS_GC3_all_plt, ncol = 1, base_width = 4, base_height = 8)
save_plot(filename = paste0(jump_common_dir, "HEG_dNdS_dS_yn00_comparisons.png"), plot = dNdS_GC3_all_plt, ncol = 1, base_width = 4, base_height = 8)
```

#### write
```{r write_dNdS_table }
sel_jumps <- setdiff(rownames(HEG_mean_dNdS)[which(HEG_mean_dNdS$dNdSRatio>1.25 | HEG_mean_dNdS$dNdSRatio<0.75)], symbiont_clades)
sel_jumps <- sel_jumps[order(HEG_mean_dNdS[sel_jumps, "dNdSRatio"], decreasing = T)]
req_df <- all_features_df_sel[sel_jumps,]

write.table(req_df, file = paste0(jump_common_dir, "selected_dNdS_comparison_table.tab"))
View(req_df)
```

```{r combine_all_data_dNdS_sets }
HEG_mean_dNdS <- subset(HEG_mean_dNdS, !is.na(dNdSRatio))
rownames(HEG_mean_dNdS) <- HEG_mean_dNdS$focal_jump

all_data_dNdS_sets <- all_features_df_sel[rownames(HEG_mean_dNdS),]
```


#### 16s vs CUB
```{r 16s_bLength_vs_CUB }
common_jumps <- setdiff(intersect(rownames(all_16s_comparisons), rownames(all_CUB_comparisons)), c(symbiont_clades))#, rownames(HEG_mean_dNdS)))

all_16s_CUB_comparisons <- cbind.data.frame(all_16s_comparisons[common_jumps,c("focal_GC", "sis_GC", "SixteenSratio")], all_CUB_comparisons[common_jumps, c("focal_CUB", "sis_CUB","CUBratio", "focal_rRNA_avg", "sis_rRNA_avg", "focal_tRNA_avg", "sis_tRNA_avg")])

CUB_16s_deltaGC_plt <- ggplot(subset(all_16s_CUB_comparisons, CUBratio>0 & SixteenSratio>10^(-3)), aes(x = CUBratio, y = SixteenSratio, fill=focal_GC - sis_GC)) + geom_point(size=3, shape=21, stroke=0.5) + scale_x_continuous(name="fold change in CUB\n(focal/related)") + scale_y_continuous(name="fold change in 16s branch length\n(focal/sister)", trans = "log2") + scale_fill_gradient2() + labs(fill=expression(Delta*"GC")) + theme(legend.title.align = 0.5, legend.position = c(0.025,0.25))

CUB_16s_deltaGC_plt <- CUB_16s_deltaGC_plt + geom_hline(yintercept = 2, col="grey", linetype="dashed") + geom_hline(yintercept = 0.5, col="grey", linetype="dashed") + geom_vline(xintercept = 0.75, col="grey", linetype="dashed") + geom_vline(xintercept = 1.25, col="grey", linetype="dashed")

CUB_16s_deltaGC_plt <- CUB_16s_deltaGC_plt + annotate("text", x=0,y=3,label="Reduced efficiency\nof selection", col="black", hjust=0) + annotate("text", x=2.25, y=0.05,label="Enhanced\nefficiency of selection", col="black", hjust=1)

CUB_16s_deltaGC_plt
save_plot(filename = paste0(jump_common_dir, "16s_vs_CUB.png"), plot = CUB_16s_deltaGC_plt, ncol = 1, base_width = 6.5, base_height = 5)
```

#### write
```{r write_16s_CUB_table }
sel_jumps <- setdiff(rownames(all_16s_CUB_comparisons)[which((all_16s_CUB_comparisons$SixteenSratio>2 & all_16s_CUB_comparisons$CUBratio<0.75) | (all_16s_CUB_comparisons$SixteenSratio<0.5 & all_16s_CUB_comparisons$CUBratio>1.25))], symbiont_clades)

req_cols <- c("focal GC (%)", "sister GC (%)", "focal branch length", "sister branch length", "focal CUB", "sister CUB", "focal genome size (Mb)", "sister genome size (Mb)", "focal coding density (%)", "sister coding density (%)", "focal pseudogenes (%)", "sister pseudogenes (%)", "focal IGR length (bp)", "sister IGR length (bp)")

req_df <- all_features_df[sel_jumps, req_cols]

write.table(req_df, file = paste0(jump_common_dir, "selected_16s_CUB_comparison_table.tab"))
View(req_df)
```

#### CUB vs deltaGC3
```{r CUB_analysis }
CUBratio_deltaGC_plt <- ggplot(data = all_CUB_comparisons[-c(which(rownames(all_CUB_comparisons) %in% symbiont_clades), which(is.infinite(all_CUB_comparisons$CUBratio))),], aes(x=focal_GC - sis_GC, y=CUBratio)) + geom_point(shape=21, color="black", size=3) + scale_x_continuous(name = expression(Delta*"GC (focal clade - sister clade)")) + scale_y_continuous(name = "focal CUB/\n related CUB", trans = "log2", breaks = c(0.0625, 0.125, 0.25, 0.5,1,2,4, 8), labels = c("1/16", "1/8", "1/4", "1/2", "1", "2", "4", "8"), limits = c(1/16,6)) + geom_hline(yintercept = 1, color="grey") + geom_hline(yintercept = 0.75, color="grey", linetype="dashed") + geom_hline(yintercept = 1.5, color="grey", linetype="dashed") + geom_hline(yintercept = 2, color="grey", linetype="dashed") + geom_hline(yintercept = 0.5, color="grey", linetype="dashed")

save_plot(filename = paste0(jump_common_dir, "CUBratio_vs_deltaGC.png"), plot = CUBratio_deltaGC_plt, ncol = 1, base_width = 6.5, base_height = 5)
CUBratio_deltaGC_plt
```

#### write
```{r get_CUB_exclusive }
all_CUB_exclusive_comparisons <- all_CUB_comparisons[setdiff(rownames(all_CUB_comparisons), c(rownames(HEG_mean_dNdS), rownames(all_16s_comparisons))),]

sel_jumps <- setdiff(rownames(all_CUB_exclusive_comparisons)[which((all_CUB_exclusive_comparisons$CUBratio<0.75 | all_CUB_exclusive_comparisons$CUBratio>1.25) & all_CUB_exclusive_comparisons$focal_CUB>0 & all_CUB_exclusive_comparisons$sis_CUB>0 & !is.infinite(all_CUB_exclusive_comparisons$focal_CUB) & !is.infinite(all_CUB_exclusive_comparisons$sis_CUB))], symbiont_clades)

req_cols <- c("focal GC (%)", "sister GC (%)", "focal CUB", "sister CUB", "focal genome size (Mb)", "sister genome size (Mb)", "focal coding density (%)", "sister coding density (%)", "focal pseudogenes (%)", "sister pseudogenes (%)", "focal IGR length (bp)", "sister IGR length (bp)")

req_df <- all_features_df[sel_jumps, req_cols] 

write.table(req_df[order(req_df$`focal CUB`/req_df$`sister CUB`),], file = paste0(jump_common_dir, "selected_CUB_exclusive_comparison_table.tab"))

CUB_exclusive_plt <- ggplot(data = all_CUB_exclusive_comparisons[!is.infinite(all_CUB_exclusive_comparisons$focal_CUB),], aes(x=focal_GC - sis_GC, y=focal_CUB/sis_CUB)) + geom_point(color="grey", size=3) + scale_x_continuous(name = expression(Delta*"GC (focal clade - sister clade)")) + scale_y_continuous(name = "focal CUB/\n related CUB", trans = "log2", breaks = c(0.25, 0.5,1,2,4, 8), labels = c("1/4", "1/2", "1", "2", "4", "8"), limits = c(1/4,6)) + geom_hline(yintercept = 1, color="grey") + geom_hline(yintercept = 0.75, color="grey", linetype="dashed") + geom_hline(yintercept = 1.25, color="grey", linetype="dashed")# + geom_hline(yintercept = 2, color="grey", linetype="dashed") + geom_hline(yintercept = 0.5, color="grey", linetype="dashed")

save_plot(filename = paste0(jump_common_dir, "CUBratio_exclusive_vs_deltaGC.png"), plot = CUB_exclusive_plt, ncol = 1, base_width = 4.5, base_height = 4)
CUB_exclusive_plt
```

#### deltacDensity vs deltagSize
```{r all_cDensity_gSize_deltaGC }
gSize_cDensity_deltaGC_plt <- ggplot(all_feature_changes, aes(x=`gSize`, y=`cDensity`,fill=`GC`, shape=`endoSymb_status`)) + geom_point(size=3)

gSize_cDensity_deltaGC_plt <- gSize_cDensity_deltaGC_plt + scale_x_continuous(name="fold change in genome size", trans = "log2") + scale_y_continuous(name = "change in coding density\n(% pts)") + scale_fill_gradient2(name=expression(Delta*"GC")) + theme(legend.title.align = 0.5, legend.position = c(0.8,0.25), legend.key.height = unit(1.5,"picas"), legend.text.align = 1) + scale_shape_manual(values = c(21, 24), labels=c("free-living", "endosymbiont"), name="dataset")

gSize_cDensity_deltaGC_plt <- gSize_cDensity_deltaGC_plt + geom_vline(xintercept = 0.75, col="grey", linetype="dashed") +  geom_hline(yintercept = -5, col="grey", linetype="dashed") + geom_hline(yintercept = 5, col="grey", linetype="dashed")# +geom_vline(xintercept = 1.5, col="grey", linetype="dashed")

gSize_cDensity_deltaGC_plt <- gSize_cDensity_deltaGC_plt + annotate("text", x=0.2,y=-25,label="Reduced efficiency\nof selection", col="black", hjust=0) + annotate("text", x=0.2, y=7.5,label="Streamlining\n selection", col="black", hjust=0)

save_plot(filename = paste0(jump_common_dir, "cDensity_vs_gSize_deltaGC_all.png"), plot = gSize_cDensity_deltaGC_plt, ncol = 1, base_width = 5.6, base_height = 4.5)

gSize_cDensity_deltaGC_plt

sel_gSize_cDensity <- which(all_features_df_sel$`genome size ratio (checkM)`<0.75 & (all_features_df_sel$`focal coding density (%)` - all_features_df_sel$`sister coding density (%)`)<(-5))

write.table(all_features_df_sel[sel_gSize_cDensity,], file = paste0(jump_common_dir,"selected_gSize_cDensity_comparison_table.tab"))

```

```{r all_gSize_deltaGC_cDensity }
gSize_cDensity_deltaGC_plt <- ggplot(all_feature_changes, aes(x=`GC`, y=`gSize`,fill=`cDensity`, shape=`endoSymb_status`)) + geom_point(size=3)

gSize_cDensity_deltaGC_plt <- gSize_cDensity_deltaGC_plt + scale_x_continuous(name=expression(Delta*"GC")) + scale_y_continuous(name = "fold change in genome size", trans = "log2") + scale_fill_gradient2(name=expression(Delta*"coding density\n(% pts)")) + theme(legend.title.align = 0.5, legend.position = c(0.8,0.25), legend.key.height = unit(1.5,"picas"), legend.text.align = 1) + scale_shape_manual(values = c(21, 24), labels=c("free-living", "endosymbiont"), name="dataset")

gSize_cDensity_deltaGC_plt <- gSize_cDensity_deltaGC_plt + geom_hline(yintercept = 0.75, col="grey", linetype="dashed")# +geom_hline(yintercept = 1.5, col="grey", linetype="dashed")

# gSize_cDensity_deltaGC_plt <- gSize_cDensity_deltaGC_plt + annotate("text", x=0.2,y=-25,label="Reduced efficiency\nof selection", col="black", hjust=0) + annotate("text", x=0.2, y=7.5,label="Streamlining\n selection", col="black", hjust=0)

save_plot(filename = paste0(jump_common_dir, "gSize_vs_deltaGC_cDensity_all.png"), plot = gSize_cDensity_deltaGC_plt, ncol = 1, base_width = 5.6, base_height = 4.5)
```

```{r plot_genome_comparisons }
# all_genome_IGR_comparisons <- cbind.data.frame(all_genome_comparisons,all_IGR_comparisons[,c("focal_igrLength", "sis_igrLength")])
# gSize_deltaGC_cDensity_plt <- ggplot(all_genome_comparisons[setdiff(rownames(all_genome_comparisons), symbiont_clades),], aes(x=focal_GC - sis_GC, y=focal_gSize/sis_gSize,fill=focal_cDensity - sis_cDensity)) + geom_point(shape=21, size=3)
# 
# gSize_deltaGC_cDensity_plt <- gSize_deltaGC_cDensity_plt + scale_x_continuous(name=expression(Delta*"GC"), breaks = seq(-30,30,10)) + scale_y_continuous(name="fold change in genome size", trans = "log2") + scale_fill_gradient2(name = "change in coding\ndensity (% pts)") + theme(legend.title.align = 0, legend.position = c(0.025,0.85), legend.title = element_text(size = 10), legend.text = element_text(size = 10), legend.key.height = unit(0.85,"picas"))
# 
# gSize_deltaGC_cDensity_plt <- gSize_deltaGC_cDensity_plt + geom_hline(yintercept = 0.5, col="grey", linetype="dashed") + geom_hline(yintercept = 1.5, col="grey", linetype="dashed")# + geom_vline(xintercept = -5, col="grey", linetype="dashed") + geom_vline(xintercept = 5, col="grey", linetype="dashed")
# 
# save_plot(filename = paste0(jump_common_dir, "gSize_vs_deltaGC_cDensity.png"), plot = gSize_deltaGC_cDensity_plt, ncol = 1, base_width = 5, base_height = 4)
# gSize_deltaGC_cDensity_plt

gSize_cDensity_deltaGC_plt <- ggplot(all_features_df_sel, aes(x=`focal genome size, checkM (Mb)`/`sister genome size, checkM (Mb)`, y=`focal coding density (%)` - `sister coding density (%)`,fill=`focal GC (%)` - `sister GC (%)`)) + geom_point(shape=21, size=3)

gSize_cDensity_deltaGC_plt <- gSize_cDensity_deltaGC_plt + scale_x_continuous(name="fold change in genome size", trans = "log2") + scale_y_continuous(name = "change in coding density\n(% pts)", limits = c(-35,12)) + scale_fill_gradient2(name=expression(Delta*"GC")) + theme(legend.title.align = 0.5, legend.position = c(0.8,0.25), legend.key.height = unit(1.5,"picas"), legend.text.align = 1)

gSize_cDensity_deltaGC_plt <- gSize_cDensity_deltaGC_plt + geom_vline(xintercept = 0.75, col="grey", linetype="dashed") +  geom_hline(yintercept = -5, col="grey", linetype="dashed") + geom_hline(yintercept = 5, col="grey", linetype="dashed")# +geom_vline(xintercept = 1.5, col="grey", linetype="dashed")

gSize_cDensity_deltaGC_plt <- gSize_cDensity_deltaGC_plt + annotate("text", x=0.35,y=-25,label="Reduced efficiency\nof selection", col="black", hjust=0) + annotate("text", x=0.35, y=8.5,label="Streamlining\n selection", col="black", hjust=0)

save_plot(filename = paste0(jump_common_dir, "cDensity_vs_gSize_deltaGC.png"), plot = gSize_cDensity_deltaGC_plt, ncol = 1, base_width = 5.6, base_height = 4.5)

gSize_cDensity_deltaGC_plt

sel_gSize_cDensity <- which(all_features_df_sel$`genome size ratio (checkM)`<0.75 & (all_features_df_sel$`focal coding density (%)` - all_features_df_sel$`sister coding density (%)`)<(-5))

write.table(all_features_df_sel[sel_gSize_cDensity,], file = paste0(jump_common_dir,"selected_gSize_cDensity_comparison_table.tab"))

View(all_features_df_sel[sel_gSize_cDensity,])
```

#### deltaIGR vs deltagSize
```{r plot_gSize_IGR_comparisons }
gSize_IGR_deltaGC_plt <- ggplot(all_features_df_sel, aes(x=`focal genome size, checkM (Mb)`/`sister genome size, checkM (Mb)`, y=`focal IGR length (bp)` - `sister IGR length (bp)`,fill=`focal GC (%)` - `sister GC (%)`)) + geom_point(shape=21, size=3)

gSize_IGR_deltaGC_plt <- gSize_IGR_deltaGC_plt + scale_x_continuous(name="fold change in genome size", trans = "log2") + scale_y_continuous(name = "change in IGR length\n(bp)", limits = c(-35,100)) + scale_fill_gradient2(name=expression(Delta*"GC")) + theme(legend.title.align = 0.5, legend.position = c(0.8,0.75), legend.key.height = unit(1.5,"picas"))

gSize_IGR_deltaGC_plt <- gSize_IGR_deltaGC_plt + geom_vline(xintercept = 0.75, col="grey", linetype="dashed") + geom_hline(yintercept = -20, col="grey", linetype="dashed") + geom_hline(yintercept = 20, col="grey", linetype="dashed")# + geom_vline(xintercept = 1.5, col="grey", linetype="dashed") 

gSize_IGR_deltaGC_plt <- gSize_IGR_deltaGC_plt + annotate("text", x=0.25,y=-30,label="Streamlining\nselection", col="black", hjust=0) + annotate("text", x=0.25, y=60,label="Reduced efficiency\nof selection", col="black", hjust=0)

save_plot(filename = paste0(jump_common_dir, "IGR_vs_gSize_deltaGC.png"), plot = gSize_IGR_deltaGC_plt, ncol = 1, base_width = 5.6, base_height = 4.5)

gSize_IGR_deltaGC_plt

sel_gSize_IGR <- which(all_features_df_sel$`genome size ratio (checkM)`<0.75 & ((all_features_df_sel$`focal IGR length (bp)` - all_features_df_sel$`sister IGR length (bp)`)<(-20) | (all_features_df_sel$`focal IGR length (bp)` - all_features_df_sel$`sister IGR length (bp)`)>20))

write.table(all_features_df_sel[sel_gSize_IGR,], file = paste0(jump_common_dir,"selected_gSize_IGR_comparison_table.tab"))

View(all_features_df_sel[sel_gSize_IGR,])
```
