---
title: "GTDB_tree_pruning"
output: html_notebook
---

#### Background
I obtained a large phylogeny (https://data.ace.uq.edu.au/public/gtdb/release80/bac120_r80.tree) representing all bacterial genomes sequence until 2017 from the Genome Taxonomy Database. It has about ~20,000 genomes incorporating representatives of a larger set of ~97,000 genomes. However, even in this phylogeny, many species are represented by multiple strains, which add only redundant information about GC content. Moreover, downstream analysis workflows used by me cannot handle such large pylogenies. Therefore, I decided to prune the phylogeny such that we remove redundant genomes that are related to each other more than a threshold.

The metadata such as full taxonomic nomenclature and GC content was obtained from- (https://data.ace.uq.edu.au/public/gtdb/release80/bac_metadata_r80.tsv).

To figure out an appropriate threshold, I will first subsample the Enterobacteriale subclade of the original tree (~20,000 taxa).

#### Choosing the Enterobacteriale subclade
I need to read the full tree and the metadata, then order the metadata according to the tips, and figure out the extent of the Enterobacterial clade.
```{r basic_setup }
require(phytools)
require(rncl)

data_folder <- "../Input_data/phylogenies/"
GTDB_22k_tree <- read_newick_phylo(file = paste0(data_folder, "GTDB_bac120_r80.tree"))
GTDB_metadata <- read.delim(file = paste0(data_folder,"GTDB_bac_metadata_r80.tsv"), header = T, sep = "\t", row.names = 1, stringsAsFactors = F)

GTDB_metadata_22k <- GTDB_metadata[GTDB_22k_tree$tip.label,]
Enterobacteriale_taxa <- rownames(GTDB_metadata_22k)[grep(pattern = "Enterobacteria", x = GTDB_metadata_22k[,"gtdb_taxonomy"], value = F, fixed = T)]
MRCA_Enterobacteria <- findMRCA(tree = GTDB_22k_tree, tips = Enterobacteriale_taxa, type = "node")
descendant_Enterobacteria <- getDescendants(tree = GTDB_22k_tree, node = MRCA_Enterobacteria)
Enterobacteria_tree <- extract.clade(phy = GTDB_22k_tree, node = MRCA_Enterobacteria)
Enterobacteria_tree_labeled <- Enterobacteria_tree
Enterobacteria_tree_labeled$tip.label <- sapply(Enterobacteria_tree$tip.label, function(x) GTDB_metadata[x, "gtdb_taxonomy"])
plot(Enterobacteria_tree_labeled)
```

This gives us a nice subtree of Enterobacteria, with multiple genomes representing the well studied species. Lets prune this tree with various thresholds to check which threshold is appropriate for us.

First, the general pruning function.
```{r pruning_function }
prune_tree <- function(tree_pruned, dist_cutoff){
    pruning_done <- F
    iter_oi <- 1
  
  while(pruning_done==F){
    print(paste0("beginning iteration number:     ", as.character(iter_oi)))
    iter_oi <- iter_oi + 1
    
    #get parameters for remaining tree
    n_tips <- length(tree_pruned$tip.label)
    
    edges_oi <- cbind(tree_pruned$edge,tree_pruned$edge.length) ##all edges and their lengths
    edges_oi <- subset(edges_oi,edges_oi[,3]<dist_cutoff) ##here we determine the edges that are shorter than dist cutoff
    edges_oi <- subset(edges_oi,edges_oi[,2]<length(tree_pruned$tip.label)) ##here we remove edges that are deep in the tree i.e only deal with edges whose desccendants are the tips of the tree; this way we cut down on the number of nodes the program analyzes in the next step
    int_nodes <- edges_oi[,1]
    
    for(int_node in int_nodes){
      node_height_oi <- nodeheight(tree_pruned, int_node)
      all_desc <- getDescendants(tree_pruned, int_node)
      leaves <- all_desc[which(all_desc %in% 1:n_tips)]
      dist_leaves <- sapply(leaves, function(x) nodeheight(tree_pruned,x)-node_height_oi)
      node_oi <- int_node
      if(length(which(dist_leaves<dist_cutoff))>1) break  #break if you find a node with > 1 descendant worth pruning
    }
    print(paste0("removing descendants of ", as.character(node_oi)))
    
    keep_leaves <- leaves[which(dist_leaves>dist_cutoff)] #we will keep all tips which are on edges longer than the threshold
    keep_leaves <- c(keep_leaves, leaves[sample(which(dist_leaves<dist_cutoff), 1)]) #sample one from tips which are within the threshold
    drop_leaves <- setdiff(leaves, keep_leaves) #remaining tips to be dropped
    print("dropped leaves...")
    print(as.character(tree_pruned$tip.label[drop_leaves]))
    tree_pruned <- drop.tip(tree_pruned, drop_leaves)
    
    if(int_node==tail(int_nodes,1)) {
      return(tree_pruned)
      break #break if you have scanned all nodes and did not find a node with descendants within threshold
    }
  }
}
```

Lets prune with various thresholds in parallel-
```{r pruning_thresholds }
require(parallel)

pruned_trees <- mclapply(c(0.1,0.05, 0.01), function(x, y) prune_tree(x,y), x=Enterobacteria_tree, mc.preschedule = F, mc.cores = 3)
```

This seems to happen in a jiffy, lets check the trees-
```{r visualize_pruned_Entero_trees }
par(mfrow=c(1,3))
for(pruned_tree in pruned_trees){
  pruned_tree_labeled <- pruned_tree
  pruned_tree_labeled$tip.label <- sapply(pruned_tree$tip.label, function(x) GTDB_metadata_22k[x, "gtdb_taxonomy"])
  plot(pruned_tree_labeled, cex=0.2)
}
```

Pruning at threshold of 0.01 seems to preserve many distinct species, but not strains. Lets prune the tree with this threshold.
```{r pruned_GTDB_22k_0.01 }
GTDB_22k_tree_pruned0.01 <- prune_tree(tree_pruned = GTDB_22k_tree, dist_cutoff = 0.01)
```

```{r write_pruned_tree_0.01 }
GTDB_22k_tree_pruned0.01_labeled <- GTDB_22k_tree_pruned0.01
GTDB_22k_tree_pruned0.01_labeled$tip.label <- sapply(GTDB_22k_tree_pruned0.01$tip.label, function(x) GTDB_metadata_22k[x, "gtdb_taxonomy"])
write.tree(phy = GTDB_22k_tree_pruned0.01, file = paste0(data_folder, "GTDB_bac120_r80_pruned0.01.newick"))
write.tree(phy = GTDB_22k_tree_pruned0.01_labeled, file = paste0(data_folder, "GTDB_bac120_r80_pruned0.01_labeled.newick"))
GTDB_metadata_14k <- GTDB_metadata_22k[GTDB_22k_tree_pruned0.01$tip.label,]
write.table(x = GTDB_metadata_14k, file = paste0(data_folder, "GTDB_bac_metadata_r80_pruned0.01.tsv", quote = T, sep = "\t", row.names = T, col.names = T))
```
We can now proceed with visualizing the tree and mapping GC content on this tree.